// @vuepress-plume/plugin-search@1.0.0-rc.193/lib/node/index.js
import MiniSearch from "minisearch";
import pMap from "p-map";
import { colors, getDirname, logger, path } from "vuepress/utils";
import { addViteOptimizeDepsInclude, getFullLocaleConfig } from "@vuepress/helper";
export * from "../shared/index.js";
//#region src/node/prepareSearchIndex.ts
/** Directory path for storing search index files / 存储搜索索引文件的目录路径 */
const SEARCH_INDEX_DIR = "internal/minisearchIndex/";
/** Map of locale paths to their MiniSearch instances / 语言路径到 MiniSearch 实例的映射 */
const indexByLocales = /* @__PURE__ */ new Map();
/** Cache for index objects by file path / 按文件路径缓存索引对象 */
const indexCache = /* @__PURE__ */ new Map();
/**
* Get or create a MiniSearch index for a specific locale.
*
* 获取或创建特定语言的 MiniSearch 索引。
*
* @param locale - Locale path (e.g., '/', '/zh/') / 语言路径
* @param lang - Language code for tokenization / 用于分词的语言代码
* @param options - Search index options / 搜索索引选项
* @returns MiniSearch instance for the locale / 该语言的 MiniSearch 实例
*/
function getIndexByLocale(locale, lang, options) {
	let index = indexByLocales.get(locale);
	if (!index) {
		const segmenter = new Intl.Segmenter(lang, { granularity: "word" });
		index = new MiniSearch({
			fields: [
				"title",
				"titles",
				"text"
			],
			storeFields: ["title", "titles"],
			tokenize(text) {
				return Array.from(segmenter.segment(text)).map((s) => s.segment);
			},
			...options.miniSearch?.options
		});
		indexByLocales.set(locale, index);
	}
	return index;
}
/**
* Get or create an index cache for a specific file path.
*
* 获取或创建特定文件路径的索引缓存。
*
* @param filepath - File path to get cache for / 要获取缓存的文件路径
* @returns Array of index objects for the file / 该文件的索引对象数组
*/
function getIndexCache(filepath) {
	let index = indexCache.get(filepath);
	if (!index) {
		index = [];
		indexCache.set(filepath, index);
	}
	return index;
}
/**
* Write a placeholder search index file for development mode.
*
* 为开发模式写入占位符搜索索引文件。
*
* This is used to provide an empty index before the actual index is prepared,
* preventing errors when the search component loads before indexing completes.
*
* 用于在实际索引准备完成之前提供空索引，
* 防止搜索组件在索引完成之前加载时出错。
*
* @param app - VuePress application instance / VuePress 应用实例
*/
async function prepareSearchIndexPlaceholder(app) {
	await app.writeTemp(`${SEARCH_INDEX_DIR}index.js`, "export const searchIndex = {}");
}
/**
* Prepare search indexes for all pages in the VuePress application.
*
* 为 VuePress 应用中的所有页面准备搜索索引。
*
* This function indexes all pages concurrently, creating separate MiniSearch
* instances for each locale, and writes the indexes to temporary files.
*
* 此函数并发索引所有页面，为每个语言创建独立的 MiniSearch 实例，
* 并将索引写入临时文件。
*
* @param options - Search index preparation options / 搜索索引准备选项
* @param options.app - VuePress application instance / VuePress 应用实例
* @param options.isSearchable - Function to filter searchable pages / 过滤可搜索页面的函数
* @param options.searchOptions - MiniSearch configuration / MiniSearch 配置
*/
async function prepareSearchIndex({ app, isSearchable, searchOptions }) {
	const start = performance.now();
	indexByLocales.clear();
	indexCache.clear();
	await pMap(app.pages, (p) => indexFile(p, searchOptions, isSearchable), { concurrency: 64 });
	await writeTemp(app);
	if (app.env.isBuild) {
		indexByLocales.clear();
		indexCache.clear();
	}
	if (app.env.isDebug) logger.info(`\n[${colors.green("@vuepress-plume/plugin-search")}] prepare search time spent: ${(performance.now() - start).toFixed(2)}ms`);
}
/**
* Handle search index update when a page is modified.
*
* 当页面被修改时处理搜索索引更新。
*
* @param app - VuePress application instance / VuePress 应用实例
* @param options - Search index preparation options / 搜索索引准备选项
* @param options.page - VuePress page instance / VuePress 页面实例
* @param options.isSearchable - Function to filter searchable pages / 过滤可搜索页面的函数
* @param options.searchOptions - MiniSearch configuration / MiniSearch 配置
*/
async function onSearchIndexUpdated(app, { page, isSearchable, searchOptions }) {
	if (isSearchable && !isSearchable(page)) return;
	await indexFile(page, searchOptions, isSearchable);
	await writeTemp(app);
}
/**
* Handle search index update when a page is removed.
*
* 当页面被删除时处理搜索索引更新。
*
* @param app - VuePress application instance / VuePress 应用实例
* @param options - Search index preparation options / 搜索索引准备选项
* @param options.page - VuePress page instance / VuePress 页面实例
* @param options.isSearchable - Function to filter searchable pages / 过滤可搜索页面的函数
* @param options.searchOptions - MiniSearch configuration / MiniSearch 配置
*/
async function onSearchIndexRemoved(app, { page, isSearchable, searchOptions }) {
	if (isSearchable && !isSearchable(page)) return;
	if (page.filePathRelative) {
		const fileId = page.path;
		const locale = page.pathLocale;
		const lang = page.lang;
		const index = getIndexByLocale(locale, lang, searchOptions);
		const cache = getIndexCache(fileId);
		if (cache && cache.length) index.removeAll(cache);
		await writeTemp(app);
	}
}
/**
* Write all search indexes to temporary files.
*
* 将所有搜索索引写入临时文件。
*
* Creates separate JavaScript files for each locale's search index,
* enabling lazy loading of indexes in the client.
*
* 为每个语言的搜索索引创建独立的 JavaScript 文件，
* 支持客户端延迟加载索引。
*
* @param app - VuePress application instance / VuePress 应用实例
*/
async function writeTemp(app) {
	const records = [];
	const promises = [];
	for (const [locale] of indexByLocales) {
		const index = indexByLocales.get(locale);
		const filename = `searchBox-${locale.replace(/^\/|\/$/g, "").replace(/\//g, "_") || "default"}.js`;
		records.push(`${JSON.stringify(locale)}: () => import('@${SEARCH_INDEX_DIR}${filename}')`);
		promises.push(app.writeTemp(`${SEARCH_INDEX_DIR}${filename}`, `export default ${JSON.stringify(JSON.stringify(index) ?? {})}`));
	}
	promises.push(app.writeTemp(`${SEARCH_INDEX_DIR}index.js`, `export const searchIndex = {${records.join(",")}}${app.env.isDev ? `\n${genHmrCode("searchIndex")}` : ""}`));
	await Promise.all(promises);
}
/**
* Index a single page for search.
*
* 为单个页面建立搜索索引。
*
* Extracts content from the page, splits it into sections based on headings,
* and adds each section to the appropriate locale's MiniSearch index.
*
* 从页面提取内容，根据标题将其分割为章节，
* 并将每个章节添加到相应语言的 MiniSearch 索引中。
*
* @param page - VuePress page object / VuePress 页面对象
* @param options - MiniSearch options / MiniSearch 选项
* @param isSearchable - Function to check if page should be indexed / 检查页面是否应被索引的函数
*/
async function indexFile(page, options, isSearchable) {
	if (!page.filePath || page.frontmatter?.search === false) return;
	if (isSearchable && !isSearchable(page)) return;
	const fileId = page.path;
	const locale = page.pathLocale;
	const lang = page.lang;
	const index = getIndexByLocale(locale, lang, options);
	const cache = getIndexCache(fileId);
	const sections = splitPageIntoSections(`<h1><a href="#"><span>${page.frontmatter.title || page.title}</span></a></h1>
${page.contentRendered}`);
	if (cache && cache.length) index.removeAll(cache);
	for await (const section of sections) {
		if (!section || !(section.text || section.titles)) break;
		const { anchor, text, titles } = section;
		const id = anchor ? [fileId, anchor].join("#") : fileId;
		if (index.has(id)) if (anchor) logger.error(`${colors.green("[@vuepress-plume/plugin-search]")} duplicate heading anchor : ${colors.cyan(titles.join(" >> "))} \n at ${colors.cyan(fileId)}`);
		else logger.error(`${colors.green("[@vuepress-plume/plugin-search]")} duplicate page permalink : ${colors.cyan(fileId)}`);
		else {
			const item = {
				id,
				text,
				title: titles.at(-1),
				titles: titles.slice(0, -1)
			};
			index.add(item);
			cache.push(item);
		}
	}
}
/** Regex pattern for matching heading tags / 匹配标题标签的正则表达式 */
const headingRegex = /<h(\d*).*?>(<a.*? href="#.*?".*?>[\s\S]*?<\/a>)<\/h\1>/gi;
/** Regex pattern for extracting heading content / 提取标题内容的正则表达式 */
const headingContentRegex = /<a.*? href="#(.*?)".*?><span>([\s\S]*?)<\/span><\/a>/i;
/** Regex pattern for ignoring template content / 忽略模板内容的正则表达式 */
const ignoreHeadingRegex = /<template[^>]*>[\s\S]*<\/template>/gi;
/**
* Split HTML content into sections based on heading elements.
*
* 根据标题元素将 HTML 内容分割为章节。
*
* This generator function parses HTML content and yields section objects
* containing anchor, titles hierarchy, and searchable text.
*
* 此生成器函数解析 HTML 内容并生成包含锚点、标题层级和可搜索文本的章节对象。
*
* @param html - HTML content to split / 要分割的 HTML 内容
* @yields Section objects with anchor, titles, and text / 包含锚点、标题和文本的章节对象
*/
function* splitPageIntoSections(html) {
	const result = html.split(headingRegex);
	result.shift();
	let parentTitles = [];
	for (let i = 0; i < result.length; i += 3) {
		const level = Number.parseInt(result[i]) - 1;
		const heading = result[i + 1];
		const headingResult = headingContentRegex.exec(heading);
		const title = clearHtmlTags(headingResult?.[2] ?? "").trim();
		const anchor = headingResult?.[1] ?? "";
		const content = result[i + 2];
		if (!title || !content) continue;
		if (level === 0) parentTitles = [title];
		else parentTitles[level] = title;
		let titles = parentTitles.slice(0, level);
		titles[level] = title;
		titles = titles.filter(Boolean);
		yield {
			anchor,
			titles,
			text: getSearchableText(content)
		};
	}
}
/**
* Extract searchable text from HTML content by removing tags.
*
* 通过移除标签从 HTML 内容中提取可搜索文本。
*
* @param content - HTML content / HTML 内容
* @returns Plain text content / 纯文本内容
*/
function getSearchableText(content) {
	content = clearHtmlTags(content);
	return content;
}
/**
* Remove all HTML tags from a string.
*
* 移除字符串中的所有 HTML 标签。
*
* @param str - String containing HTML / 包含 HTML 的字符串
* @returns String with HTML tags removed / 移除 HTML 标签后的字符串
*/
function clearHtmlTags(str) {
	str = str.replace(ignoreHeadingRegex, "");
	return str.replace(/<[^>]*>/g, "");
}
/**
* Generate HMR (Hot Module Replacement) code for the search index.
*
* 为搜索索引生成 HMR（热模块替换）代码。
*
* @param m - Module name / 模块名称
* @returns HMR code string / HMR 代码字符串
*/
function genHmrCode(m) {
	const func = `update${m[0].toUpperCase()}${m.slice(1)}`;
	return `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept()
  if (__VUE_HMR_RUNTIME__.${m}) {
    __VUE_HMR_RUNTIME__.${func}(${m})
  }
}

if (import.meta.hot) {
  import.meta.hot.accept(({ ${m} }) => {
    __VUE_HMR_RUNTIME__.${func}(${m})
  })
}
`;
}
//#endregion
//#region src/node/locales/index.ts
/**
* Default locale configurations for search plugin.
*
* 搜索插件的默认语言配置。
*
* Maps language codes to their respective locale strings.
*
* 将语言代码映射到相应的本地化字符串。
*/
const SEARCH_LOCALES = [
	[["en", "en-US"], {
		placeholder: "Search",
		resetButtonTitle: "Reset search",
		backButtonTitle: "Close search",
		noResultsText: "No results for",
		footer: {
			selectText: "to select",
			selectKeyAriaLabel: "enter",
			navigateText: "to navigate",
			navigateUpKeyAriaLabel: "up arrow",
			navigateDownKeyAriaLabel: "down arrow",
			closeText: "to close",
			closeKeyAriaLabel: "escape"
		}
	}],
	[[
		"zh",
		"zh-CN",
		"zh-Hans",
		"zh-Hant"
	], {
		placeholder: "搜索文档",
		resetButtonTitle: "重置搜索",
		backButtonTitle: "关闭",
		noResultsText: "无搜索结果：",
		footer: {
			selectText: "选择",
			selectKeyAriaLabel: "输入",
			navigateText: "切换",
			navigateUpKeyAriaLabel: "向上",
			navigateDownKeyAriaLabel: "向下",
			closeText: "关闭",
			closeKeyAriaLabel: "退出"
		}
	}],
	[["zh-TW"], {
		placeholder: "搜尋文件",
		resetButtonTitle: "重設搜尋",
		backButtonTitle: "關閉",
		noResultsText: "無搜尋結果：",
		footer: {
			selectText: "選擇",
			selectKeyAriaLabel: "輸入",
			navigateText: "切換",
			navigateUpKeyAriaLabel: "向上",
			navigateDownKeyAriaLabel: "向下",
			closeText: "關閉",
			closeKeyAriaLabel: "退出"
		}
	}],
	[["de", "de-DE"], {
		placeholder: "Dokumente durchsuchen",
		resetButtonTitle: "Suche zurücksetzen",
		backButtonTitle: "Schließen",
		noResultsText: "Keine Suchergebnisse:",
		footer: {
			selectText: "Auswählen",
			selectKeyAriaLabel: "Eingabe",
			navigateText: "Wechseln",
			navigateUpKeyAriaLabel: "Nach oben",
			navigateDownKeyAriaLabel: "Nach unten",
			closeText: "Schließen",
			closeKeyAriaLabel: "Beenden"
		}
	}],
	[["fr", "fr-FR"], {
		placeholder: "Rechercher dans la documentation",
		resetButtonTitle: "Réinitialiser la recherche",
		backButtonTitle: "Fermer",
		noResultsText: "Aucun résultat trouvé :",
		footer: {
			selectText: "sélectionner",
			selectKeyAriaLabel: "Entrée",
			navigateText: "naviguer",
			navigateUpKeyAriaLabel: "haut",
			navigateDownKeyAriaLabel: "bas",
			closeText: "fermer",
			closeKeyAriaLabel: "sortie"
		}
	}],
	[["ru", "ru-RU"], {
		placeholder: "Поиск по документации",
		resetButtonTitle: "Сбросить поиск",
		backButtonTitle: "Закрыть",
		noResultsText: "Нет результатов поиска:",
		footer: {
			selectText: "Выбрать",
			selectKeyAriaLabel: "Ввод",
			navigateText: "Переключить",
			navigateUpKeyAriaLabel: "Вверх",
			navigateDownKeyAriaLabel: "Вниз",
			closeText: "Закрыть",
			closeKeyAriaLabel: "Выход"
		}
	}],
	[["ja", "ja-JP"], {
		placeholder: "ドキュメントを検索",
		resetButtonTitle: "検索をリセット",
		backButtonTitle: "閉じる",
		noResultsText: "検索結果がありません：",
		footer: {
			selectText: "選択",
			selectKeyAriaLabel: "入力",
			navigateText: "切り替え",
			navigateUpKeyAriaLabel: "上へ",
			navigateDownKeyAriaLabel: "下へ",
			closeText: "閉じる",
			closeKeyAriaLabel: "終了"
		}
	}]
];
//#endregion
//#region src/node/searchPlugin.ts
const __dirname = getDirname(import.meta.url);
/**
* Create a VuePress search plugin instance.
*
* 创建 VuePress 搜索插件实例。
*
* @param options - Plugin configuration options / 插件配置选项
* @param options.locales - Locale-specific search configurations / 特定语言的搜索配置
* @param options.isSearchable - Function to determine if a page should be indexed / 判断页面是否应被索引的函数
* @param options.searchOptions - MiniSearch options / MiniSearch 配置选项
* @returns VuePress plugin object / VuePress 插件对象
* @example
* // Basic usage
* export default {
*   plugins: [
*     searchPlugin()
*   ]
* }
*
* // With custom options
* export default {
*   plugins: [
*     searchPlugin({
*       locales: {
*         '/zh/': { placeholder: '搜索文档' }
*       },
*       isSearchable: (page) => page.path !== '/secret/'
*     })
*   ]
* }
*/
function searchPlugin({ locales = {}, isSearchable, ...searchOptions } = {}) {
	return (app) => ({
		name: "@vuepress-plume/plugin-search",
		clientConfigFile: path.resolve(__dirname, "../client/config.js"),
		define: {
			__SEARCH_LOCALES__: getFullLocaleConfig({
				app,
				name: "@vuepress-plume/plugin-search",
				default: SEARCH_LOCALES,
				config: locales
			}),
			__SEARCH_OPTIONS__: searchOptions
		},
		extendsBundlerOptions(bundlerOptions) {
			addViteOptimizeDepsInclude(bundlerOptions, app, [
				"mark.js/src/vanilla.js",
				"@vueuse/integrations/useFocusTrap",
				"minisearch"
			]);
		},
		onPrepared: async (app) => {
			if (app.env.isBuild) await prepareSearchIndex({
				app,
				isSearchable,
				searchOptions
			});
			else {
				await prepareSearchIndexPlaceholder(app);
				prepareSearchIndex({
					app,
					isSearchable,
					searchOptions
				});
			}
		},
		onPageUpdated: async (app, type, page) => {
			if (!page?.filePathRelative) return;
			if (type === "create" || type === "update") await onSearchIndexUpdated(app, {
				page,
				isSearchable,
				searchOptions
			});
			else if (type === "delete") await onSearchIndexRemoved(app, {
				page,
				isSearchable,
				searchOptions
			});
		}
	});
}
//#endregion
export { prepareSearchIndex, searchPlugin };
