// @vuepress-plume/plugin-search@1.0.0-rc.193/lib/client/composables/index.js
import { computed, shallowRef, toRef } from "vue";
import { useRouteLocale } from "vuepress/client";
import { searchIndex } from "@internal/minisearchIndex";
//#region src/client/composables/locale.ts
/**
* Default locale configuration for search.
*
* 搜索的默认语言配置。
*/
const defaultLocales = { "/": {
	placeholder: "Search",
	resetButtonTitle: "Reset search",
	backButtonTitle: "Close search",
	noResultsText: "No results for",
	footer: {
		selectText: "to select",
		selectKeyAriaLabel: "enter",
		navigateText: "to navigate",
		navigateUpKeyAriaLabel: "up arrow",
		navigateDownKeyAriaLabel: "down arrow",
		closeText: "to close",
		closeKeyAriaLabel: "escape"
	}
} };
/**
* Get locale configuration for the current route.
*
* 获取当前路由的语言配置。
*
* @param locales - Locale configuration object / 语言配置对象
* @returns Computed ref to the current locale settings / 当前语言设置的计算引用
* @example
* const locale = useLocale(locales)
* console.log(locale.value.placeholder) // 'Search' or localized value
*/
function useLocale(locales) {
	const localesRef = toRef(locales);
	const routeLocale = useRouteLocale();
	return computed(() => localesRef.value[routeLocale.value] ?? defaultLocales[routeLocale.value] ?? defaultLocales["/"]);
}
//#endregion
//#region src/client/composables/searchIndex.ts
/**
* Reactive reference to the search index data.
*
* 搜索索引数据的响应式引用。
*/
const searchIndexData = shallowRef(searchIndex);
/**
* Get the search index data for all locales.
*
* 获取所有语言的搜索索引数据。
*
* @returns Shallow ref to the search index data / 搜索索引数据的浅层引用
* @example
* const indexData = useSearchIndex()
* const localeIndex = await indexData.value['/zh/']()
*/
function useSearchIndex() {
	return searchIndexData;
}
if (__VUEPRESS_DEV__ && (import.meta.webpackHot || import.meta.hot)) __VUE_HMR_RUNTIME__.updateSearchIndex = (data) => {
	searchIndexData.value = data;
};
//#endregion
export { useLocale, useSearchIndex };
