// vuepress-plugin-md-power@1.0.0-rc.193/lib/node/index.js
import { createRequire } from "node:module";
import { tab } from "@mdit/plugin-tab";
import { addViteOptimizeDepsInclude, encodeData, ensureEndingSlash, ensureLeadingSlash, getFullLocaleConfig, isArray, isLinkAbsolute, isLinkHttp, isPlainObject } from "@vuepress/helper";
import { Buffer } from "node:buffer";
import http from "node:https";
import { URL, URLSearchParams } from "node:url";
import { attempt, camelCase, debounce, deleteKey, isBoolean, isEmptyObject, isNull, isNumber, isPlainObject as isPlainObject$1, isString, isUndefined, kebabCase, notNullish, objectMap, omit, toArray, uniqWith, withTimeout } from "@pengzhanbo/utils";
import imageSize from "image-size";
import pMap from "p-map";
import { colors, fs, getDirname, hash, logger, ora, path, tinyglobby } from "vuepress/utils";
import path$1 from "node:path";
import { ensureLeadingSlash as ensureLeadingSlash$1, isLinkHttp as isLinkHttp$1, isLinkWithProtocol, removeEndingSlash, removeLeadingSlash, resolveLocalePath } from "vuepress/shared";
import fs$1, { promises } from "node:fs";
import process from "node:process";
import container from "markdown-it-container";
import { getRandomValues, webcrypto } from "node:crypto";
import { isPackageExists, resolveModule } from "local-pkg";
import { LRUCache } from "lru-cache";
import { watch } from "chokidar";
import { customAlphabet } from "nanoid";
import { removeLeadingSlash as removeLeadingSlash$1 } from "@vuepress/shared";
import { path as path$2 } from "@vuepress/utils";
import { attrs } from "@mdit/plugin-attrs";
import { footnote } from "@mdit/plugin-footnote";
import { mark } from "@mdit/plugin-mark";
import { sub } from "@mdit/plugin-sub";
import { sup } from "@mdit/plugin-sup";
import { tasklist } from "@mdit/plugin-tasklist";
import cjsFriendly from "markdown-it-cjk-friendly";
//#region src/node/fileIcons/definitions.ts
const defaultFolder = "vscode-icons:default-folder";
const defaultFile = "vscode-icons:default-file";
const definitions = {
	named: {
		"pnpm": "vscode-icons:file-type-light-pnpm",
		"PNPM": "vscode-icons:file-type-light-pnpm",
		"npm": "logos:npm-icon",
		"NPM": "logos:npm-icon",
		"yarn": "vscode-icons:file-type-yarn",
		"Yarn": "vscode-icons:file-type-yarn",
		"bun": "vscode-icons:file-type-bun",
		"Bun": "vscode-icons:file-type-bun",
		"deno": "vscode-icons:file-type-light-deno",
		"Deno": "vscode-icons:file-type-light-deno",
		"rollup": "vscode-icons:file-type-rollup",
		"Rollup": "vscode-icons:file-type-rollup",
		"webpack": "vscode-icons:file-type-webpack",
		"Webpack": "vscode-icons:file-type-webpack",
		"vite": "vscode-icons:file-type-vite",
		"Vite": "vscode-icons:file-type-vite",
		"esbuild": "vscode-icons:file-type-esbuild",
		"Esbuild": "vscode-icons:file-type-esbuild",
		"vue": "vscode-icons:file-type-vue",
		"Vue": "vscode-icons:file-type-vue",
		"svelte": "vscode-icons:file-type-svelte",
		"Svelte": "vscode-icons:file-type-svelte",
		"sveltekit": "vscode-icons:file-type-svelte",
		"SvelteKit": "vscode-icons:file-type-svelte",
		"angular": "vscode-icons:file-type-angular",
		"Angular": "vscode-icons:file-type-angular",
		"react": "vscode-icons:file-type-reactjs",
		"React": "vscode-icons:file-type-reactjs",
		"next": "vscode-icons:file-type-light-next",
		"Next": "vscode-icons:file-type-light-next",
		"Nextjs": "vscode-icons:file-type-light-next",
		"NextJS": "vscode-icons:file-type-light-next",
		"nuxt": "vscode-icons:file-type-nuxt",
		"Nuxt": "vscode-icons:file-type-nuxt",
		"Nuxtjs": "vscode-icons:file-type-nuxt",
		"NuxtJS": "vscode-icons:file-type-nuxt",
		"solid": "logos:solidjs-icon",
		"Solid": "logos:solidjs-icon",
		"solidjs": "logos:solidjs-icon",
		"astro": "vscode-icons:file-type-light-astro",
		"Astro": "vscode-icons:file-type-light-astro",
		"vitest": "vscode-icons:file-type-vitest",
		"Vitest": "vscode-icons:file-type-vitest",
		"playwright": "vscode-icons:file-type-playwright",
		"Playwright": "vscode-icons:file-type-playwright",
		"jest": "vscode-icons:file-type-jest",
		"Jest": "vscode-icons:file-type-jest",
		"cypress": "vscode-icons:file-type-cypress",
		"Cypress": "vscode-icons:file-type-cypress",
		"docker": "vscode-icons:file-type-docker",
		"Docker": "vscode-icons:file-type-docker",
		"html": "vscode-icons:file-type-html",
		"Html": "vscode-icons:file-type-html",
		"HTML": "vscode-icons:file-type-html",
		"javascript": "vscode-icons:file-type-js-official",
		"Javascript": "vscode-icons:file-type-js-official",
		"JavaScript": "vscode-icons:file-type-js-official",
		"js": "vscode-icons:file-type-js-official",
		"JS": "vscode-icons:file-type-js-official",
		"typescript": "vscode-icons:file-type-typescript-official",
		"Typescript": "vscode-icons:file-type-typescript-official",
		"TypeScript": "vscode-icons:file-type-typescript-official",
		"ts": "vscode-icons:file-type-typescript-official",
		"TS": "vscode-icons:file-type-typescript-official",
		"css": "vscode-icons:file-type-css",
		"CSS": "vscode-icons:file-type-css",
		"less": "vscode-icons:file-type-less",
		"Less": "vscode-icons:file-type-less",
		"LESS": "vscode-icons:file-type-less",
		"scss": "vscode-icons:file-type-scss",
		"Scss": "vscode-icons:file-type-scss",
		"SCSS": "vscode-icons:file-type-scss",
		"sass": "vscode-icons:file-type-sass",
		"Sass": "vscode-icons:file-type-sass",
		"SASS": "vscode-icons:file-type-sass",
		"stylus": "vscode-icons:file-type-light-stylus",
		"Stylus": "vscode-icons:file-type-light-stylus",
		"postcss": "vscode-icons:file-type-postcss",
		"Postcss": "vscode-icons:file-type-postcss",
		"PostCSS": "vscode-icons:file-type-postcss",
		"sh": "vscode-icons:file-type-shell",
		"shell": "vscode-icons:file-type-shell",
		"Shell": "vscode-icons:file-type-shell",
		"bash": "vscode-icons:file-type-shell",
		"Bash": "vscode-icons:file-type-shell",
		"java": "vscode-icons:file-type-java",
		"Java": "vscode-icons:file-type-java",
		"JAVA": "vscode-icons:file-type-java",
		"php": "vscode-icons:file-type-php3",
		"PHP": "vscode-icons:file-type-php3",
		"c": "vscode-icons:file-type-c",
		"C": "vscode-icons:file-type-c",
		"python": "vscode-icons:file-type-python",
		"Python": "vscode-icons:file-type-python",
		"kotlin": "vscode-icons:file-type-kotlin",
		"Kotlin": "vscode-icons:file-type-kotlin",
		"go": "vscode-icons:file-type-go-gopher",
		"golang": "vscode-icons:file-type-go-gopher",
		"Go": "vscode-icons:file-type-go-gopher",
		"GoLang": "vscode-icons:file-type-go-gopher",
		"rust": "vscode-icons:file-type-rust",
		"Rust": "vscode-icons:file-type-rust",
		"zig": "vscode-icons:file-type-zig",
		"Zig": "vscode-icons:file-type-zig",
		"swift": "vscode-icons:file-type-swift",
		"Swift": "vscode-icons:file-type-swift",
		"c#": "vscode-icons:file-type-csharp",
		"CSharp": "vscode-icons:file-type-csharp",
		"c++": "vscode-icons:file-type-cpp",
		"C++": "vscode-icons:file-type-cpp",
		"ruby": "vscode-icons:file-type-ruby",
		"Ruby": "vscode-icons:file-type-ruby",
		"makefile": "vscode-icons:file-type-makefile",
		"Makefile": "vscode-icons:file-type-makefile",
		"Object-C": "vscode-icons:file-type-objectivec",
		"object-c": "vscode-icons:file-type-objectivec",
		"SQL": "vscode-icons:file-type-sql",
		"sql": "vscode-icons:file-type-sql",
		"mysql": "vscode-icons:file-type-mysql",
		"MySQL": "vscode-icons:file-type-mysql",
		"MySql": "vscode-icons:file-type-mysql",
		"pgsql": "vscode-icons:file-type-pgsql",
		"PostgreSQL": "vscode-icons:file-type-pgsql",
		"PGSQL": "vscode-icons:file-type-pgsql",
		"xml": "vscode-icons:file-type-xml",
		"XML": "vscode-icons:file-type-xml",
		"wasm": "vscode-icons:file-type-wasm",
		"webassembly": "vscode-icons:file-type-wasm",
		"WebAssembly": "vscode-icons:file-type-wasm",
		"WASM": "vscode-icons:file-type-wasm",
		"toml": "vscode-icons:file-type-light-toml",
		"Toml": "vscode-icons:file-type-light-toml",
		"TOML": "vscode-icons:file-type-light-toml",
		"yaml": "vscode-icons:file-type-light-yaml",
		"Yaml": "vscode-icons:file-type-light-yaml",
		"YAML": "vscode-icons:file-type-light-yaml"
	},
	folders: {
		"default": "vscode-icons:default-folder",
		"src": "vscode-icons:folder-type-src",
		"srcs": "vscode-icons:folder-type-src",
		"source": "vscode-icons:folder-type-src",
		"sources": "vscode-icons:folder-type-src",
		"code": "vscode-icons:folder-type-src",
		"tauri-src": "vscode-icons:folder-type-tauri",
		"dist": "vscode-icons:folder-type-dist",
		"out": "vscode-icons:folder-type-dist",
		"output": "vscode-icons:folder-type-dist",
		"release": "vscode-icons:folder-type-dist",
		"bin": "vscode-icons:folder-type-dist",
		"distribution": "vscode-icons:folder-type-dist",
		"docs": "vscode-icons:folder-type-docs",
		"doc": "vscode-icons:folder-type-docs",
		"document": "vscode-icons:folder-type-docs",
		"documents": "vscode-icons:folder-type-docs",
		"documentation": "vscode-icons:folder-type-docs",
		"post": "vscode-icons:folder-type-docs",
		"posts": "vscode-icons:folder-type-docs",
		"article": "vscode-icons:folder-type-docs",
		"articles": "vscode-icons:folder-type-docs",
		"scripts": "vscode-icons:folder-type-script",
		"script": "vscode-icons:folder-type-script",
		"node_modules": "vscode-icons:folder-type-light-node",
		"cli": "vscode-icons:folder-type-cli",
		"template": "vscode-icons:folder-type-template",
		"templates": "vscode-icons:folder-type-template",
		"theme": "vscode-icons:folder-type-theme",
		"themes": "vscode-icons:folder-type-theme",
		"color": "vscode-icons:folder-type-theme",
		"colors": "vscode-icons:folder-type-theme",
		"design": "vscode-icons:folder-type-theme",
		"designs": "vscode-icons:folder-type-theme",
		"packages": "vscode-icons:folder-type-package",
		"package": "vscode-icons:folder-type-package",
		"pkg": "vscode-icons:folder-type-package",
		"pkgs": "vscode-icons:folder-type-package",
		"shared": "vscode-icons:folder-type-shared",
		"utils": "vscode-icons:folder-type-tools",
		"util": "vscode-icons:folder-type-tools",
		"utility": "vscode-icons:folder-type-tools",
		"utilities": "vscode-icons:folder-type-tools",
		"helper": "vscode-icons:folder-type-helper",
		"helpers": "vscode-icons:folder-type-helper",
		"tools": "vscode-icons:folder-type-tools",
		"toolkit": "vscode-icons:folder-type-tools",
		"toolkits": "vscode-icons:folder-type-tools",
		"tooling": "vscode-icons:folder-type-tools",
		"devtools": "vscode-icons:folder-type-tools",
		"component": "vscode-icons:folder-type-component",
		"components": "vscode-icons:folder-type-component",
		"Component": "vscode-icons:folder-type-component",
		"Components": "vscode-icons:folder-type-component",
		"widget": "vscode-icons:folder-type-component",
		"widgets": "vscode-icons:folder-type-component",
		"fragments": "vscode-icons:folder-type-component",
		"hooks": "vscode-icons:folder-type-hook",
		"composables": "vscode-icons:folder-type-hook",
		"public": "vscode-icons:folder-type-public",
		"www": "vscode-icons:folder-type-public",
		"web": "vscode-icons:folder-type-public",
		"wwwroot": "vscode-icons:folder-type-public",
		"website": "vscode-icons:folder-type-public",
		"site": "vscode-icons:folder-type-public",
		"browser": "vscode-icons:folder-type-public",
		"browsers": "vscode-icons:folder-type-public",
		"fonts": "vscode-icons:folder-type-fonts",
		"font": "vscode-icons:folder-type-fonts",
		"images": "vscode-icons:folder-type-images",
		"image": "vscode-icons:folder-type-images",
		"imgs": "vscode-icons:folder-type-images",
		"img": "vscode-icons:folder-type-images",
		"icon": "vscode-icons:folder-type-images",
		"icons": "vscode-icons:folder-type-images",
		"ico": "vscode-icons:folder-type-images",
		"icos": "vscode-icons:folder-type-images",
		"figure": "vscode-icons:folder-type-images",
		"figures": "vscode-icons:folder-type-images",
		"fig": "vscode-icons:folder-type-images",
		"figs": "vscode-icons:folder-type-images",
		"screenshot": "vscode-icons:folder-type-images",
		"screenshots": "vscode-icons:folder-type-images",
		"screengrab": "vscode-icons:folder-type-images",
		"screengrabs": "vscode-icons:folder-type-images",
		"pic": "vscode-icons:folder-type-images",
		"pics": "vscode-icons:folder-type-images",
		"picture": "vscode-icons:folder-type-images",
		"pictures": "vscode-icons:folder-type-images",
		"photo": "vscode-icons:folder-type-images",
		"photos": "vscode-icons:folder-type-images",
		"photograph": "vscode-icons:folder-type-images",
		"photographs": "vscode-icons:folder-type-images",
		"asset": "vscode-icons:folder-type-asset",
		"assets": "vscode-icons:folder-type-asset",
		"resource": "vscode-icons:folder-type-asset",
		"resources": "vscode-icons:folder-type-asset",
		"res": "vscode-icons:folder-type-asset",
		"static": "vscode-icons:folder-type-asset",
		"report": "vscode-icons:folder-type-asset",
		"reports": "vscode-icons:folder-type-asset",
		"apis": "vscode-icons:folder-type-api",
		"api": "vscode-icons:folder-type-api",
		"restapi": "vscode-icons:folder-type-api",
		"style": "vscode-icons:folder-type-style",
		"styles": "vscode-icons:folder-type-style",
		"stylesheet": "vscode-icons:folder-type-style",
		"stylesheets": "vscode-icons:folder-type-style",
		"css": "vscode-icons:folder-type-css",
		"scss": "vscode-icons:folder-type-light-sass",
		"sass": "vscode-icons:folder-type-light-sass",
		"less": "vscode-icons:folder-type-less",
		"plugin": "vscode-icons:folder-type-plugin",
		"plugins": "vscode-icons:folder-type-plugin",
		"typings": "vscode-icons:folder-type-typings",
		"types": "vscode-icons:folder-type-typings",
		"mock": "vscode-icons:folder-type-mock",
		"i18n": "vscode-icons:folder-type-locale",
		"locales": "vscode-icons:folder-type-locale",
		"locale": "vscode-icons:folder-type-locale",
		"lang": "vscode-icons:folder-type-locale",
		"langs": "vscode-icons:folder-type-locale",
		"language": "vscode-icons:folder-type-locale",
		"languages": "vscode-icons:folder-type-locale",
		"l10n": "vscode-icons:folder-type-locale",
		"localization": "vscode-icons:folder-type-locale",
		"translation": "vscode-icons:folder-type-locale",
		"translate": "vscode-icons:folder-type-locale",
		"translations": "vscode-icons:folder-type-locale",
		"tx": "vscode-icons:folder-type-locale",
		"config": "vscode-icons:folder-type-config",
		"configs": "vscode-icons:folder-type-config",
		".config": "vscode-icons:folder-type-config",
		".configs": "vscode-icons:folder-type-config",
		"cfg": "vscode-icons:folder-type-config",
		"cfgs": "vscode-icons:folder-type-config",
		"conf": "vscode-icons:folder-type-config",
		"confs": "vscode-icons:folder-type-config",
		"configuration": "vscode-icons:folder-type-config",
		"configurations": "vscode-icons:folder-type-config",
		"setting": "vscode-icons:folder-type-config",
		"settings": "vscode-icons:folder-type-config",
		"option": "vscode-icons:folder-type-config",
		"options": "vscode-icons:folder-type-config",
		"controller": "vscode-icons:folder-type-controller",
		"controllers": "vscode-icons:folder-type-controller",
		"Controller": "vscode-icons:folder-type-controller",
		"Controllers": "vscode-icons:folder-type-controller",
		"model": "vscode-icons:folder-type-model",
		"models": "vscode-icons:folder-type-model",
		"Model": "vscode-icons:folder-type-model",
		"Models": "vscode-icons:folder-type-model",
		"service": "vscode-icons:folder-type-services",
		"services": "vscode-icons:folder-type-services",
		"Service": "vscode-icons:folder-type-services",
		"Services": "vscode-icons:folder-type-services",
		"view": "vscode-icons:folder-type-view",
		"views": "vscode-icons:folder-type-view",
		"View": "vscode-icons:folder-type-view",
		"Views": "vscode-icons:folder-type-view",
		"page": "vscode-icons:folder-type-view",
		"pages": "vscode-icons:folder-type-view",
		"html": "vscode-icons:folder-type-view",
		"app": "vscode-icons:folder-type-app",
		"apps": "vscode-icons:folder-type-app",
		"client": "vscode-icons:folder-type-client",
		"clients": "vscode-icons:folder-type-client",
		"frontend": "vscode-icons:folder-type-client",
		"frontends": "vscode-icons:folder-type-client",
		"server": "vscode-icons:folder-type-server",
		"servers": "vscode-icons:folder-type-server",
		"backend": "vscode-icons:folder-type-server",
		"backends": "vscode-icons:folder-type-server",
		"db": "vscode-icons:folder-type-db",
		"database": "vscode-icons:folder-type-db",
		"databases": "vscode-icons:folder-type-db",
		"data": "vscode-icons:folder-type-db",
		"sql": "vscode-icons:folder-type-db",
		"e2e": "vscode-icons:folder-type-e2e",
		"cypress": "vscode-icons:folder-type-light-cypress",
		"test": "vscode-icons:folder-type-test",
		"tests": "vscode-icons:folder-type-test",
		"testing": "vscode-icons:folder-type-test",
		"snapshots": "vscode-icons:folder-type-test",
		"spec": "vscode-icons:folder-type-test",
		"specs": "vscode-icons:folder-type-test",
		"lib": "vscode-icons:folder-type-library",
		"libs": "vscode-icons:folder-type-library",
		"library": "vscode-icons:folder-type-library",
		"libraries": "vscode-icons:folder-type-library",
		"vendor": "vscode-icons:folder-type-library",
		"vendors": "vscode-icons:folder-type-library",
		"third-party": "vscode-icons:folder-type-library",
		"lib64": "vscode-icons:folder-type-library",
		"include": "vscode-icons:folder-type-include",
		"inc": "vscode-icons:folder-type-include",
		"includes": "vscode-icons:folder-type-include",
		"partial": "vscode-icons:folder-type-include",
		"partials": "vscode-icons:folder-type-include",
		"inc64": "vscode-icons:folder-type-include",
		"temp": "vscode-icons:folder-type-temp",
		"tmp": "vscode-icons:folder-type-temp",
		"cache": "vscode-icons:folder-type-temp",
		"cached": "vscode-icons:folder-type-temp",
		".temp": "vscode-icons:folder-type-temp",
		".cache": "vscode-icons:folder-type-temp",
		"log": "vscode-icons:folder-type-log",
		"logs": "vscode-icons:folder-type-log",
		"logging": "vscode-icons:folder-type-log",
		".svelte-kit": "vscode-icons:folder-type-svelte",
		".git": "vscode-icons:folder-type-git",
		".github": "vscode-icons:folder-type-github",
		".gitlab": "vscode-icons:folder-type-gitlab",
		".vscode": "vscode-icons:folder-type-vscode",
		".husky": "vscode-icons:folder-type-husky",
		".idea": "vscode-icons:folder-type-idea",
		".changesets": "vscode-icons:folder-type-changesets",
		".vercel": "vscode-icons:folder-type-vercel",
		".netlify": "vscode-icons:folder-type-netlify",
		".claude": "vscode-icons:folder-type-claude",
		".cursor": "vscode-icons:folder-type-cursor",
		".gemini": "vscode-icons:folder-type-gemini",
		".windsurf": "vscode-icons:folder-type-windsurf"
	},
	files: {
		"package.json": "vscode-icons:file-type-node",
		"pnpm-debug.log": "vscode-icons:file-type-light-pnpm",
		"pnpm-lock.yaml": "vscode-icons:file-type-light-pnpm",
		"pnpm-workspace.yaml": "vscode-icons:file-type-light-pnpm",
		".pnpmfile.cjs": "vscode-icons:file-type-light-pnpm",
		"pnpmfile.js": "vscode-icons:file-type-light-pnpm",
		"biome.json": "vscode-icons:file-type-biome",
		"bun.lockb": "vscode-icons:file-type-bun",
		"COMMIT_EDITMSG": "vscode-icons:file-type-git",
		"MERGE_MSG": "vscode-icons:file-type-git",
		"karma.conf.js": "vscode-icons:file-type-karma",
		"karma.conf.cjs": "vscode-icons:file-type-karma",
		"karma.conf.mjs": "vscode-icons:file-type-karma",
		"karma.conf.coffee": "vscode-icons:file-type-karma",
		"README.md": "flat-color-icons:info",
		"README.txt": "flat-color-icons:info",
		"README": "flat-color-icons:info",
		"CHANGELOG.md": "catppuccin:changelog",
		"CHANGELOG.txt": "catppuccin:changelog",
		"CHANGELOG": "catppuccin:changelog",
		"CHANGES.md": "catppuccin:changelog",
		"CHANGES.txt": "catppuccin:changelog",
		"CHANGES": "catppuccin:changelog",
		"VERSION.md": "catppuccin:changelog",
		"VERSION.txt": "catppuccin:changelog",
		"VERSION": "catppuccin:changelog",
		"mvnw": "vscode-icons:file-type-maven",
		"pom.xml": "vscode-icons:file-type-maven",
		"tsconfig.json": "vscode-icons:file-type-tsconfig",
		"swagger.json": "vscode-icons:file-type-swagger",
		"swagger.yml": "vscode-icons:file-type-swagger",
		"swagger.yaml": "vscode-icons:file-type-swagger",
		"mime.types": "vscode-icons:file-type-light-config",
		"Jenkinsfile": "vscode-icons:file-type-jenkins",
		"babel.config.js": "vscode-icons:file-type-babel2",
		"babel.config.json": "vscode-icons:file-type-babel2",
		"babel.config.cjs": "vscode-icons:file-type-babel2",
		"BUILD": "vscode-icons:file-type-bazel",
		"BUILD.bazel": "vscode-icons:file-type-bazel",
		"WORKSPACE": "vscode-icons:file-type-bazel",
		"WORKSPACE.bazel": "vscode-icons:file-type-bazel",
		"bower.json": "vscode-icons:file-type-bower2",
		"Bower.json": "vscode-icons:file-type-bower2",
		"eslint.config.js": "vscode-icons:file-type-eslint",
		"eslint.config.ts": "vscode-icons:file-type-eslint",
		"firebase.json": "vscode-icons:file-type-firebase",
		"geckodriver": "openmoji:firefox",
		"Gruntfile.js": "vscode-icons:file-type-grunt",
		"gruntfile.babel.js": "vscode-icons:file-type-grunt",
		"Gruntfile.babel.js": "vscode-icons:file-type-grunt",
		"gruntfile.js": "vscode-icons:file-type-grunt",
		"Gruntfile.coffee": "vscode-icons:file-type-grunt",
		"gruntfile.coffee": "vscode-icons:file-type-grunt",
		"ionic.config.json": "vscode-icons:file-type-ionic",
		"Ionic.config.json": "vscode-icons:file-type-ionic",
		"ionic.project": "vscode-icons:file-type-ionic",
		"Ionic.project": "vscode-icons:file-type-ionic",
		"platformio.ini": "vscode-icons:file-type-platformio",
		"rollup.config.js": "vscode-icons:file-type-rollup",
		"sass-lint.yml": "vscode-icons:file-type-sass",
		"stylelint.config.js": "vscode-icons:file-type-light-stylelint",
		"stylelint.config.cjs": "vscode-icons:file-type-light-stylelint",
		"stylelint.config.mjs": "vscode-icons:file-type-light-stylelint",
		"yarn.clean": "vscode-icons:file-type-yarn",
		"yarn.lock": "vscode-icons:file-type-yarn",
		"webpack.config.js": "vscode-icons:file-type-webpack",
		"webpack.config.cjs": "vscode-icons:file-type-webpack",
		"webpack.config.mjs": "vscode-icons:file-type-webpack",
		"webpack.config.ts": "vscode-icons:file-type-webpack",
		"webpack.config.build.js": "vscode-icons:file-type-webpack",
		"webpack.config.build.cjs": "vscode-icons:file-type-webpack",
		"webpack.config.build.mjs": "vscode-icons:file-type-webpack",
		"webpack.config.build.ts": "vscode-icons:file-type-webpack",
		"webpack.common.js": "vscode-icons:file-type-webpack",
		"webpack.common.cjs": "vscode-icons:file-type-webpack",
		"webpack.common.mjs": "vscode-icons:file-type-webpack",
		"webpack.common.ts": "vscode-icons:file-type-webpack",
		"webpack.dev.js": "vscode-icons:file-type-webpack",
		"webpack.dev.cjs": "vscode-icons:file-type-webpack",
		"webpack.dev.mjs": "vscode-icons:file-type-webpack",
		"webpack.dev.ts": "vscode-icons:file-type-webpack",
		"webpack.prod.js": "vscode-icons:file-type-webpack",
		"webpack.prod.cjs": "vscode-icons:file-type-webpack",
		"webpack.prod.mjs": "vscode-icons:file-type-webpack",
		"webpack.prod.ts": "vscode-icons:file-type-webpack",
		"vite.config.js": "vscode-icons:file-type-vite",
		"vite.config.ts": "vscode-icons:file-type-vite",
		"vite.config.mjs": "vscode-icons:file-type-vite",
		"vite.config.cjs": "vscode-icons:file-type-vite",
		"vitest.config.ts": "vscode-icons:file-type-vitest",
		"vitest.config.js": "vscode-icons:file-type-vitest",
		"vitest.config.mjs": "vscode-icons:file-type-vitest",
		"vitest.config.cjs": "vscode-icons:file-type-vitest",
		"turbo.json": "vscode-icons:file-type-light-turbo",
		"vercel.json": "vscode-icons:file-type-light-vercel",
		"netlify.toml": "vscode-icons:file-type-light-netlify",
		"Cargo.toml": "vscode-icons:file-type-cargo",
		"Cargo.lock": "vscode-icons:file-type-cargo",
		"npm-debug.log": "vscode-icons:file-type-npm",
		"components.json": "vscode-icons:file-type-light-shadcn",
		".postcssrc": "vscode-icons:file-type-postcssconfig",
		".postcssrc.json": "vscode-icons:file-type-postcssconfig",
		".postcssrc.yaml": "vscode-icons:file-type-postcssconfig",
		".postcssrc.yml": "vscode-icons:file-type-postcssconfig",
		".postcssrc.ts": "vscode-icons:file-type-postcssconfig",
		".postcssrc.cts": "vscode-icons:file-type-postcssconfig",
		".postcssrc.mts": "vscode-icons:file-type-postcssconfig",
		".postcssrc.js": "vscode-icons:file-type-postcssconfig",
		".postcssrc.cjs": "vscode-icons:file-type-postcssconfig",
		".postcssrc.mjs": "vscode-icons:file-type-postcssconfig",
		"postcss.config.ts": "vscode-icons:file-type-postcssconfig",
		"postcss.config.cts": "vscode-icons:file-type-postcssconfig",
		"postcss.config.mts": "vscode-icons:file-type-postcssconfig",
		"postcss.config.js": "vscode-icons:file-type-postcssconfig",
		"postcss.config.cjs": "vscode-icons:file-type-postcssconfig",
		"postcss.config.mjs": "vscode-icons:file-type-postcssconfig",
		"uno.config.js": "vscode-icons:file-type-unocss",
		"uno.config.mjs": "vscode-icons:file-type-unocss",
		"uno.config.ts": "vscode-icons:file-type-unocss",
		"unocss.config.js": "vscode-icons:file-type-unocss",
		"unocss.config.mjs": "vscode-icons:file-type-unocss",
		"unocss.config.ts": "vscode-icons:file-type-unocss",
		"rolldown.config.js": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.cjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.mjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.ts": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.common.js": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.common.cjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.common.mjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.common.ts": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.dev.js": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.dev.cjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.dev.mjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.dev.ts": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.prod.js": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.prod.cjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.prod.mjs": "vscode-icons:file-type-light-rolldown",
		"rolldown.config.prod.ts": "vscode-icons:file-type-light-rolldown",
		"tsdown.config.js": "vscode-icons:file-type-tsdown",
		"tsdown.config.cjs": "vscode-icons:file-type-tsdown",
		"tsdown.config.mjs": "vscode-icons:file-type-tsdown",
		"tsdown.config.ts": "vscode-icons:file-type-tsdown",
		"tsdown.config.json": "vscode-icons:file-type-tsdown",
		".oxlintignore": "vscode-icons:file-type-oxc",
		".oxlintrc.json": "vscode-icons:file-type-oxc",
		".oxlintrc.jsonc": "vscode-icons:file-type-oxc",
		".oxfmtrc.json": "vscode-icons:file-type-oxc",
		".oxfmtrc.jsonc": "vscode-icons:file-type-oxc",
		"agents.md": "vscode-icons:file-type-agents",
		"AGENTS.md": "vscode-icons:file-type-agents",
		"claude.md": "vscode-icons:file-type-claude",
		"CLAUDE.md": "vscode-icons:file-type-claude",
		"copilot-instructions.md": "vscode-icons:file-type-copilot",
		"COPILOT-INSTRUCTIONS.md": "vscode-icons:file-type-copilot",
		"github-copilot.xml": "vscode-icons:file-type-copilot",
		"instructions.md": "vscode-icons:file-type-copilot",
		"INSTRUCTIONS.md": "vscode-icons:file-type-copilot"
	},
	extensions: {
		".astro": "vscode-icons:file-type-light-astro",
		".mdx": "vscode-icons:file-type-light-mdx",
		".cls": "vscode-icons:file-type-apex",
		".apex": "vscode-icons:file-type-apex",
		".asm": "vscode-icons:file-type-assembly",
		".s": "vscode-icons:file-type-assembly",
		".bicep": "vscode-icons:file-type-bicep",
		".bzl": "vscode-icons:file-type-bazel",
		".bazel": "vscode-icons:file-type-bazel",
		".BUILD": "vscode-icons:file-type-bazel",
		".WORKSPACE": "vscode-icons:file-type-bazel",
		".bazelignore": "vscode-icons:file-type-bazel",
		".bazelversion": "vscode-icons:file-type-bazel",
		".c": "vscode-icons:file-type-c",
		".h": "vscode-icons:file-type-c",
		".m": "vscode-icons:file-type-c",
		".cs": "vscode-icons:file-type-csharp",
		".cshtml": "vscode-icons:file-type-html",
		".aspx": "vscode-icons:file-type-aspx",
		".ascx": "vscode-icons:file-type-aspx",
		".asax": "vvscode-icons:file-type-aspx",
		".master": "vscode-icons:file-type-html",
		".cc": "vscode-icons:file-type-cpp",
		".cpp": "vscode-icons:file-type-cpp",
		".cxx": "vscode-icons:file-type-cpp",
		".c++": "vscode-icons:file-type-cpp",
		".hh": "vscode-icons:file-type-cppheader",
		".hpp": "vscode-icons:file-type-cppheader",
		".hxx": "vscode-icons:file-type-cppheader",
		".h++": "vscode-icons:file-type-cppheader",
		".mm": "vscode-icons:file-type-cpp",
		".clj": "vscode-icons:file-type-clojure",
		".cljs": "vscode-icons:file-type-clojure",
		".cljc": "vscode-icons:file-type-clojure",
		".edn": "vscode-icons:file-type-clojure",
		".cfc": "vscode-icons:file-type-cfc",
		".cfm": "vscode-icons:file-type-cfm",
		".coffee": "vscode-icons:file-type-coffeescript",
		".litcoffee": "vscode-icons:file-type-coffeescript",
		".config": "vscode-icons:file-type-config",
		".cfg": "vscode-icons:file-type-config",
		".conf": "vscode-icons:file-type-config",
		".cr": "vscode-icons:file-type-light-crystal",
		".dll": "vscode-icons:file-type-binary",
		".ecr": "vscode-icons:file-type-light-crystal",
		".slang": "vscode-icons:file-type-slang",
		".cson": "vscode-icons:file-type-json",
		".css": "vscode-icons:file-type-css",
		".css.map": "vscode-icons:file-type-cssmap",
		".sss": "vscode-icons:file-type-sss",
		".csv": "vscode-icons:file-type-excel",
		".xls": "vscode-icons:file-type-excel2",
		".xlsx": "vscode-icons:file-type-excel2",
		".cu": "vscode-icons:file-type-cuda",
		".cuh": "vscode-icons:file-type-cuda",
		".hu": "vscode-icons:file-type-cuda",
		".cake": "vscode-icons:file-type-cake",
		".ctp": "vscode-icons:file-type-cakephp",
		".d": "vscode-icons:file-type-dependencies",
		".doc": "vscode-icons:file-type-word2",
		".docx": "vscode-icons:file-type-word2",
		".ejs": "vscode-icons:file-type-ejs",
		".ex": "vscode-icons:file-type-elixir",
		".exs": "vscode-icons:file-type-elixir",
		".elm": "vscode-icons:file-type-elm",
		".ico": "vscode-icons:file-type-favicon",
		".fs": "vscode-icons:file-type-fsharp",
		".fsx": "vscode-icons:file-type-fsharp",
		".gitignore": "vscode-icons:file-type-git",
		".gitconfig": "vscode-icons:file-type-git",
		".gitkeep": "vscode-icons:file-type-git",
		".gitattributes": "vscode-icons:file-type-git",
		".gitmodules": "vscode-icons:file-type-git",
		".go": "vscode-icons:file-type-go",
		".slide": "vscode-icons:file-type-go",
		".article": "vscode-icons:file-type-go",
		".gd": "vscode-icons:file-type-godot",
		".godot": "vscode-icons:file-type-godot",
		".tres": "vscode-icons:file-type-godot",
		".tscn": "vscode-icons:file-type-godot",
		".gradle": "vscode-icons:file-type-light-gradle",
		".groovy": "vscode-icons:file-type-groovy",
		".gsp": "vscode-icons:file-type-groovy",
		".gql": "vscode-icons:file-type-graphql",
		".graphql": "vscode-icons:file-type-graphql",
		".graphqls": "vscode-icons:file-type-graphql",
		".hack": "logos:hack",
		".haml": "vscode-icons:file-type-haml",
		".handlebars": "vscode-icons:file-type-handlebars",
		".hbs": "vscode-icons:file-type-handlebars",
		".hjs": "vscode-icons:file-type-handlebars",
		".hs": "vscode-icons:file-type-haskell",
		".lhs": "vscode-icons:file-type-haskell",
		".hx": "vscode-icons:file-type-haxe",
		".hxs": "vscode-icons:file-type-haxe",
		".hxp": "vscode-icons:file-type-haxe",
		".hxml": "vscode-icons:file-type-haxe",
		".html": "vscode-icons:file-type-html",
		".jade": "file-icons:jade",
		".java": "vscode-icons:file-type-java",
		".class": "vscode-icons:file-type-java",
		".classpath": "vscode-icons:file-type-java",
		".properties": "vscode-icons:file-type-java",
		".js": "vscode-icons:file-type-js",
		".js.map": "vscode-icons:file-type-jsmap",
		".cjs": "vscode-icons:file-type-js",
		".cjs.map": "vscode-icons:file-type-jsmap",
		".mjs": "vscode-icons:file-type-js",
		".mjs.map": "vscode-icons:file-type-jsmap",
		".spec.js": "vscode-icons:file-type-light-testjs",
		".spec.cjs": "vscode-icons:file-type-light-testjs",
		".spec.mjs": "vscode-icons:file-type-light-testjs",
		".test.js": "vscode-icons:file-type-light-testjs",
		".test.cjs": "vscode-icons:file-type-light-testjs",
		".test.mjs": "vscode-icons:file-type-light-testjs",
		".es": "vscode-icons:file-type-js",
		".es5": "vscode-icons:file-type-js",
		".es6": "vscode-icons:file-type-js",
		".es7": "vscode-icons:file-type-js",
		".jinja": "vscode-icons:file-type-jinja",
		".jinja2": "vscode-icons:file-type-jinja",
		".json": "vscode-icons:file-type-json",
		".jl": "vscode-icons:file-type-julia",
		".kt": "vscode-icons:file-type-kotlin",
		".kts": "vscode-icons:file-type-kotlin",
		".dart": "vscode-icons:file-type-dartlang",
		".less": "vscode-icons:file-type-less",
		".liquid": "vscode-icons:file-type-liquid",
		".ls": "vscode-icons:file-type-livescript",
		".lua": "vscode-icons:file-type-lua",
		".markdown": "vscode-icons:file-type-markdown",
		".md": "vscode-icons:file-type-markdown",
		".mustache": "vscode-icons:file-type-light-mustache",
		".stache": "vscode-icons:file-type-light-mustache",
		".nim": "vscode-icons:file-type-light-nim",
		".nims": "vscode-icons:file-type-light-nim",
		".github-issues": "mdi:github",
		".ipynb": "vscode-icons:file-type-jupyter",
		".njk": "vscode-icons:file-type-nunjucks",
		".nunjucks": "vscode-icons:file-type-nunjucks",
		".nunjs": "vscode-icons:file-type-nunjucks",
		".nunj": "vscode-icons:file-type-nunjucks",
		".njs": "vscode-icons:file-type-nunjucks",
		".nj": "vscode-icons:file-type-nunjucks",
		".npm-debug.log": "vscode-icons:file-type-npm",
		".npmignore": "catppuccin:npm-ignore",
		".npmrc": "vscode-icons:file-type-npm",
		".ml": "vscode-icons:file-type-ocaml",
		".mli": "vscode-icons:file-type-ocaml",
		".cmx": "vscode-icons:file-type-ocaml",
		".cmxa": "vscode-icons:file-type-ocaml",
		".pl": "vscode-icons:file-type-perl",
		".php": "vscode-icons:file-type-php",
		".php.inc": "vscode-icons:file-type-php",
		".pipeline": "vscode-icons:file-type-pipeline",
		".pddl": "vscode-icons:file-type-pddl",
		".plan": "vscode-icons:file-type-pddl-plan",
		".happenings": "vscode-icons:file-type-pddl-happenings",
		".ps1": "vscode-icons:file-type-powershell",
		".psd1": "vscode-icons:file-type-powershell",
		".psm1": "vscode-icons:file-type-powershell",
		".prisma": "vscode-icons:file-type-light-prisma",
		".pug": "vscode-icons:file-type-pug",
		".pp": "vscode-icons:file-type-puppet",
		".epp": "vscode-icons:file-type-puppet",
		".purs": "vscode-icons:file-type-light-purescript",
		".py": "vscode-icons:file-type-python",
		".jsx": "vscode-icons:file-type-reactjs",
		".spec.jsx": "vscode-icons:file-type-reactjs",
		".test.jsx": "vscode-icons:file-type-reactjs",
		".cjsx": "vscode-icons:file-type-reactjs",
		".tsx": "vscode-icons:file-type-reactts",
		".spec.tsx": "vscode-icons:file-type-reactts",
		".test.tsx": "vscode-icons:file-type-reactts",
		".res": "vscode-icons:file-type-rescript",
		".resi": "vscode-icons:file-type-rescript",
		".R": "vscode-icons:file-type-r",
		".rmd": "vscode-icons:file-type-rmd",
		".rb": "vscode-icons:file-type-ruby",
		".erb": "vscode-icons:file-type-html",
		".erb.html": "vscode-icons:file-type-html",
		".html.erb": "vscode-icons:file-type-html",
		".rs": "vscode-icons:file-type-rust",
		".sass": "vscode-icons:file-type-sass",
		".scss": "vscode-icons:file-type-sass",
		".springBeans": "seti:spring",
		".slim": "vscode-icons:file-type-slim",
		".smarty.tpl": "vscode-icons:file-type-smarty",
		".tpl": "vscode-icons:file-type-smarty",
		".sbt": "vscode-icons:file-type-sbt",
		".scala": "vscode-icons:file-type-scala",
		".sol": "logos:ethereum-color",
		".styl": "vscode-icons:file-type-light-stylus",
		".svelte": "vscode-icons:file-type-svelte",
		".swift": "vscode-icons:file-type-swift",
		".sql": "vscode-icons:file-type-sql",
		".soql": "vscode-icons:file-type-sql",
		".tf": "vscode-icons:file-type-terraform",
		".tf.json": "vscode-icons:file-type-terraform",
		".tfvars": "vscode-icons:file-type-terraform",
		".tfvars.json": "vscode-icons:file-type-terraform",
		".tex": "vscode-icons:file-type-light-tex",
		".sty": "vscode-icons:file-type-light-tex",
		".dtx": "vscode-icons:file-type-light-tex",
		".ins": "vscode-icons:file-type-light-tex",
		".txt": "vscode-icons:file-type-text",
		".toml": "vscode-icons:file-type-light-toml",
		".twig": "vscode-icons:file-type-twig",
		".ts": "vscode-icons:file-type-typescript",
		".spec.ts": "vscode-icons:file-type-testts",
		".test.ts": "vscode-icons:file-type-testts",
		".vala": "vscode-icons:file-type-vala",
		".vapi": "vscode-icons:file-type-vala",
		".component": "vscode-icons:file-type-html",
		".vue": "vscode-icons:file-type-vue",
		".wasm": "vscode-icons:file-type-wasm",
		".wat": "vscode-icons:file-type-wasm",
		".xml": "vscode-icons:file-type-xml",
		".yml": "vscode-icons:file-type-light-yaml",
		".yaml": "vscode-icons:file-type-light-yaml",
		".pro": "vscode-icons:file-type-prolog",
		".zig": "vscode-icons:file-type-zig",
		".jar": "vscode-icons:file-type-jar",
		".zip": "vscode-icons:file-type-zip",
		".ai": "vscode-icons:file-type-ai",
		".psd": "vscode-icons:file-type-photoshop",
		".pdf": "vscode-icons:file-type-pdf2",
		".eot": "vscode-icons:file-type-light-font",
		".ttf": "vscode-icons:file-type-light-font",
		".woff": "vscode-icons:file-type-light-font",
		".woff2": "vscode-icons:file-type-light-font",
		".otf": "vscode-icons:file-type-light-font",
		".avif": "vscode-icons:file-type-avif",
		".gif": "vscode-icons:file-type-image",
		".jpg": "vscode-icons:file-type-image",
		".jpeg": "vscode-icons:file-type-image",
		".png": "vscode-icons:file-type-image",
		".pxm": "vscode-icons:file-type-image",
		".svg": "vscode-icons:file-type-svg",
		".svgx": "vscode-icons:file-type-image",
		".tiff": "vscode-icons:file-type-image",
		".webp": "vscode-icons:file-type-image",
		".sublime-project": "vscode-icons:file-type-sublime",
		".sublime-workspace": "vscode-icons:file-type-sublime",
		".code-search": "vscode-icons:file-type-search-result",
		".sh": "vscode-icons:file-type-shell",
		".zsh": "vscode-icons:file-type-shell",
		".fish": "vscode-icons:file-type-shell",
		".zshrc": "vscode-icons:file-type-shell",
		".bashrc": "vscode-icons:file-type-shell",
		".mov": "vscode-icons:file-type-video",
		".ogv": "vscode-icons:file-type-video",
		".webm": "vscode-icons:file-type-video",
		".avi": "vscode-icons:file-type-video",
		".mpg": "vscode-icons:file-type-video",
		".mp4": "vscode-icons:file-type-video",
		".mp3": "vscode-icons:file-type-audio",
		".ogg": "vscode-icons:file-type-audio",
		".wav": "vscode-icons:file-type-audio",
		".flac": "vscode-icons:file-type-audio",
		".bat": "vscode-icons:file-type-bat",
		".cmd": "vscode-icons:file-type-shell",
		".babelrc": "vscode-icons:file-type-light-babel",
		".babelrc.js": "vscode-icons:file-type-light-babel",
		".babelrc.cjs": "vscode-icons:file-type-light-babel",
		".bazelrc": "vscode-icons:file-type-bazel",
		".bowerrc": "vscode-icons:file-type-bower2",
		".codeclimate.yml": "vscode-icons:file-type-codeclimate",
		".eslintrc": "vscode-icons:file-type-eslint",
		".eslintrc.js": "vscode-icons:file-type-eslint",
		".eslintrc.cjs": "vscode-icons:file-type-eslint",
		".eslintrc.yaml": "vscode-icons:file-type-eslint",
		".eslintrc.yml": "vscode-icons:file-type-eslint",
		".eslintrc.json": "vscode-icons:file-type-eslint",
		".eslintignore": "vscode-icons:file-type-eslint",
		".firebaserc": "vscode-icons:file-type-firebase",
		".gitlab-ci.yml": "vscode-icons:file-type-gitlab",
		".jshintrc": "vscode-icons:file-type-jshint",
		".jscsrc": "vscode-icons:file-type-js",
		".stylelintrc": "vscode-icons:file-type-light-stylelint",
		".stylelintrc.json": "vscode-icons:file-type-light-stylelint",
		".stylelintrc.yaml": "vscode-icons:file-type-light-stylelint",
		".stylelintrc.yml": "vscode-icons:file-type-light-stylelint",
		".stylelintrc.js": "vscode-icons:file-type-light-stylelint",
		".stylelintignore": "vscode-icons:file-type-light-stylelint",
		".direnv": "vscode-icons:file-type-dotenv",
		".env": "vscode-icons:file-type-dotenv",
		".static": "vscode-icons:file-type-config",
		".editorconfig": "vscode-icons:file-type-editorconfig",
		".slugignore": "vscode-icons:file-type-config",
		".tmp": "ph:clock-bold",
		".htaccess": "vscode-icons:file-type-config",
		".key": "vscode-icons:file-type-key",
		".cert": "vscode-icons:file-type-cert",
		".cer": "vscode-icons:file-type-cert",
		".crt": "vscode-icons:file-type-cert",
		".pem": "vscode-icons:file-type-cert",
		".DS_Store": "octicon:diff-ignored-16"
	},
	partials: {
		"Gemfile": "vscode-icons:file-type-ruby",
		"gemfile": "vscode-icons:file-type-ruby",
		"dockerfile": "vscode-icons:file-type-docker",
		"Dockerfile": "vscode-icons:file-type-docker",
		"DOCKERFILE": "vscode-icons:file-type-docker",
		".dockerignore": "vscode-icons:file-type-docker",
		"docker-healthcheck": "vscode-icons:file-type-docker",
		"docker-compose.yml": "vscode-icons:file-type-docker",
		"docker-compose.yaml": "vscode-icons:file-type-docker",
		"docker-compose.override.yml": "vscode-icons:file-type-docker",
		"docker-compose.override.yaml": "vscode-icons:file-type-docker",
		"GULPFILE": "vscode-icons:file-type-gulp",
		"Gulpfile": "vscode-icons:file-type-gulp",
		"gulpfile": "vscode-icons:file-type-gulp",
		"gulpfile.js": "vscode-icons:file-type-gulp",
		"LICENSE": "vscode-icons:file-type-license",
		"LICENCE": "vscode-icons:file-type-license",
		"LICENSE.txt": "vscode-icons:file-type-license",
		"LICENCE.txt": "vscode-icons:file-type-license",
		"LICENSE.md": "vscode-icons:file-type-license",
		"LICENCE.md": "vscode-icons:file-type-license",
		"COPYING": "vscode-icons:file-type-license",
		"COPYING.txt": "vscode-icons:file-type-license",
		"COPYING.md": "vscode-icons:file-type-license",
		"COMPILING": "vscode-icons:file-type-license",
		"COMPILING.txt": "vscode-icons:file-type-license",
		"COMPILING.md": "vscode-icons:file-type-license",
		"CONTRIBUTING": "vscode-icons:file-type-license",
		"CONTRIBUTING.txt": "vscode-icons:file-type-license",
		"CONTRIBUTING.md": "vscode-icons:file-type-license",
		"MAKEFILE": "vscode-icons:file-type-makefile",
		"Makefile": "vscode-icons:file-type-makefile",
		"makefile": "vscode-icons:file-type-makefile",
		"QMAKEFILE": "vscode-icons:file-type-makefile",
		"QMakefile": "vscode-icons:file-type-makefile",
		"qmakefile": "vscode-icons:file-type-makefile",
		"OMAKEFILE": "vscode-icons:file-type-makefile",
		"OMakefile": "vscode-icons:file-type-makefile",
		"omakefile": "vscode-icons:file-type-makefile",
		"CMAKELISTS.TXT": "vscode-icons:file-type-makefile",
		"CMAKELISTS.txt": "vscode-icons:file-type-makefile",
		"CMakeLists.txt": "vscode-icons:file-type-makefile",
		"cmakelists.txt": "vscode-icons:file-type-makefile",
		"Procfile": "vscode-icons:file-type-procfile",
		"TODO": "vscode-icons:file-type-light-todo",
		"TODO.txt": "vscode-icons:file-type-light-todo",
		"TODO.md": "vscode-icons:file-type-light-todo"
	}
};
//#endregion
//#region src/node/fileIcons/findIcon.ts
function getFileIcon(fileName, type) {
	const name = getFileIconName(fileName, type);
	if (!name) return type !== "folder" ? defaultFile : defaultFolder;
	return name;
}
function getFileIconName(fileName, type = "file") {
	if (type === "folder") {
		const icon = definitions.folders[fileName];
		if (icon) return icon;
		if (fileName.includes("/")) return definitions.folders[fileName.slice(fileName.lastIndexOf("/") + 1)];
		return;
	}
	let icon = definitions.named[fileName] || definitions.files[fileName];
	if (icon) return icon;
	icon = getFileIconTypeFromExtension(fileName) || void 0;
	if (icon) return icon;
	for (const [partial, partialIcon] of Object.entries(definitions.partials)) if (fileName.includes(partial)) return partialIcon;
	return icon;
}
function getFileIconTypeFromExtension(fileName) {
	const firstDotIndex = fileName.indexOf(".");
	if (firstDotIndex === -1) return;
	let extension = fileName.slice(firstDotIndex);
	while (extension !== "") {
		const icon = definitions.extensions[extension];
		if (icon) return icon;
		const nextDotIndex = extension.indexOf(".", 1);
		if (nextDotIndex === -1) return;
		extension = extension.slice(nextDotIndex);
	}
}
//#endregion
//#region src/node/utils/cleanMarkdownEnv.ts
/**
* Whitelist of environment keys to preserve
*
* 要保留的环境键白名单
*/
const WHITE_LIST = [
	"base",
	"filePath",
	"filePathRelative",
	"references",
	"abbreviations",
	"annotations"
];
/**
* Clean markdown environment for inline rendering.
*
* 清理 markdown 环境以用于行内渲染。
*
* When using `md.renderInline()` in custom renderers, some environment properties
* may cause issues. This function creates a clean environment object with only
* the necessary properties preserved.
*
* 在自定义渲染器中使用 `md.renderInline()` 时，某些环境属性可能会导致问题。
* 该函数创建一个只保留必要属性的干净环境对象。
*
* @param env - Markdown environment / Markdown 环境
* @param excludes - Keys to exclude / 要排除的键
* @returns Cleaned environment / 清理后的环境
*
* @example
* ```ts
* const cleanEnv = cleanMarkdownEnv(env)
* const rendered = md.renderInline(content, cleanEnv)
* ```
*/
function cleanMarkdownEnv(env, excludes = []) {
	const result = {};
	for (const key of WHITE_LIST) {
		if (excludes.includes(key)) continue;
		result[key] = env[key];
	}
	return result;
}
//#endregion
//#region src/node/utils/stringifyProp.ts
/**
* Stringify a property value for use in Vue templates
*
* 将属性值字符串化为可在 Vue 模板中使用的格式
*
* @param data - Data to stringify / 要字符串化的数据
* @returns Stringified data with single quotes escaped / 字符串化后的数据，单引号已转义
*/
function stringifyProp(data) {
	return JSON.stringify(data).replace(/'/g, "&#39");
}
//#endregion
//#region src/node/container/codeTabs.ts
/**
* Create code tab icon getter function
*
* 创建代码标签页图标获取函数
*
* @param options - Code tabs options / 代码标签页选项
* @returns Icon getter function / 图标获取函数
*/
function createCodeTabIconGetter(options = {}) {
	const noop = () => void 0;
	if (options.icon === false) return noop;
	const { named, extensions } = isPlainObject(options.icon) ? options.icon : {};
	return function getIcon(filename) {
		if (named === false && definitions.named[filename]) return void 0;
		if (extensions === false && getFileIconTypeFromExtension(filename)) return;
		const hasNamed = named && named.length;
		const hasExt = extensions && extensions.length;
		if (hasNamed || hasExt) {
			if (hasNamed && named.includes(filename)) return definitions.named[filename];
			if (hasExt && extensions.some((ext) => filename.endsWith(ext))) return getFileIconTypeFromExtension(filename);
			return;
		}
		return getFileIconName(filename);
	};
}
/**
* Code tabs plugin - Enable code tabs container
*
* 代码标签页插件 - 启用代码标签页容器
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param options - Code tabs options / 代码标签页选项
*/
const codeTabs = (md, options = {}) => {
	const getIcon = createCodeTabIconGetter(options);
	tab(md, {
		name: "code-tabs",
		openRender: ({ active, data }, tokens, index, _, env) => {
			const { meta } = tokens[index];
			const titles = data.map(({ title }) => md.renderInline(title, cleanMarkdownEnv(env)));
			const tabsData = data.map((item, dataIndex) => {
				const { id = titles[dataIndex] } = item;
				return { id };
			});
			const titlesContent = titles.map((title, index) => {
				const icon = getIcon(title);
				return `<template #title${index}="{ value, isActive }">${icon ? `<VPIcon provider="iconify" name="${icon}"/>` : ""}<span>${title}</span></template>`;
			}).join("");
			return `<CodeTabs id="${index}" :data='${stringifyProp(tabsData)}'${active === -1 ? "" : ` :active="${active}"`}${meta.id ? ` tab-id="${meta.id}"` : ""}>${titlesContent}`;
		},
		closeRender: () => `</CodeTabs>`,
		tabOpenRender: ({ index }, tokens, tokenIndex) => {
			let foundFence = false;
			for (let i = tokenIndex; i < tokens.length; i++) {
				const { type } = tokens[i];
				if (type === "code-tabs_tab_close") break;
				if ((type === "fence" || type === "import_code") && !foundFence) {
					foundFence = true;
					continue;
				}
				tokens[i].type = "code_tab_empty";
				tokens[i].hidden = true;
			}
			return `<template #tab${index}="{ value, isActive }">`;
		},
		tabCloseRender: () => `</template>`
	});
};
//#endregion
//#region src/node/utils/resolveAttrs.ts
/**
* Regular expression for matching attribute values
*
* 匹配属性值的正则表达式
*/
const RE_ATTR_VALUE = /(?:^|\s+)(?<attr>[\w-]+)(?:=(?<quote>['"])(?<valueWithQuote>.+?)\k<quote>|=(?<valueWithoutQuote>\S+))?(?:\s+|$)/;
/**
* Resolve attribute string to object
*
* 将属性字符串解析为对象
*
* @param info - Attribute string / 属性字符串
* @returns Object with attrs and rawAttrs / 包含 attrs 和 rawAttrs 的对象
* @typeParam T - Attribute type / 属性类型
*/
function resolveAttrs(info) {
	info = info.trim();
	if (!info) return {
		rawAttrs: "",
		attrs: {}
	};
	const attrs = {};
	const rawAttrs = info;
	let matched;
	while (matched = info.match(RE_ATTR_VALUE)) {
		const { attr, valueWithQuote, valueWithoutQuote } = matched.groups;
		const value = valueWithQuote || valueWithoutQuote || true;
		let v = typeof value === "string" ? value.trim() : value;
		if (v === "true") v = true;
		else if (v === "false") v = false;
		else if (v === "\"\"" || v === "''") v = "";
		attrs[camelCase(attr)] = v;
		info = info.slice(matched[0].length);
	}
	return {
		attrs,
		rawAttrs
	};
}
/**
* Resolve single attribute value from info string
*
* 从信息字符串中解析单个属性值
*
* @param info - Info string / 信息字符串
* @param key - Attribute key / 属性键
* @returns Attribute value or undefined / 属性值或 undefined
*/
function resolveAttr(info, key) {
	const pattern = new RegExp(`(?:^|\\s+)${key}(?:=(?<quote>['"])(?<valueWithQuote>.+?)\\k<quote>|=(?<valueWithoutQuote>\\S+))?(?:\\s+|$)`);
	const groups = info.match(pattern)?.groups;
	return groups?.valueWithQuote || groups?.valueWithoutQuote;
}
//#endregion
//#region src/node/enhance/imageSize.ts
/**
* Regular expression for matching markdown image syntax
*
* 匹配 markdown 图片语法的正则表达式
*/
const REG_IMG = /!\[[^\]]*\]\([^)]*\)/g;
/**
* Regular expression for matching HTML img tag
*
* 匹配 HTML img 标签的正则表达式
*/
const REG_IMG_TAG = /<img([^>]*)>/g;
/**
* Regular expression for matching src/srcset attribute
*
* 匹配 src/srcset 属性的正则表达式
*/
const REG_IMG_TAG_SRC = /src(?:set)?=(['"])(.+?)\1/g;
/**
* List of badge URLs to exclude
*
* 要排除的徽章 URL 列表
*/
const BADGE_LIST = [
	"https://img.shields.io",
	"https://badge.fury.io",
	"https://badgen.net",
	"https://forthebadge.com",
	"https://vercel.com/button"
];
/**
* Image size plugin - Add width and height attributes to images
*
* 图片尺寸插件 - 为图片添加宽度和高度属性
*
* @param app - VuePress app / VuePress 应用
* @param md - Markdown instance / Markdown 实例
* @param type - Image size type: 'local', 'all', or false / 图片尺寸类型：'local'、'all' 或 false
*/
async function imageSizePlugin(app, md, type = false) {
	if (!app.env.isBuild || !type) return;
	const start = performance.now();
	const sizes = await getAllImageOriginalSize(await scanImage(app), type === "all");
	if (app.env.isDebug) logger.info(`[vuepress-plugin-md-power] imageSizePlugin: scan all images time spent: ${performance.now() - start}ms`);
	const imageRule = md.renderer.rules.image.bind(md);
	md.renderer.rules.image = (tokens, idx, options, env, self) => {
		const token = tokens[idx];
		const width = token.attrGet("width");
		const height = token.attrGet("height");
		const src = token.attrGet("src");
		const url = resolveImagePath(app, src, env.filePath);
		if (width && height || !url || !sizes[url] || src?.startsWith("data:")) return imageRule(tokens, idx, options, env, self);
		const size = resolveSize(sizes[url], width, height);
		token.attrSet("width", `${size.width}`);
		token.attrSet("height", `${size.height}`);
		return imageRule(tokens, idx, options, env, self);
	};
	md.renderer.rules.html_block = createHtmlRule(md.renderer.rules.html_block.bind(md));
	md.renderer.rules.html_inline = createHtmlRule(md.renderer.rules.html_inline.bind(md));
	/**
	* Create HTML rule for processing img tags
	*
	* 创建处理 img 标签的 HTML 规则
	*
	* @param rawHtmlRule - Original HTML rule / 原始 HTML 规则
	* @returns New HTML rule / 新的 HTML 规则
	*/
	function createHtmlRule(rawHtmlRule) {
		return (tokens, idx, options, env, self) => {
			const token = tokens[idx];
			token.content = token.content.replace(REG_IMG_TAG, (raw, info) => {
				const { attrs } = resolveAttrs(info);
				const src = attrs.src || attrs.srcset;
				const url = resolveImagePath(app, src, env.filepath);
				const { width, height } = attrs;
				if (width && height || !url || !sizes[url] || src?.startsWith("data:")) return raw;
				const size = resolveSize(sizes[url], width, height);
				attrs.width = size.width;
				attrs.height = size.height;
				return `<img ${Object.entries(attrs).map(([key, value]) => typeof value === "boolean" ? key : `${key}="${value}"`).join(" ")}>`;
			});
			return rawHtmlRule(tokens, idx, options, env, self);
		};
	}
	/**
	* Resolve image size from source
	*
	* 从源解析图片尺寸
	*
	* @param original - Image source / 图片源
	* @param width - Existing width / 现有宽度
	* @param height - Existing height / 现有高度
	* @returns Image size / 图片尺寸
	*/
	function resolveSize(original, width, height) {
		const { width: originalWidth, height: originalHeight } = original;
		const ratio = originalWidth / originalHeight;
		if (width && !height) {
			const w = Number.parseInt(width, 10);
			return {
				width: w,
				height: Math.round(w / ratio)
			};
		}
		if (height && !width) {
			const h = Number.parseInt(height, 10);
			return {
				width: Math.round(h * ratio),
				height: h
			};
		}
		return {
			width: originalWidth,
			height: originalHeight
		};
	}
}
/**
* Scan all images in the source directory
*
* 扫描源目录中的所有图片
*
* @param app - VuePress app / VuePress 应用
* @returns List of image URLs / 图片 URL 列表
*/
async function scanImage(app) {
	if (!app.env.isBuild) return [];
	const cwd = app.dir.source();
	const files = await fs.readdir(cwd, { recursive: true });
	const result = /* @__PURE__ */ new Set();
	await pMap(files, async (file) => {
		const filepath = path.join(cwd, file);
		if ((await fs.stat(filepath)).isFile() && filepath.endsWith(".md") && !filepath.includes(".vuepress") && !filepath.includes("node_modules")) {
			const content = await fs.readFile(filepath, "utf-8");
			const syntaxMatched = content.match(REG_IMG) ?? [];
			for (const img of syntaxMatched) {
				const url = resolveImagePath(app, img.slice(img.indexOf("](") + 2, -1).split(/\s+/)[0], filepath);
				url && result.add(url);
			}
			const tagMatched = content.match(REG_IMG_TAG) ?? [];
			for (const img of tagMatched) {
				const url = resolveImagePath(app, img.match(REG_IMG_TAG_SRC)?.[2] ?? "", filepath);
				url && result.add(url);
			}
		}
	}, { concurrency: 64 });
	return Array.from(result);
}
/**
* Get original size of all images
*
* 获取所有图片的原始尺寸
*
* @param images - List of image URLs / 图片 URL 列表
* @param includeRemote - Whether to include remote images / 是否包含远程图片
* @returns Record of image URLs and their sizes / 图片 URL 及其尺寸的记录
*/
async function getAllImageOriginalSize(images, includeRemote = false) {
	const result = {};
	for (const src of images) {
		const size = await getImageOriginalSize(src, includeRemote);
		if (size) result[src] = size;
	}
	return result;
}
async function getImageOriginalSize(image, includeRemote = false) {
	if (!image) return null;
	const isRemote = isLinkHttp(image);
	if (isRemote && includeRemote && !BADGE_LIST.some((badge) => image.startsWith(badge))) {
		const { width, height } = await fetchRemoteImageSize(image.startsWith("//") ? `https:${image}` : image);
		if (width && height) return {
			width,
			height
		};
	}
	if (!isRemote) {
		const [, data] = attempt(() => imageSize(fs.readFileSync(image)));
		if (data?.width && data?.height) return {
			width: data.width,
			height: data.height
		};
	}
	return null;
}
/**
* Resolve image path from source
*
* 从源解析图片路径
*
* @param app - VuePress app / VuePress 应用
* @param src - Image source / 图片源
* @param currentPath - Current path / 当前路径
* @returns Image path / 图片路径
*/
function resolveImagePath(app, src, currentPath) {
	if (!src) return "";
	if (isLinkHttp(src)) return src;
	if (src[0] === "/") return app.dir.public(src.slice(1));
	return currentPath ? path.resolve(currentPath, src) : "";
}
/**
* Fetch image size from remote URL
*
* 从远程 URL 获取图片尺寸
*
* @param src - Image URL / 图片 URL
* @returns Image size / 图片尺寸
*/
async function fetchRemoteImageSize(src) {
	const link = new URL(src);
	const promise = new Promise((resolve) => {
		http.get(link, async (stream) => {
			const chunks = [];
			for await (const chunk of stream) {
				chunks.push(chunk);
				const [, data] = attempt(imageSize, Buffer.concat(chunks));
				if (data && data.width && data.height) return resolve(data);
			}
			resolve({
				width: 0,
				height: 0
			});
		}).on("error", () => resolve({
			width: 0,
			height: 0
		}));
	});
	try {
		return await withTimeout(() => promise, 3e3);
	} catch {}
	return {
		width: 0,
		height: 0
	};
}
//#endregion
//#region src/node/demo/supports/file.ts
const require = createRequire(process.cwd());
function findFile(app, env, url) {
	if (url.startsWith("/")) return app.dir.source(url.slice(1));
	if (url.startsWith("./") || url.startsWith("../")) return app.dir.source(path$1.dirname(env.filePathRelative), url);
	if (url.startsWith("@source/")) return app.dir.source(url.slice(8));
	try {
		return require.resolve(url);
	} catch {
		return url;
	}
}
function readFileSync(filepath) {
	try {
		return fs$1.readFileSync(filepath, "utf-8");
	} catch {
		return false;
	}
}
function writeFileSync(filepath, content) {
	const dirname = path$1.dirname(filepath);
	fs$1.mkdirSync(dirname, { recursive: true });
	fs$1.writeFileSync(filepath, content, "utf-8");
}
//#endregion
//#region src/node/embed/createEmbedRuleBlock.ts
/**
* Create embed rule block
*
* 创建嵌入规则块
*
* Syntax: \@\[name]()
*
* 语法：\@\[name]()
*
* @param md - Markdown instance / Markdown 实例
* @param {EmbedRuleBlockOptions} options - Embed rule block options / 嵌入规则块选项
* @typeParam Meta - Metadata type / 元数据类型
*/
function createEmbedRuleBlock(md, { type, name = type, syntaxPattern, beforeName = "import_code", ruleOptions = { alt: [
	"paragraph",
	"reference",
	"blockquote",
	"list"
] }, meta, content }) {
	const MIN_LENGTH = type.length + 5;
	const START_CODES = [
		64,
		91,
		...type.split("").map((c) => c.charCodeAt(0))
	];
	md.block.ruler.before(beforeName, name, (state, startLine, endLine, silent) => {
		const pos = state.bMarks[startLine] + state.tShift[startLine];
		const max = state.eMarks[startLine];
		if (pos + MIN_LENGTH > max) return false;
		for (let i = 0; i < START_CODES.length; i += 1) if (state.src.charCodeAt(pos + i) !== START_CODES[i]) return false;
		const content = state.src.slice(pos, max);
		const match = content.match(syntaxPattern);
		if (!match) return false;
		/* istanbul ignore if -- @preserve */
		if (silent) return true;
		const token = state.push(name, "", 0);
		token.meta = meta(match);
		token.content = content;
		token.map = [startLine, startLine + 1];
		state.line = startLine + 1;
		return true;
	}, ruleOptions);
	md.renderer.rules[name] = (tokens, index, _, env) => {
		const token = tokens[index];
		token.content = content(token.meta, token.content, env);
		return token.content;
	};
}
//#endregion
//#region src/node/utils/parseRect.ts
/**
* Parse rect size string, add unit if it's a number
*
* 解析矩形尺寸字符串，如果是数字则添加单位
*
* @param str - Size string / 尺寸字符串
* @param unit - Unit to append (default: 'px') / 要添加的单位（默认：'px'）
* @returns Size string with unit / 带单位的尺寸字符串
*/
function parseRect(str, unit = "px") {
	if (Number.parseFloat(str) === Number(str)) return `${str}${unit}`;
	return str;
}
//#endregion
//#region src/node/utils/stringifyAttrs.ts
/**
* Stringify attributes object to HTML attribute string
*
* 将属性对象字符串化为 HTML 属性字符串
*
* @param attrs - Attributes object / 属性对象
* @param withUndefinedOrNull - Whether to include undefined/null values / 是否包含 undefined/null 值
* @param forceStringify - Keys to force stringify / 强制字符串化的键
* @returns HTML attribute string / HTML 属性字符串
* @typeParam T - Attribute type / 属性类型
*/
function stringifyAttrs(attrs, withUndefinedOrNull = false, forceStringify = []) {
	const result = Object.entries(attrs).map(([key, value]) => {
		const k = kebabCase(key);
		if (isUndefined(value) || value === "undefined") return withUndefinedOrNull ? `:${k}="undefined"` : "";
		if (isNull(value) || value === "null") return withUndefinedOrNull ? `:${k}="null"` : "";
		if (value === "true") value = true;
		if (value === "false") value = false;
		if (isBoolean(value)) return value ? `${k}` : "";
		if (isNumber(value)) return `:${k}="${value}"`;
		if (isString(value) && (value[0] === "{" || value[0] === "[")) {
			const v = value.replaceAll("\"", "'");
			if (forceStringify.includes(key)) return `${k}="${v}"`;
			return `:${k}="${v}"`;
		}
		return `${key[0] === ":" ? ":" : ""}${k}="${String(value)}"`;
	}).filter(Boolean).join(" ");
	return result ? ` ${result}` : "";
}
//#endregion
//#region src/node/container/createContainer.ts
/**
* Create markdown-it custom container plugin
*
* 创建 markdown-it 的自定义容器插件
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param type - Container type (e.g., 'tip', 'warning') / 容器类型（如 'tip', 'warning' 等）
* @param options - Optional before/after render hooks / 可选的 before/after 渲染钩子
* @param options.before - Callback for rendering container opening tag / 渲染容器起始标签时的回调函数
* @param options.after - Callback for rendering container closing tag / 渲染容器结束标签时的回调函数
*/
function createContainerPlugin(md, type, { before, after } = {}) {
	const render = (tokens, index, options, env) => {
		const token = tokens[index];
		const info = token.info.trim().slice(type.length).trim() || "";
		if (token.nesting === 1) return before?.(info, tokens, index, options, env) ?? `<div class="custom-container ${type}">`;
		else return after?.(info, tokens, index, options, env) ?? "</div>";
	};
	md.use(container, type, { render });
}
/**
* Create a custom container rule where content is not processed by markdown-it
* Requires custom content processing logic
* ```md
* ::: type
*  xxxx    <-- content: this part will not be processed by markdown-it
* :::
* ```
*
* @example
* ```ts
* const example = createContainerSyntaxPlugin(md, 'example', (tokens, index, options, env) => {
*   const { content, meta } = tokens[index]
*   return `<div class="example">${meta.title} | ${content}</div>`
* })
* ```
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param type - Container type / 容器类型
* @param render - Custom render rule / 自定义渲染规则
*/
function createContainerSyntaxPlugin(md, type, render) {
	const maker = ":";
	const markerMinLen = 3;
	/**
	* Custom container block rule definition
	*
	* 自定义容器的 block 规则定义
	*
	* @param state - Current block state / 当前 block 状态
	* @param startLine - Start line / 起始行
	* @param endLine - End line / 结束行
	* @param silent - Silent mode / 是否为静默模式
	* @returns Whether matched / 是否匹配到自定义容器
	*/
	function defineContainer(state, startLine, endLine, silent) {
		const start = state.bMarks[startLine] + state.tShift[startLine];
		const max = state.eMarks[startLine];
		let pos = start;
		if (state.src[pos] !== maker) return false;
		for (pos = start + 1; pos <= max; pos++) if (state.src[pos] !== maker) break;
		if (pos - start < markerMinLen) return false;
		const markup = state.src.slice(start, pos);
		const info = state.src.slice(pos, max).trim();
		if (!info.startsWith(type)) return false;
		/* istanbul ignore if -- @preserve */
		if (silent) return true;
		let line = startLine;
		let content = "";
		while (++line < endLine) {
			if (state.src.slice(state.bMarks[line], state.eMarks[line]).trim() === markup) break;
			content += `${state.src.slice(state.bMarks[line], state.eMarks[line])}\n`;
		}
		const token = state.push(`${type}_container`, "", 0);
		token.meta = resolveAttrs(info.slice(type.length)).attrs;
		token.content = content;
		token.markup = `${markup} ${type}`;
		token.map = [startLine, line + 1];
		state.line = line + 1;
		return true;
	}
	const defaultRender = (tokens, index) => {
		const { content } = tokens[index];
		return `<div class="custom-container ${type}">${content}</div>`;
	};
	md.block.ruler.before("fence", `${type}_definition`, defineContainer);
	md.renderer.rules[`${type}_container`] = render ?? defaultRender;
}
//#endregion
//#region src/node/container/codeTree.ts
/**
* Unsupported file types for code tree
*
* 代码树不支持的文件类型
*/
const UNSUPPORTED_FILE_TYPES = [
	"jpg",
	"jpeg",
	"png",
	"gif",
	"avif",
	"webp",
	"mp3",
	"mp4",
	"ogg",
	"m3u8",
	"m3u",
	"flv",
	"webm",
	"wav",
	"flac",
	"aac",
	"pdf",
	"doc",
	"docx",
	"ppt",
	"pptx",
	"xls",
	"xlsx"
];
/**
* Parse file paths to file tree node structure
*
* 将文件路径数组解析为文件树节点结构
*
* @param files - File path array / 文件路径数组
* @returns File tree node array / 文件树节点数组
*/
function parseFileNodes(files) {
	const nodes = [];
	for (const file of files) {
		const parts = removeLeadingSlash(file).split("/");
		let node = nodes;
		for (let i = 0; i < parts.length; i++) {
			const part = parts[i];
			const isFile = i === parts.length - 1;
			let child = node.find((n) => n.filename === part);
			if (!child) {
				child = {
					level: i + 1,
					filename: part,
					filepath: isFile ? file : void 0,
					children: isFile ? void 0 : []
				};
				node.push(child);
			}
			if (!isFile && child.children) node = child.children;
		}
	}
	return nodes;
}
/**
* Code tree plugin - Register code-tree container and embed syntax
*
* 注册 code-tree 容器和嵌入语法的 markdown 插件
*
* @param md - Markdown-it instance / markdown-it 实例
* @param app - VuePress app instance / vuepress app 实例
* @param options - Code tree options / code-tree 配置项
*/
function codeTreePlugin(md, app, options = {}) {
	/**
	* Get file or folder icon
	*
	* 获取文件或文件夹的图标
	*/
	const getIcon = (filename, type, mode) => {
		mode ||= options.icon || "colored";
		if (mode === "simple") return type === "folder" ? defaultFolder : defaultFile;
		return getFileIcon(filename, type);
	};
	/**
	* Render file tree nodes to component string
	*
	* 渲染文件树节点为组件字符串
	*/
	function renderFileTree(nodes, mode) {
		return nodes.map((node) => {
			const props = {
				filename: node.filename,
				level: node.level,
				type: node.children?.length ? "folder" : "file",
				expanded: true,
				filepath: node.filepath
			};
			return `<FileTreeNode${stringifyAttrs(props, false, ["filename", "filepath"])}>
  <template #icon><VPIcon provider="iconify" name="${getIcon(node.filename, props.type, mode)}" /></template>
  ${node.children?.length ? renderFileTree(node.children, mode) : ""}
</FileTreeNode>`;
		}).join("\n");
	}
	createContainerPlugin(md, "code-tree", {
		before: (info, tokens, index) => {
			const files = [];
			let activeFile;
			for (let i = index + 1; !(tokens[i].nesting === -1 && tokens[i].type === "container_code-tree_close"); i++) {
				const token = tokens[i];
				if (token.type === "fence" && token.tag === "code") {
					const fenceInfo = md.utils.unescapeAll(token.info);
					const title = resolveAttr(fenceInfo, "title");
					if (title) {
						files.push(title);
						if (fenceInfo.includes(":active")) activeFile = title;
					}
				}
			}
			const { attrs } = resolveAttrs(info);
			const { title, icon, height, entry } = attrs;
			const fileTreeNodes = parseFileNodes(files);
			const entryFile = activeFile || entry || files[0];
			const h = height || String(options.height);
			return `<VPCodeTree${stringifyAttrs({
				title,
				entryFile,
				height: h ? parseRect(h) : void 0
			})}><template #file-tree>${renderFileTree(fileTreeNodes, icon)}</template>`;
		},
		after: () => "</VPCodeTree>"
	});
	createEmbedRuleBlock(md, {
		type: "code-tree",
		syntaxPattern: /^@\[code-tree([^\]]*)\]\(([^)]*)\)/,
		meta: ([, info, dir]) => {
			const { attrs } = resolveAttrs(info);
			const h = attrs.height || String(options.height);
			return {
				title: attrs.title,
				entryFile: attrs.entry,
				icon: attrs.icon,
				height: h ? parseRect(h) : void 0,
				dir
			};
		},
		content: ({ dir, icon, ...props }, _, env) => {
			const codeTreeFiles = env.codeTreeFiles ??= [];
			const root = findFile(app, env, dir);
			const files = tinyglobby.globSync("**/*", {
				cwd: root,
				onlyFiles: true,
				dot: true,
				ignore: [
					"**/node_modules/**",
					"**/.DS_Store",
					"**/.gitkeep"
				]
			}).sort((a, b) => {
				const al = a.split("/").length;
				return b.split("/").length - al;
			});
			props.entryFile ||= files[0];
			const codeContent = files.map((file) => {
				const ext = path$1.extname(file).slice(1);
				if (UNSUPPORTED_FILE_TYPES.includes(ext)) return "";
				const filepath = path$1.join(root, file);
				codeTreeFiles.push(filepath);
				const content = readFileSync(filepath);
				return `\`\`\`${ext || "txt"} title="${file}"\n${content}\n\`\`\``;
			}).filter(Boolean).join("\n");
			const fileTreeNodes = parseFileNodes(files);
			return `<VPCodeTree${stringifyAttrs(props)}><template #file-tree>${renderFileTree(fileTreeNodes, icon)}</template>${md.render(codeContent, cleanMarkdownEnv(env))}</VPCodeTree>`;
		}
	});
}
/**
* Extend page dependencies with codeTreeFiles
*
* 扩展页面依赖，将 codeTreeFiles 添加到页面依赖中
*
* @param page - VuePress page object / vuepress 页面对象
*/
function extendsPageWithCodeTree(page) {
	const codeTreeFiles = page.markdownEnv.codeTreeFiles ?? [];
	if (codeTreeFiles.length) page.deps.push(...codeTreeFiles);
}
//#endregion
//#region src/node/utils/findLocales.ts
/**
* Find locale configurations for a specific key
*
* 查找特定键的本地化配置
*
* @param locales - Locale configuration / 本地化配置
* @param key - Key to find / 要查找的键
* @returns Record of locale paths to locale data / 本地化路径到本地化数据的记录
*/
function findLocales(locales, key) {
	const res = {};
	for (const [locale, value] of Object.entries(locales)) res[locale] = value[key] ?? {};
	return res;
}
//#endregion
//#region src/node/container/align.ts
/**
* Align plugin - Enable text alignment containers
*
* 对齐插件 - 启用文本对齐容器
*
* Syntax:
* - ::: left
* - ::: center
* - ::: right
* - ::: justify
* - ::: flex [options]
*
* @param md - Markdown instance / Markdown 实例
*/
function alignPlugin(md) {
	for (const name of [
		"left",
		"center",
		"right",
		"justify"
	]) createContainerPlugin(md, name, { before: () => `<div style="text-align:${name}">` });
	createContainerPlugin(md, "flex", { before: (info) => {
		const { attrs } = resolveAttrs(info);
		const styles = ["margin:16px 0;display:flex"];
		const align = attrs.start ? "flex-start" : attrs.end ? "flex-end" : attrs.center ? "center" : "";
		const justify = attrs.between ? "space-between" : attrs.around ? "space-around" : attrs.center ? "center" : "";
		if (align) styles.push(`align-items:${align}`);
		if (justify) styles.push(`justify-content:${justify}`);
		if (attrs.column) styles.push("flex-direction:column");
		if (attrs.wrap) styles.push("flex-wrap:wrap");
		styles.push(`gap:${parseRect(attrs.gap || "16")}`);
		return `<div style="${styles.join(";")}">`;
	} });
}
//#endregion
//#region src/node/container/card.ts
/**
* Card plugin - Enable card containers
*
* 卡片插件 - 启用卡片容器
*
* Syntax:
* - ::: card title="xxx" icon="xxx"
* - ::: card-grid
* - ::: card-masonry cols="2" gap="10"
*
* @param md - Markdown instance / Markdown 实例
*/
function cardPlugin(md) {
	/**
	* ::: card title="xxx" icon="xxx"
	* xxx
	* :::
	*/
	createContainerPlugin(md, "card", {
		before(info) {
			const { attrs } = resolveAttrs(info);
			return `<VPCard${stringifyAttrs(attrs)}>`;
		},
		after: () => "</VPCard>"
	});
	/**
	* :::: card-grid
	* ::: card
	* xxx
	* :::
	* ::: card
	* xxx
	* :::
	* ::::
	*/
	createContainerPlugin(md, "card-grid", {
		before: () => "<VPCardGrid>",
		after: () => "</VPCardGrid>"
	});
	/**
	* ::: card-masonry cols="2" gap="10"
	* ::: card
	* xxx
	* :::
	* ::: card
	* xxx
	* :::
	* ::::
	*/
	createContainerPlugin(md, "card-masonry", {
		before: (info) => {
			const { attrs } = resolveAttrs(info);
			if (attrs.cols) attrs.cols = attrs.cols[0] === "{" ? attrs.cols : Number.parseInt(`${attrs.cols}`);
			if (attrs.gap) attrs.gap = Number(attrs.gap);
			return `<VPCardMasonry${stringifyAttrs(attrs)}>`;
		},
		after: () => "</VPCardMasonry>"
	});
}
//#endregion
//#region src/node/container/chat.ts
/**
* Render chat messages to HTML
*
* 渲染聊天消息为 HTML
*
* @param md - Markdown instance / Markdown 实例
* @param env - Markdown environment / Markdown 环境
* @param messages - Chat messages / 聊天消息数组
* @returns Rendered HTML / 渲染后的 HTML
*/
function chatMessagesRender(md, env, messages) {
	let currentDate = "";
	return messages.map(({ sender, username, date, content }) => {
		let messageContent = "";
		if (!currentDate || currentDate !== date) {
			currentDate = date;
			messageContent += `<div class="vp-chat-date"><span>${currentDate}</span></div>\n`;
		}
		messageContent += `<div class="vp-chat-message ${sender}">
      <div class="vp-chat-message-body">\
      ${sender === "user" ? `\n<p class="vp-chat-username">${username}</p>` : ""}
        <div class="message-content">
          ${md.render(content.join("\n"), cleanMarkdownEnv(env)).trim()}
        </div>
      </div>
    </div>`;
		return messageContent;
	}).join("\n");
}
/**
* Parse chat content to messages
*
* 解析聊天内容为消息数组
*
* @param content - Raw chat content / 原始聊天内容
* @returns Parsed chat messages / 解析后的聊天消息
*/
function parseChatContent(content) {
	const lines = content.split("\n");
	const messages = [];
	let currentDate = "";
	let message;
	for (const line of lines) {
		const lineStr = line.trim();
		if (lineStr.startsWith("{:") && lineStr.endsWith("}")) {
			currentDate = lineStr.slice(2, -1).trim();
			continue;
		}
		if (lineStr.startsWith("{") && lineStr.endsWith("}")) {
			const username = lineStr.slice(1, -1).trim();
			message = {
				sender: username === "." ? "self" : "user",
				username,
				date: currentDate,
				content: []
			};
			messages.push(message);
			continue;
		}
		if (message?.sender) message.content.push(line);
	}
	return messages;
}
/**
* Chat plugin - Enable chat container
*
* 聊天插件 - 启用聊天容器
*
* @param md - Markdown-it instance / Markdown-it 实例
*/
const chatPlugin = (md) => createContainerSyntaxPlugin(md, "chat", (tokens, idx, _, env) => `<div class="vp-chat">
  <div class="vp-chat-header">
    <p class="vp-chat-title">${tokens[idx].meta?.title || "Chat"}</p>
  </div>
  <div class="vp-chat-content">
    ${chatMessagesRender(md, env, parseChatContent(tokens[idx].content))}
  </div>
</div>`);
//#endregion
//#region src/node/container/collapse.ts
/**
* Collapse plugin - Enable collapse container
*
* 折叠面板插件 - 启用折叠面板容器
*
* @param md - Markdown instance / Markdown 实例
*/
function collapsePlugin(md) {
	createContainerPlugin(md, "collapse", {
		before: (info, tokens, index) => {
			const { attrs } = resolveAttrs(info);
			const idx = parseCollapse(tokens, index, attrs);
			const { accordion } = attrs;
			return `<VPCollapse${stringifyAttrs({
				accordion,
				index: idx
			})}>`;
		},
		after: () => `</VPCollapse>`
	});
	md.renderer.rules.collapse_item_open = (tokens, idx) => {
		const { expand, index } = tokens[idx].meta;
		return `<VPCollapseItem${stringifyAttrs({
			expand,
			index
		})}>`;
	};
	md.renderer.rules.collapse_item_close = () => "</VPCollapseItem>";
	md.renderer.rules.collapse_item_title_open = () => "<template #title>";
	md.renderer.rules.collapse_item_title_close = () => "</template>";
}
/**
* Parse collapse tokens
*
* 解析折叠面板令牌
*
* @param tokens - Token array / 令牌数组
* @param index - Start index / 起始索引
* @param attrs - Collapse metadata / 折叠面板元数据
* @returns Default expanded index / 默认展开索引
*/
function parseCollapse(tokens, index, attrs) {
	const listStack = [];
	let idx = -1;
	let defaultIndex;
	let hashExpand = false;
	for (let i = index + 1; i < tokens.length; i++) {
		const token = tokens[i];
		if (token.type === "container_collapse_close") break;
		if (token.type === "bullet_list_open" || token.type === "ordered_list_open") {
			listStack.push(0);
			if (listStack.length === 1) token.hidden = true;
		} else if (token.type === "bullet_list_close" || token.type === "ordered_list_close") {
			listStack.pop();
			if (listStack.length === 0) token.hidden = true;
		} else if (token.type === "list_item_open") {
			if (listStack.length === 1) {
				token.type = "collapse_item_open";
				tokens[i + 1].type = "collapse_item_title_open";
				tokens[i + 3].type = "collapse_item_title_close";
				idx++;
				const firstToken = tokens[i + 2].children?.[0];
				let flag = "";
				let expand;
				if (firstToken?.type === "text") firstToken.content = firstToken.content.trim().replace(/^:[+\-]\s*/, (match) => {
					flag = match.trim();
					return "";
				});
				if (attrs.accordion) {
					if (!hashExpand && flag === ":+") {
						expand = hashExpand = true;
						defaultIndex = idx;
					}
				} else if (flag === ":+") expand = true;
				else if (flag === ":-") expand = false;
				else expand = !!attrs.expand;
				token.meta = {
					index: idx,
					expand
				};
			}
		} else if (token.type === "list_item_close") {
			if (listStack.length === 1) token.type = "collapse_item_close";
		}
	}
	if (attrs.accordion && attrs.expand && !hashExpand) defaultIndex = 0;
	return defaultIndex;
}
//#endregion
//#region src/node/utils/encryptContent.ts
/**
* Get key material from password
*
* 从密码获取密钥材料
*
* @see https://developer.mozilla.org/zh-CN/docs/Web/API/SubtleCrypto/deriveKey#pbkdf2_2
* @param password - Password string / 密码字符串
* @returns CryptoKey / 加密密钥
*/
function getKeyMaterial(password) {
	const enc = new TextEncoder();
	return webcrypto.subtle.importKey("raw", enc.encode(password), "PBKDF2", false, ["deriveBits", "deriveKey"]);
}
/**
* Get derived key from key material
*
* 从密钥材料获取派生密钥
*
* @param keyMaterial - Key material / 密钥材料
* @param salt - Salt for key derivation / 密钥派生盐值
* @returns Derived CryptoKey / 派生加密密钥
*/
function getCryptoDeriveKey(keyMaterial, salt) {
	return webcrypto.subtle.deriveKey({
		name: "PBKDF2",
		salt,
		iterations: 1e5,
		hash: "SHA-256"
	}, keyMaterial, {
		name: "AES-CBC",
		length: 256
	}, true, ["encrypt", "decrypt"]);
}
/**
* Encrypt content using AES-CBC
*
* 使用 AES-CBC 加密内容
*
* @see https://github.com/mdn/dom-examples/blob/main/web-crypto/encrypt-decrypt/aes-cbc.js
* @param content - Content to encrypt / 要加密的内容
* @param options - Encryption options / 加密选项
* @param options.password - Password for encryption / 加密密码
* @param options.iv - Initialization vector / 初始化向量
* @param options.salt - Salt for key derivation / 密钥派生盐值
* @returns Encrypted content / 加密后的内容
*/
async function encryptContent(content, options) {
	const { password, iv, salt } = options;
	const key = await getCryptoDeriveKey(await getKeyMaterial(password), salt);
	const enc = new TextEncoder();
	const cipherTextData = await webcrypto.subtle.encrypt({
		name: "AES-CBC",
		iv
	}, key, enc.encode(content));
	return String.fromCharCode(...new Uint8Array(cipherTextData));
}
//#endregion
//#region src/node/utils/logger.ts
/* istanbul ignore file -- @preserve */
/**
* Logger utility class for plugin
*
* 插件日志工具类
*/
var Logger = class {
	/**
	* Create a logger instance
	*
	* 创建日志记录器实例
	*
	* @param name - Plugin/Theme name / 插件/主题名称
	*/
	constructor(name = "") {
		this.name = name;
	}
	/**
	* Initialize spinner
	*
	* 初始化加载动画
	*
	* @param subname - Subname / 子名称
	* @param text - Loading text / 加载文本
	* @returns Ora spinner instance / Ora 加载动画实例
	*/
	init(subname, text) {
		return ora({
			prefixText: colors.blue(`${this.name}${subname ? `:${subname}` : ""}: `),
			text
		});
	}
	/**
	* Create a loading spinner with text
	*
	* 创建带文本的加载动画
	*
	* @param subname - Subname / 子名称
	* @param msg - Message / 消息
	* @returns Object with succeed and fail methods / 包含 succeed 和 fail 方法的对象
	*/
	load(subname, msg) {
		const instance = this.init(subname, msg);
		return {
			succeed: (text) => instance.succeed(text),
			fail: (text) => instance.succeed(text)
		};
	}
	/**
	* Log info message
	*
	* 记录信息消息
	*
	* @param subname - Subname / 子名称
	* @param text - Message text / 消息文本
	* @param args - Additional arguments / 额外参数
	*/
	info(subname, text = "", ...args) {
		this.init(subname, colors.blue(text)).info();
		if (args.length) console.info(...args);
	}
	/**
	* Log success message
	*
	* 记录成功消息
	*
	* @param subname - Subname / 子名称
	* @param text - Message text / 消息文本
	* @param args - Additional arguments / 额外参数
	*/
	succeed(subname, text = "", ...args) {
		this.init(subname, colors.green(text)).succeed();
		if (args.length) console.log(...args);
	}
	/**
	* Log warning message
	*
	* 记录警告消息
	*
	* @param subname - Subname / 子名称
	* @param text - Message text / 消息文本
	* @param args - Additional arguments / 额外参数
	*/
	warn(subname, text = "", ...args) {
		this.init(subname, colors.yellow(text)).warn();
		if (args.length) console.warn(...args);
	}
	/**
	* Log error message
	*
	* 记录错误消息
	*
	* @param subname - Subname / 子名称
	* @param text - Message text / 消息文本
	* @param args - Additional arguments / 额外参数
	*/
	error(subname, text = "", ...args) {
		this.init(subname, colors.red(text)).fail();
		if (args.length) console.error(...args);
	}
};
/**
* Default logger instance for vuepress-plugin-md-power
*
* vuepress-plugin-md-power 的默认日志记录器实例
*/
const logger$1 = new Logger("vuepress-plugin-md-power");
//#endregion
//#region src/node/container/encrypt.ts
/**
* Encrypt plugin - Enable encrypted content container
*
* 加密插件 - 启用加密内容容器
*
* @param app - VuePress app / VuePress 应用
* @param md - Markdown instance / Markdown 实例
* @param options - Encrypt snippet options / 加密片段选项
*/
function encryptPlugin(app, md, options) {
	const encrypted = /* @__PURE__ */ new Set();
	const entryFile = "internal/encrypt-snippets/index.js";
	/**
	* Write encrypted content to temp file
	*
	* 将加密内容写入临时文件
	*/
	const writeTemp = async (hash, content, options) => {
		const encrypted = await encryptContent(content, options);
		await app.writeTemp(`internal/encrypt-snippets/${hash}.js`, `export default ${JSON.stringify(encrypted)}`);
	};
	/**
	* Write entry file with all encrypted snippets
	*
	* 写入包含所有加密片段的入口文件
	*/
	const writeEntry = debounce(150, async () => {
		let content = `export default {\n`;
		for (const hash of encrypted) content += `  '${hash}': () => import('./${hash}.js' /* webpackChunkName: "snippet-${hash}" */),\n`;
		content += "\n}\n";
		app.writeTemp(entryFile, content);
	});
	if (!fs.existsSync(app.dir.temp(entryFile))) app.writeTemp(entryFile, "export default {}\n");
	const localKeys = Object.keys(app.options.locales || {}).filter((key) => key !== "/");
	/**
	* Get locale from relative path
	*
	* 从相对路径获取本地化
	*/
	const getLocale = (relativePath) => {
		const relative = ensureLeadingSlash(relativePath);
		return localKeys.find((key) => relative.startsWith(key)) || "/";
	};
	createContainerSyntaxPlugin(md, "encrypt", (tokens, index, _, env) => {
		const { meta, content } = tokens[index];
		const { password, pwd, hint } = meta;
		const rendered = md.render(content, cleanMarkdownEnv(env));
		const _pwd = password || pwd || options.password;
		if (!_pwd) {
			logger$1.warn(`${colors.cyan("[encrypt snippet]")} ${colors.green("::: encrypt")} container missing password. ${colors.gray(`(${env.filePathRelative})`)}`);
			return rendered;
		}
		const contentHash = hash(content);
		encrypted.add(contentHash);
		const salt = getRandomValues(new Uint8Array(16));
		const iv = getRandomValues(new Uint8Array(16));
		writeEntry();
		writeTemp(contentHash, rendered, {
			salt,
			iv,
			password: String(_pwd)
		});
		return `<VPEncryptSnippet data="${encodeData(JSON.stringify({
			hash: contentHash,
			salt: Array.from(salt),
			iv: Array.from(iv)
		}))}" hint="${hint || ""}" path-locale="${getLocale(env.filePathRelative)}" />`;
	});
}
//#endregion
//#region src/node/container/field.ts
/**
* Field plugin - Enable field container for API documentation
*
* 字段插件 - 启用用于 API 文档的字段容器
*
* Syntax: ::: field name="xxx" type="string" required
* 语法：::: field name="xxx" type="string" required
*
* @param md - Markdown instance / Markdown 实例
*/
function fieldPlugin(md) {
	createContainerPlugin(md, "field", {
		before: (info) => {
			const { attrs } = resolveAttrs(info);
			const { name, type, required, optional, deprecated, default: defaultValue } = attrs;
			return `<VPField${stringifyAttrs({
				name,
				required,
				optional,
				deprecated
			})}${!isUndefined(type) ? ` type="${type}"` : ""}${!isUndefined(defaultValue) ? ` default-value="${defaultValue}"` : ""}>`;
		},
		after: () => "</VPField>"
	});
	createContainerPlugin(md, "field-group", { before: () => "<div class=\"vp-field-group\">" });
}
//#endregion
//#region src/node/container/fileTree.ts
/**
* Parse raw file tree content to node tree structure
*
* 解析原始文件树内容为节点树结构
*
* @param content - Raw file tree text content / 文件树的原始文本内容
* @returns File tree node array / 文件树节点数组
*/
function parseFileTreeRawContent(content) {
	const root = {
		level: -1,
		children: []
	};
	const stack = [root];
	const lines = content.trimEnd().split("\n");
	const spaceLength = lines[0].match(/^\s*/)?.[0].length ?? 0;
	for (const line of lines) {
		const match = line.match(/^(\s*)-(.*)$/);
		if (!match) continue;
		const level = Math.floor((match[1].length - spaceLength) / 2);
		const info = match[2].trim();
		while (stack.length > 0 && stack[stack.length - 1].level >= level) stack.pop();
		const parent = stack[stack.length - 1];
		const node = {
			level,
			children: [],
			...parseFileTreeNodeInfo(info)
		};
		parent.children.push(node);
		stack.push(node);
	}
	return root.children;
}
/**
* Regex for focus marker
*
* 高亮标记正则
*/
const RE_FOCUS = /^\*\*(.*)\*\*(?:$|\s+)/;
/**
* Parse single node info string, extract filename, comment, type, etc.
*
* 解析单个节点的 info 字符串，提取文件名、注释、类型等属性
*
* @param info - Node description string / 节点描述字符串
* @returns File tree node props / 文件树节点属性
*/
function parseFileTreeNodeInfo(info) {
	let filename = "";
	let comment = "";
	let focus = false;
	let expanded = true;
	let type = "file";
	let diff;
	if (info.startsWith("++")) {
		info = info.slice(2).trim();
		diff = "add";
	} else if (info.startsWith("--")) {
		info = info.slice(2).trim();
		diff = "remove";
	}
	info = info.replace(RE_FOCUS, (_, matched) => {
		filename = matched;
		focus = true;
		return "";
	});
	if (filename === "" && !focus) {
		const sharpIndex = info.indexOf("#");
		filename = info.slice(0, sharpIndex === -1 ? info.length : sharpIndex).trim();
		info = sharpIndex === -1 ? "" : info.slice(sharpIndex);
	}
	comment = info.trim();
	if (filename.endsWith("/")) {
		type = "folder";
		expanded = false;
		filename = removeEndingSlash(filename);
	}
	return {
		filename,
		comment,
		focus,
		expanded,
		type,
		diff
	};
}
/**
* File tree markdown plugin main function
*
* 文件树 markdown 插件主函数
*
* @param md - Markdown instance / markdown 实例
* @param options - File tree render options / 文件树渲染选项
* @param locales - Locale data / 本地化数据
*/
function fileTreePlugin(md, options = {}, locales) {
	/**
	* Get file or folder icon
	*
	* 获取文件或文件夹的图标
	*/
	const getIcon = (filename, type, mode) => {
		mode ||= options.icon || "colored";
		if (mode === "simple") return type === "folder" ? defaultFolder : defaultFile;
		return getFileIcon(filename, type);
	};
	/**
	* Recursively render file tree nodes
	*
	* 递归渲染文件树节点
	*/
	const renderFileTree = (nodes, meta) => nodes.map((node) => {
		const { level, children, filename, comment, focus, expanded, type, diff } = node;
		const isOmit = filename === "…" || filename === "...";
		if (children.length === 0 && type === "folder") children.push({
			level: level + 1,
			children: [],
			filename: "…",
			type: "file"
		});
		const nodeType = children.length > 0 ? "folder" : type;
		const renderedComment = comment ? `<template #comment>${md.renderInline(comment.replaceAll("#", "#"))}</template>` : "";
		const renderedIcon = !isOmit ? `<template #icon><VPIcon provider="iconify" name="${getIcon(filename, nodeType, meta.icon)}" /></template>` : "";
		return `<FileTreeNode${stringifyAttrs({
			expanded: nodeType === "folder" ? expanded : false,
			focus,
			type: nodeType,
			diff,
			filename,
			level
		}, false, ["filename"])}>
${renderedIcon}${renderedComment}${children.length > 0 ? renderFileTree(children, meta) : ""}
</FileTreeNode>`;
	}).join("\n");
	return createContainerSyntaxPlugin(md, "file-tree", (tokens, index, _, env) => {
		const token = tokens[index];
		const nodes = parseFileTreeRawContent(token.content);
		const meta = token.meta;
		const cmdText = fileTreeToCMDText(nodes).trim();
		const data = locales[resolveLocalePath(locales, ensureLeadingSlash$1(env.filePathRelative || ""))] ?? {};
		return `<div class="vp-file-tree">${meta.title ? `<p class="vp-file-tree-title">${meta.title}</p>` : ""}<VPCopyButton text="${encodeData(cmdText)}" encode aria-label="${data.copy || "Copy"}" data-copied="${data.copied || "Copied"}" />${renderFileTree(nodes, meta)}</div>\n`;
	});
}
/**
* Convert file tree to command line text format
*
* 将文件树转换为命令行文本格式
*
* @param nodes - File tree nodes / 文件树节点
* @param prefix - Line prefix / 行前缀
* @returns CMD text / CMD 文本
*/
function fileTreeToCMDText(nodes, prefix = "") {
	let content = prefix ? "" : ".\n";
	for (let i = 0, l = nodes.length; i < l; i++) {
		const { filename, children } = nodes[i];
		content += `${prefix + (i === l - 1 ? "└── " : "├── ")}${filename}\n`;
		const child = children.filter((n) => n.filename !== "…");
		if (child.length) content += fileTreeToCMDText(child, prefix + (i === l - 1 ? "    " : "│   "));
	}
	return content;
}
//#endregion
//#region src/node/container/langRepl.ts
/**
* Language REPL plugin - Enable interactive code playgrounds
*
* 语言 REPL 插件 - 启用交互式代码游乐场
*
* Supports: kotlin, go, rust, python
*
* @param app - VuePress app / VuePress 应用
* @param md - Markdown-it instance / Markdown-it 实例
* @param options - REPL options / REPL 选项
* @param options.theme - Editor theme / 编辑器主题
* @param options.go - Enable Go playground / 启用 Go 游乐场
* @param options.kotlin - Enable Kotlin playground / 启用 Kotlin 游乐场
* @param options.rust - Enable Rust playground / 启用 Rust 游乐场
* @param options.python - Enable Python playground / 启用 Python 游乐场
*/
async function langReplPlugin(app, md, { theme, go = false, kotlin = false, rust = false, python = false }) {
	/**
	* Create container for specific language
	*
	* 为特定语言创建容器
	*/
	const container = (lang) => createContainerPlugin(md, `${lang}-repl`, {
		before(info) {
			const { attrs } = resolveAttrs(info);
			const { editable, title } = attrs;
			return `<CodeRepl${stringifyAttrs({
				editable,
				title: title || `${lang} playground`
			})}>`;
		},
		after: () => "</CodeRepl>"
	});
	if (kotlin) container("kotlin");
	if (go) container("go");
	if (rust) container("rust");
	if (python) container("python");
	theme ??= {
		light: "github-light",
		dark: "github-dark"
	};
	const data = { grammars: {} };
	try {
		const themesPath = path.dirname(resolveModule("tm-themes"));
		const grammarsPath = path.dirname(resolveModule("tm-grammars"));
		const readTheme = (theme) => read(path.join(themesPath, "themes", `${theme}.json`));
		const readGrammar = (grammar) => read(path.join(grammarsPath, "grammars", `${grammar}.json`));
		if (typeof theme === "string") data.theme = await readTheme(theme);
		else data.theme = await Promise.all([readTheme(theme.light), readTheme(theme.dark)]).then(([light, dark]) => ({
			light,
			dark
		}));
		if (kotlin) data.grammars.kotlin = await readGrammar("kotlin");
		if (go) data.grammars.go = await readGrammar("go");
		if (rust) data.grammars.rust = await readGrammar("rust");
		if (python) data.grammars.python = await readGrammar("python");
	} catch {
		/* istanbul ignore next -- @preserve */
		logger.error("[vuepress-plugin-md-power]", `Failed to load packages: ${colors.green("tm-themes")}, ${colors.green("tm-grammars")}, Please install them manually.`);
	}
	await app.writeTemp("internal/md-power/replEditorData.js", `export default ${JSON.stringify(data, null, 2)}`);
}
/**
* Read JSON file
*
* 读取 JSON 文件
*
* @param file - File path / 文件路径
* @returns Parsed JSON / 解析后的 JSON
*/
async function read(file) {
	try {
		const content = await promises.readFile(file, "utf-8");
		return JSON.parse(content);
	} catch {}
}
//#endregion
//#region src/node/container/npmToPreset.ts
/**
* Allowed package managers list
*
* 允许的包管理器列表
*/
const ALLOW_LIST = [
	"npm",
	"pnpm",
	"yarn",
	"bun",
	"deno"
];
/**
* Boolean flags for npm commands
*
* npm 命令的布尔标志
*/
const BOOL_FLAGS = [
	"--no-save",
	"-B",
	"--save-bundle",
	"--save-dev",
	"-D",
	"--save-prod",
	"-P",
	"--save-peer",
	"-O",
	"--save-optional",
	"-E",
	"--save-exact",
	"-y",
	"--yes",
	"-g",
	"--global"
];
/**
* Default tabs to display
*
* 默认显示的标签
*/
const DEFAULT_TABS = [
	"npm",
	"pnpm",
	"yarn"
];
/**
* Package manager configurations
*
* 包管理器配置
*/
const MANAGERS_CONFIG = {
	install: {
		pattern: /(?:^|\s)npm\s+(?:install|i)$/,
		pnpm: { cli: "pnpm install" },
		yarn: { cli: "yarn" },
		bun: { cli: "bun install" },
		deno: { cli: "deno install" }
	},
	add: {
		pattern: /(?:^|\s)npm\s+(?:install|i|add)(?:\s|$)/,
		pnpm: {
			cli: "pnpm add",
			flags: {
				"--no-save": "",
				"-B": "",
				"--save-bundle": ""
			}
		},
		yarn: {
			cli: "yarn add",
			flags: {
				"--save-dev": "--dev",
				"--save-prod": "--prod",
				"-P": "",
				"--save-peer": "--peer",
				"--save-optional": "--optional",
				"--no-save": "",
				"--save-exact": "--exact",
				"-B": "",
				"--save-bundle": ""
			}
		},
		bun: {
			cli: "bun add",
			flags: {
				"--save-dev": "--development",
				"-P": "",
				"--save-prod": "",
				"--save-peer": "",
				"-O": "--optional",
				"--save-optional": "--optional",
				"--no-save": "",
				"--save-exact": "--exact",
				"-B": "",
				"--save-bundle": ""
			}
		},
		deno: {
			cli: "deno add",
			flags: {
				"-g": "",
				"--global": "",
				"--save-dev": "--dev",
				"-P": "",
				"--save-prod": "",
				"--save-peer": "",
				"-O": "",
				"--save-optional": "",
				"--no-save": "",
				"-E": "",
				"--save-exact": "",
				"-B": "",
				"--save-bundle": ""
			}
		}
	},
	run: {
		pattern: /(?:^|\s)npm\s+(?:run|run-script|rum|urn)(?:\s|$)/,
		pnpm: {
			cli: "pnpm",
			flags: {
				"-w": "-F",
				"--workspace": "--filter",
				"--": ""
			}
		},
		yarn: {
			cli: "yarn",
			flags: {
				"-w": "",
				"--workspace": ""
			}
		},
		bun: {
			cli: "bun run",
			flags: {
				"-w": "--filter",
				"--workspace": "--filter"
			}
		},
		deno: {
			cli: "deno run",
			flags: {
				"-w": "",
				"--workspace": ""
			}
		}
	},
	create: {
		pattern: /(?:^|\s)npm\s+create\s/,
		pnpm: {
			cli: "pnpm create",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		yarn: {
			cli: "yarn create",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		bun: {
			cli: "bun create",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		deno: {
			cli: "deno run -A ",
			flags: {
				"-y": "",
				"--yes": ""
			}
		}
	},
	init: {
		pattern: /(?:^|\s)npm\s+init/,
		pnpm: {
			cli: "pnpm init",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		yarn: {
			cli: "yarn init",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		bun: {
			cli: "bun init",
			flags: {
				"-y": "",
				"--yes": ""
			}
		},
		deno: {
			cli: "deno init",
			flags: {
				"-y": "",
				"--yes": ""
			}
		}
	},
	npx: {
		pattern: /(?:^|\s)npx\s+/,
		pnpm: { cli: "pnpm dlx" },
		yarn: { cli: "yarn dlx" },
		bun: { cli: "bunx" },
		deno: { cli: "deno run -A" }
	},
	remove: {
		pattern: /(?:^|\s)npm\s+(?:uninstall|r|rm|remove|unlink|un)(?:\s|$)/,
		pnpm: {
			cli: "pnpm remove",
			flags: {
				"--no-save": "",
				"--save": "",
				"-S": ""
			}
		},
		yarn: {
			cli: "yarn remove",
			flags: {
				"--save-dev": "--dev",
				"--save": "",
				"-S": "",
				"-g": "",
				"--global": ""
			}
		},
		bun: {
			cli: "bun remove",
			flags: {
				"--save-dev": "--development",
				"--save": "",
				"-S": "",
				"-g": "",
				"--global": ""
			}
		},
		deno: {
			cli: "deno uninstall",
			flags: {
				"--save-dev": "--dev",
				"--save": "",
				"-S": ""
			}
		}
	},
	ci: {
		pattern: /(?:^|\s)npm\s+ci$/,
		pnpm: { cli: "pnpm install --frozen-lockfile" },
		yarn: { cli: "yarn install --immutable" },
		bun: { cli: "bun install --frozen-lockfile" },
		deno: { cli: "deno install --frozen" }
	}
};
//#endregion
//#region src/node/container/npmTo.ts
/**
* npm-to plugin - Convert npm commands to multiple package manager commands
*
* 注册 npm-to 容器插件，将 npm 代码块自动转换为多包管理器命令分组
*
* @param md - Markdown instance / Markdown 实例
* @param options - npm-to options / npm-to 选项
*/
function npmToPlugins(md, options = {}) {
	const opt = isArray(options) ? { tabs: options } : options;
	const defaultTabs = opt.tabs?.length ? opt.tabs : DEFAULT_TABS;
	createContainerPlugin(md, "npm-to", {
		before: (info, tokens, idx, _opt, env) => {
			const { attrs } = resolveAttrs(info);
			const tabs = attrs.tabs ? attrs.tabs.split(/,\s*/) : defaultTabs;
			const token = tokens[idx + 1];
			if (token.type === "fence") {
				const content = token.content;
				token.hidden = true;
				token.type = "text";
				token.content = "";
				const lines = content.split(/(\n|\s*&&\s*)/);
				return md.render(resolveNpmTo(lines, token.info.trim(), idx, tabs), cleanMarkdownEnv(env));
			}
			logger$1.warn("npm-to", `Invalid npm-to container in ${colors.gray(env.filePathRelative || env.filePath)}`);
			return "";
		},
		after: () => ""
	});
}
/**
* Convert npm commands to package manager command groups
*
* 将 npm 命令转换为各包管理器命令分组
*
* @param lines - Command line array / 命令行数组
* @param info - Code block type / 代码块类型
* @param idx - Token index / token 索引
* @param tabs - Package managers to support / 需要支持的包管理器
* @returns code-tabs formatted string / code-tabs 格式字符串
*/
function resolveNpmTo(lines, info, idx, tabs) {
	tabs = validateTabs(tabs);
	const res = [];
	const map = {};
	for (const tab of tabs) {
		const newLines = [];
		for (const line of lines) {
			const config = findConfig(line);
			if (config && config[tab]) {
				const parsed = map[line] ??= parseLine(line);
				const { cli, flags } = config[tab];
				let newLine = `${parsed.env ? `${parsed.env} ` : ""}${cli}`;
				if (parsed.args && flags) {
					let args = parsed.args;
					for (const [key, value] of Object.entries(flags)) args = args.replaceAll(key, value);
					newLine += ` ${args.replace(/\s+-/g, " -").trim()}`;
				}
				if (parsed.cmd) newLine += ` ${parsed.cmd}`;
				if (parsed.scriptArgs) newLine += ` ${parsed.scriptArgs}`;
				newLines.push(newLine.trim());
			} else newLines.push(line);
		}
		res.push(`@tab ${tab}\n\`\`\`${info}\n${newLines.join("")}\n\`\`\``);
	}
	return `:::code-tabs#npm-to-${tabs.join("-")}\n${res.join("\n")}\n:::`;
}
/**
* Find package manager config by command line content
*
* 根据命令行内容查找对应的包管理器配置
*
* @param line - Command line / 命令行
* @returns Command config / 命令配置
*/
function findConfig(line) {
	for (const { pattern, ...config } of Object.values(MANAGERS_CONFIG)) if (pattern.test(line)) return config;
}
/**
* Validate tabs, return allowed package manager list
*
* 校验 tabs 合法性，返回允许的包管理器列表
*
* @param tabs - Package managers / 包管理器
* @returns Validated tabs / 验证后的标签
*/
function validateTabs(tabs) {
	tabs = tabs.filter((tab) => ALLOW_LIST.includes(tab));
	if (tabs.length === 0) return DEFAULT_TABS;
	return tabs;
}
/**
* Regex for parsing npm/npx commands
*
* 解析 npm/npx 命令的正则
*/
const LINE_REG = /(.*)(npm|npx)\s+(.*)/;
/**
* Parse a line of npm/npx command, split env, command, args, etc.
*
* 解析一行 npm/npx 命令，拆分出环境变量、命令、参数等
*
* @param line - Command line / 命令行
* @returns Parsed result / 解析结果
*/
function parseLine(line) {
	const match = line.match(LINE_REG);
	if (!match) return false;
	const [, env, cli, rest] = match;
	const idx = rest.trim().indexOf(" ");
	if (cli === "npx") {
		let cmd = "";
		let scriptArgs = "";
		if (idx !== -1) {
			cmd = rest.slice(0, idx);
			scriptArgs = rest.slice(idx + 1).trim();
		} else cmd = rest;
		return {
			env,
			cli,
			cmd,
			scriptArgs
		};
	}
	if (idx === -1) return {
		env,
		cli: `${cli} ${rest.trim()}`,
		cmd: ""
	};
	return {
		env,
		cli: `${cli} ${rest.slice(0, idx)}`,
		...parseArgs(rest.slice(idx + 1))
	};
}
/**
* Parse npm command arguments, distinguish command, args, script args
*
* 解析 npm 命令参数，区分命令、参数、脚本参数
*
* @param line - Arguments line / 参数字符串
* @returns Parsed args / 解析后的参数
*/
function parseArgs(line) {
	line = line?.trim();
	const [npmArgs, scriptArgs] = line.split(/\s+--\s+/);
	let cmd = "";
	let args = "";
	if (npmArgs[0] !== "-") if (npmArgs[0] === "\"" || npmArgs[0] === "'") {
		const idx = npmArgs.slice(1).indexOf(npmArgs[0]);
		cmd = npmArgs.slice(0, idx + 2);
		args = npmArgs.slice(idx + 2);
	} else {
		const idx = npmArgs.indexOf(" -");
		if (idx === -1) cmd = npmArgs;
		else {
			cmd = npmArgs.slice(0, idx);
			args = npmArgs.slice(idx + 1);
		}
	}
	else {
		let newLine = "";
		let value = "";
		let isQuote = false;
		let isBool = false;
		let isNextValue = false;
		let quote = "";
		for (let i = 0; i < npmArgs.length; i++) {
			const v = npmArgs[i];
			if (!isQuote && (v === "\"" || v === "'")) {
				quote = v;
				isQuote = true;
				value += v;
			} else if (isQuote && v === quote) {
				isQuote = false;
				value += v;
			} else if ((v === " " || v === "=" || i === npmArgs.length - 1) && !isQuote && value) {
				if (i === npmArgs.length - 1) value += v;
				const isKey = value[0] === "-";
				if (isKey) {
					isBool = BOOL_FLAGS.includes(value);
					isNextValue = !isBool;
				}
				if (!isKey && !isNextValue) cmd += ` ${value}`;
				else {
					newLine += `${value}${i !== npmArgs.length - 1 ? v : ""}`;
					if (!isKey && isNextValue) isNextValue = false;
				}
				value = "";
			} else value += v;
		}
		args = newLine;
	}
	return {
		cmd: cmd.trim(),
		args: args.trim(),
		scriptArgs
	};
}
//#endregion
//#region src/node/container/steps.ts
/**
* Steps container plugin
*
* 步骤容器插件
*
* Syntax:
* ```md
* :::steps
* 1. 步骤 1
*    xxx
* 2. 步骤 2
*    xxx
* 3. ...
* :::
* ```
*
* @param md - Markdown instance / Markdown 实例
*/
function stepsPlugin(md) {
	createContainerPlugin(md, "steps", { before: () => "<div class=\"vp-steps\">" });
}
//#endregion
//#region src/node/container/table.ts
/**
* Table plugin - Wrap table with container for enhanced features
*
* 表格插件 - 通过容器语法将表格包裹起来，为表格提供增强功能
*
* @example
* ```md
* ::: table title="表格标题" max-content copy align="center" hl-rows="warning:1,2,3;error:4,5,6" hl-cols="warning:1;error:2,3" hl-cells="warning:(1,2)(2,3);"
*
* | xx | xx | xx |
* | -- | -- | -- |
* | xx | xx | xx |
* :::
* ```
*
* @param md - Markdown instance / Markdown 实例
* @param options - Table container options / 表格容器选项
*/
function tablePlugin(md, options = {}) {
	createContainerSyntaxPlugin(md, "table", (tokens, index, opt, env) => {
		const { hlCols = "", hlRows = "", hlCells = "", ...meta } = tokens[index].meta;
		const props = {
			copy: true,
			maxContent: false,
			fullWidth: false,
			...options,
			...meta
		};
		const content = tokens[index].content;
		if (props.copy) {
			props.copy = props.copy === true ? "all" : props.copy;
			if (props.copy === "all" || props.copy === "md") props.markdown = encodeData(content.trim());
		}
		if (!hlCols && !hlRows && !hlCells) return `<VPTable ${stringifyAttrs(props)}>${md.render(content, env)}</VPTable>`;
		const rows = parseHl(hlRows);
		const cols = parseHl(hlCols);
		const cells = parseHlCells(hlCells);
		const tableTokens = md.parse(content, env);
		let isTable = false;
		let colIndex = 0;
		let rowIndex = 0;
		for (const token of tableTokens) {
			if (token.type === "table_open") isTable = true;
			if (token.type === "table_close") isTable = false;
			if (!isTable) continue;
			if (token.type === "tr_open") {
				rowIndex++;
				colIndex = 0;
			}
			if (token.type === "th_open" || token.type === "td_open") {
				colIndex++;
				const classes = cells[rowIndex]?.[colIndex] || rows[rowIndex] || cols[colIndex];
				if (classes) token.attrJoin("class", classes);
			}
		}
		return `<VPTable ${stringifyAttrs(props)}>${md.renderer.render(tableTokens, opt, env)}</VPTable>`;
	});
}
/**
* Parse highlight string
*
* 解析高亮字符串
*
* @param hl - Highlight string / 高亮字符串
* @returns Parsed highlight map / 解析后的高亮映射
*/
function parseHl(hl) {
	const res = {};
	if (!hl) return res;
	hl.split(";").forEach((item) => {
		const [key, value = "1"] = item.split(":");
		String(value).split(",").forEach((v) => res[v.trim()] = key.trim());
	});
	return res;
}
/**
* Parse highlight cells string
*
* 解析高亮单元格字符串
*
* @param hl - Highlight cells string / 高亮单元格字符串
* @returns Parsed highlight cells map / 解析后的高亮单元格映射
*/
function parseHlCells(hl) {
	const res = {};
	if (!hl) return res;
	hl.split(";").forEach((item) => {
		const [key, value = ""] = item.split(":");
		value.trim().replace(/\s*\((\d+)\s*,\s*(\d+)\)\s*/g, (_, row, col) => {
			res[row] ??= {};
			res[row][col] = key.trim();
			return "";
		});
	});
	return res;
}
//#endregion
//#region src/node/container/tabs.ts
/**
* Tabs plugin - Enable tabs container
*
* 标签页插件 - 启用标签页容器
*
* Syntax:
* ```md
* :::tabs
* @tab Tab 1
* Content 1
* @tab Tab 2
* Content 2
* :::
* ```
*
* @param md - Markdown-it instance / Markdown-it 实例
*/
const tabs = (md) => {
	tab(md, {
		name: "tabs",
		openRender: ({ active, data }, tokens, index, _, env) => {
			const { meta } = tokens[index];
			const titles = data.map(({ title }) => md.renderInline(title, cleanMarkdownEnv(env)));
			return `<Tabs id="${index}" :data='${stringifyProp(data.map((item, dataIndex) => {
				const { id = titles[dataIndex] } = item;
				return { id };
			}))}'${active === -1 ? "" : ` :active="${active}"`}${meta.id ? ` tab-id="${meta.id}"` : ""}>
${titles.map((title, titleIndex) => `<template #title${titleIndex}="{ value, isActive }">${title}</template>`).join("")}`;
		},
		closeRender: () => `</Tabs>`,
		tabOpenRender: ({ index }) => `<template #tab${index}="{ value, isActive }">`,
		tabCloseRender: () => `</template>`
	});
};
//#endregion
//#region src/node/container/timeline.ts
/**
* Regex for matching attribute keys
*
* 匹配属性键的正则
*/
const RE_KEY = /(\w+)=\s*/;
/**
* Regex for searching next attribute key
*
* 搜索下一个属性键的正则
*/
const RE_SEARCH_KEY = /\s+\w+=\s*|$/;
/**
* Regex for cleaning quote values
*
* 清理引号值的正则
*/
const RE_CLEAN_VALUE = /(?<quote>["'])(.*?)(\k<quote>)/;
/**
* Timeline plugin - Enable timeline container
*
* 时间线插件 - 启用时间线容器
*
* @param md - Markdown instance / Markdown 实例
*/
function timelinePlugin(md) {
	createContainerPlugin(md, "timeline", {
		before(info, tokens, index) {
			parseTimeline(tokens, index);
			const { attrs } = resolveAttrs(info);
			attrs.card ??= void 0;
			return `<VPTimeline${stringifyAttrs(attrs, true)}>`;
		},
		after: () => "</VPTimeline>"
	});
	md.renderer.rules.timeline_item_open = (tokens, idx) => {
		const attrs = tokens[idx].meta;
		attrs.card ??= void 0;
		const icon = attrs.icon;
		return `<VPTimelineItem${stringifyAttrs(attrs, true)}>${icon ? `<template #icon><VPIcon provider="iconify" name="${icon}"/></template>` : ""}`;
	};
	md.renderer.rules.timeline_item_close = () => "</VPTimelineItem>";
	md.renderer.rules.timeline_item_title_open = () => "<template #title>";
	md.renderer.rules.timeline_item_title_close = () => "</template>";
}
/**
* Parse timeline tokens
*
* 解析时间线令牌
*
* @param tokens - Token array / 令牌数组
* @param index - Start index / 起始索引
*/
function parseTimeline(tokens, index) {
	const listStack = [];
	for (let i = index + 1; i < tokens.length; i++) {
		const token = tokens[i];
		if (token.type === "container_timeline_close") break;
		if (token.type === "bullet_list_open") {
			listStack.push(0);
			if (listStack.length === 1) token.hidden = true;
		} else if (token.type === "bullet_list_close") {
			listStack.pop();
			if (listStack.length === 0) token.hidden = true;
		} else if (token.type === "list_item_open") {
			if (listStack.length === 1) {
				token.type = "timeline_item_open";
				tokens[i + 1].type = "timeline_item_title_open";
				tokens[i + 3].type = "timeline_item_title_close";
				const inlineToken = tokens[i + 2];
				const softbreakIndex = inlineToken.children.findLastIndex((token) => token.type === "softbreak");
				if (softbreakIndex !== -1) {
					const lastToken = inlineToken.children[inlineToken.children.length - 1];
					token.meta = extractTimelineAttributes(lastToken.content.trim());
					if (!isEmptyObject(token.meta)) inlineToken.children = inlineToken.children.slice(0, softbreakIndex);
				} else token.meta = {};
			}
		} else if (token.type === "list_item_close") {
			if (listStack.length === 1) token.type = "timeline_item_close";
		}
	}
}
/**
* Extract timeline attributes from raw text
*
* 从原始文本中提取时间线属性
*
* @param rawText - Raw attribute text / 原始属性文本
* @returns Timeline item metadata / 时间线项元数据
*/
function extractTimelineAttributes(rawText) {
	const attrKeys = [
		"time",
		"type",
		"icon",
		"line",
		"color",
		"card",
		"placement"
	];
	const attrs = {};
	let buffer = rawText.trim();
	while (buffer.length) {
		const keyMatch = buffer.match(RE_KEY);
		if (!keyMatch) break;
		const matchedKey = keyMatch[1].toLowerCase();
		if (!attrKeys.includes(matchedKey)) break;
		const keyEnd = keyMatch.index + keyMatch[0].length;
		buffer = buffer.slice(keyEnd);
		let valueEnd = buffer.search(RE_SEARCH_KEY);
		/* istanbul ignore if -- @preserve */
		if (valueEnd === -1) valueEnd = buffer.length;
		attrs[matchedKey] = buffer.slice(0, valueEnd).trim().replace(RE_CLEAN_VALUE, "$2");
		buffer = buffer.slice(valueEnd);
	}
	return attrs;
}
//#endregion
//#region src/node/container/window.ts
const RE_IMAGE_SYNTAX = /^!?\[[^\]]*\]\([^)]+\)$/;
const render = (info, tokens, idx) => {
	const elms = [];
	for (let i = idx + 1; i < tokens.length; i++) {
		if (tokens[i].type === "container_window_close") break;
		elms.push(tokens[i]);
	}
	const { attrs } = resolveAttrs(info);
	let onlyImg = false;
	if (elms.length === 1) {
		const { type } = elms[0];
		const content = elms[0].content.trim();
		if (type === "html_block" && (content.startsWith("<img") || content.startsWith("<picture"))) onlyImg = true;
	}
	if (elms.length === 3) {
		const [op, img, cp] = elms;
		if (op.type === "paragraph_open" && cp.type === "paragraph_close" && img.type === "inline" && RE_IMAGE_SYNTAX.test(img.content.trim())) {
			op.type = "text";
			cp.type = "text";
			onlyImg = true;
		}
	}
	const { title, height, noPadding } = attrs;
	const gap = isString(attrs.gap) || isNumber(attrs.gap) ? parseRect(attrs.gap) : onlyImg || noPadding ? "0" : "20px";
	const classes = ["window-wrapper"];
	title && classes.push("has-title");
	return `<article class="${classes.join(" ")}">
  <header class="window-header">
    <div class="window-left"><i></i><i></i><i></i></div>
    ${title ? `<div class="window-center"><h4 class="window-title ignore-header"><span>${title}</span><i class="vpi-window-reload"></i></h4></div>` : ""}
    <div class="window-right"><i class="vpi-window-share"></i><i class="vpi-window-add"></i><i class="vpi-window-copy"></i></div>
  </header>
  <section class="window-content" style="--window-gap:${gap};${height ? `--window-height:${parseRect(height)}` : ""}">`;
};
/**
* window plugin - Enable window container
*
* 窗口插件 - 启用窗口容器
*
* Syntax: :::window title="xxx" height="100px" gap="20px"
* 语法：:::window title="xxx" height="100px" gap="20px"
*
* @param md - Markdown instance / Markdown 实例
*/
function windowPlugin(md) {
	const after = () => "</section></article>";
	createContainerPlugin(md, "window", {
		before: render,
		after
	});
	createContainerPlugin(md, "demo-wrapper", {
		before: (info, tokens, idx, _, env) => {
			logger$1.warn("container", `::: demo-wrapper container is deprecated, please use ::: window container instead. (${colors.gray(env.filePathRelative || "")})`);
			return render(info, tokens, idx, _, env);
		},
		after
	});
}
//#endregion
//#region src/node/container/index.ts
/**
* Container plugin - Register all container plugins
*
* 容器插件 - 注册所有容器插件
*
* @param app - VuePress app / VuePress 应用
* @param md - Markdown instance / Markdown 实例
* @param options - Plugin options / 插件选项
* @param locales - Locale configuration / 本地化配置
*/
async function containerPlugin(app, md, options, locales) {
	alignPlugin(md);
	tabs(md);
	codeTabs(md, options.codeTabs);
	windowPlugin(md);
	stepsPlugin(md);
	cardPlugin(md);
	if (options.encrypt) encryptPlugin(app, md, typeof options.encrypt === "boolean" ? {} : options.encrypt);
	if (options.npmTo) npmToPlugins(md, typeof options.npmTo === "boolean" ? {} : options.npmTo);
	if (options.repl) await langReplPlugin(app, md, options.repl);
	if (options.fileTree) fileTreePlugin(md, isPlainObject(options.fileTree) ? options.fileTree : {}, findLocales(locales, "common"));
	if (options.codeTree) codeTreePlugin(md, app, isPlainObject(options.codeTree) ? options.codeTree : {});
	if (options.timeline) timelinePlugin(md);
	if (options.collapse) collapsePlugin(md);
	if (options.chat) chatPlugin(md);
	if (options.field) fieldPlugin(md);
	if (options.table) tablePlugin(md, isPlainObject(options.table) ? options.table : {});
}
//#endregion
//#region src/node/demo/markdown.ts
function markdownEmbed(app, md, env, { url, title, desc, codeSetting = "", expanded = false }) {
	const filepath = findFile(app, env, url);
	const code = readFileSync(filepath);
	if (code === false) {
		logger$1.warn("demo-markdown", `Cannot read markdown file: ${colors.gray(filepath)}\n  at: ${colors.gray(env.filePathRelative || "")}`);
		return "";
	}
	const demo = {
		type: "markdown",
		path: filepath
	};
	env.demoFiles ??= [];
	if (!env.demoFiles.some((d) => d.path === filepath)) env.demoFiles.push(demo);
	return `<VPDemoBasic${stringifyAttrs({
		type: "markdown",
		title,
		desc,
		expanded
	})}>
    ${md.render(code, {
		filepath: env.filePath,
		filepathRelative: env.filePathRelative
	})}
    <template #code>
      ${md.render(`\`\`\`md ${codeSetting}\n${code}\n\`\`\``, {})}
    </template>
  </VPDemoBasic>`;
}
const markdownContainerRender = {
	before(app, md, env, meta, codeMap) {
		const { title, desc, expanded = false } = meta;
		const code = codeMap.md || "";
		return `<VPDemoBasic${stringifyAttrs({
			type: "markdown",
			title,
			desc,
			expanded
		})}>
      ${md.render(code, {
			filepath: env.filePath,
			filepathRelative: env.filePathRelative
		})}
      <template #code>`;
	},
	after: () => "</template></VPDemoBasic>"
};
//#endregion
//#region src/node/utils/package.ts
/**
* Get the default export from a module, or the module itself if no default export
*
* 从模块获取默认导出，如果没有默认导出则返回模块本身
*
* @param m - Module or Promise of module / 模块或模块的 Promise
* @returns Default export or module itself / 默认导出或模块本身
* @typeParam T - Module type / 模块类型
*/
async function interopDefault(m) {
	const resolved = await m;
	return resolved.default || resolved;
}
//#endregion
//#region src/node/demo/supports/compiler.ts
const cache = new LRUCache({ max: 64 });
const compiler = {
	script: importer(async () => {
		const { transform } = await import("esbuild");
		return transform;
	}),
	less: importer(() => import("less")),
	sass: importer(async () => {
		if (isPackageExists("sass-embedded")) return await import("sass-embedded");
		return await import("sass");
	}),
	stylus: importer(() => import("stylus"))
};
async function compileScript(source, type) {
	const key = `${type}:::${source}`;
	if (cache.has(key)) return cache.get(key);
	const res = await (await compiler.script())(source, {
		target: "es2018",
		platform: "browser",
		format: "cjs",
		loader: type === "ts" ? "ts" : "js",
		sourcemap: false
	});
	cache.set(key, res.code);
	return res.code;
}
async function compileStyle(source, type) {
	const key = `${type}:::${source}`;
	if (cache.has(key)) return cache.get(key);
	if (type === "css") return source;
	if (type === "less") {
		const res = await (await compiler.less()).render(source);
		cache.set(key, res.css);
		return res.css;
	}
	if (type === "scss") {
		const res = (await compiler.sass()).compileString(source);
		cache.set(key, res.css);
		return res.css;
	}
	if (type === "stylus") {
		const res = (await compiler.stylus()).render(source);
		cache.set(key, res);
		return res;
	}
	return source;
}
function importer(func) {
	let imported;
	return async () => {
		if (!imported) imported = interopDefault(await func());
		return imported;
	};
}
//#endregion
//#region src/node/demo/supports/insertScript.ts
const SCRIPT_RE$1 = /<script.*?>/;
function insertSetupScript({ export: name, path }, env) {
	const imports = `import ${name ? `${name} from ` : ""}'${path}';`;
	const scriptSetup = env.sfcBlocks.scriptSetup ??= {
		type: "script",
		content: "<script setup>\n<\/script>",
		contentStripped: "",
		tagOpen: "<script setup>",
		tagClose: "<\/script>"
	};
	scriptSetup.contentStripped = `${imports}\n${scriptSetup.contentStripped}`;
	scriptSetup.content = scriptSetup.content.replace(SCRIPT_RE$1, (matched) => `${matched}\n${imports}`);
}
//#endregion
//#region src/node/demo/watcher.ts
/**
* 消除异步编译 demo 代码 与 markdown 同步 render 的时间差问题
* 确保 在 vuepress onPrepared 阶段完成所有 demo 代码的编译与输出
*/
let renderDone = null;
let renderCount = 0;
let renderPromise;
function createDemoRender() {
	renderPromise = new Promise((resolve) => {
		renderDone = resolve;
	});
}
async function waitDemoRender() {
	if (renderCount === 0) {
		renderDone?.();
		renderDone = null;
	}
	await renderPromise;
}
function markDemoRender() {
	renderCount++;
}
function checkDemoRender() {
	if (renderCount > 0) renderCount--;
	if (renderCount === 0) {
		renderDone?.();
		renderDone = null;
	}
}
let watcher = null;
const tasks = {};
const target$2 = "md-power/demo/watcher.txt";
function demoWatcher(app, watchers) {
	if (!watcher) watcher = watch([], { ignoreInitial: true });
	Object.keys(tasks).forEach((path) => {
		watcher.add(path);
	});
	const code = readFileSync(app.dir.temp(target$2));
	if (code) {
		const paths = JSON.parse(code || "{}");
		Object.entries(paths).forEach(([path, output]) => {
			watcher.add(path);
			tasks[path] = output;
		});
	}
	updateWatchFiles(app);
	watcher.on("change", (path) => {
		if (tasks[path]) {
			const code = readFileSync(path);
			if (code === false) return;
			compileCode(parseEmbedCode(code), tasks[path]);
		}
	});
	watcher.on("unlink", (path) => {
		deleteKey(tasks, path);
		watcher.unwatch(path);
	});
	watchers.push({ close: () => {
		watcher.close();
		watcher = null;
	} });
}
function addTask(app, path, output) {
	if (tasks[path]) return;
	tasks[path] = output;
	if (watcher) watcher.add(path);
	updateWatchFiles(app);
}
async function updateWatchFiles(app) {
	await fs$1.promises.mkdir(app.dir.temp(path$1.dirname(target$2)), { recursive: true });
	await fs$1.promises.writeFile(app.dir.temp(target$2), JSON.stringify(tasks));
}
//#endregion
//#region src/node/demo/normal.ts
const CONFIG_RE = /<script type="config">([\s\S]*?)\n<\/script>/;
const SCRIPT_RE = /<script\s?(?:lang="(\w+)")?>([\s\S]*?)\n<\/script>/;
const STYLE_RE$1 = /<style\s?(?:lang="(\w+)")?>([\s\S]*?)\n<\/style>/;
const scriptSupported = ["ts", "js"];
const styleSupported = [
	"css",
	"scss",
	"less",
	"stylus",
	"styl"
];
const target$1 = "md-power/demo/normal";
const FENCE = "```";
function parseEmbedCode(code) {
	const res = {
		html: "",
		css: "",
		script: "",
		imports: "",
		jsType: "js",
		cssType: "css"
	};
	res.html = code.replace(CONFIG_RE, (_, config) => {
		res.imports = config;
		return "";
	}).replace(SCRIPT_RE, (_, lang, script) => {
		res.script = script;
		res.jsType = scriptSupported.includes(lang) ? lang : "js";
		return "";
	}).replace(STYLE_RE$1, (_, lang, style) => {
		res.css = style;
		res.cssType = styleSupported.includes(lang) ? lang === "styl" ? "stylus" : lang : "css";
		return "";
	});
	return res;
}
function codeToHtml(md, source, info) {
	let content = "::: code-tabs\n";
	if (source.html) content += `@tab HTML\n${FENCE}html ${info}\n${source.html.trim()}\n${FENCE}\n`;
	if (source.script) {
		const title = source.jsType === "ts" ? "Typescript" : "Javascript";
		content += `@tab ${title}\n${FENCE}${source.jsType} ${info}\n${source.script.trim()}\n${FENCE}\n`;
	}
	if (source.css) {
		const title = source.cssType === "stylus" ? "Stylus" : source.cssType.toUpperCase();
		content += `@tab ${title}\n${FENCE}${source.cssType} ${info}\n${source.css.trim()}\n${FENCE}\n`;
	}
	content += "\n:::";
	return md.render(content, {});
}
async function compileCode(code, output) {
	markDemoRender();
	const res = {
		jsLib: [],
		cssLib: [],
		script: "",
		css: "",
		html: ""
	};
	if (!fs$1.existsSync(output)) writeFileSync(output, `import { ref } from "vue"\nexport default ref(${JSON.stringify(res, null, 2)})`);
	try {
		if (code.imports) {
			const imports = JSON.parse(code.imports);
			res.jsLib = imports.jsLib ?? [];
			res.cssLib = imports.cssLib ?? [];
		}
		if (code.script) res.script = await compileScript(code.script.trim(), code.jsType);
		if (code.css) res.css = await compileStyle(code.css.trim(), code.cssType);
		if (code.html) res.html = code.html.trim();
	} catch (e) {
		logger$1.error("demo-normal", "demo parse error: \n", e);
	}
	writeFileSync(output, `import { ref } from "vue"\nexport default ref(${JSON.stringify(res, null, 2)})`);
	checkDemoRender();
}
function normalEmbed(app, md, env, { url, title, desc, codeSetting = "", expanded = false }) {
	const filepath = findFile(app, env, url);
	const code = readFileSync(filepath);
	if (code === false) {
		logger$1.warn("demo-normal", `Cannot read demo file: ${colors.gray(filepath)}\n  at: ${colors.gray(env.filePathRelative || "")}`);
		return "";
	}
	const source = parseEmbedCode(code);
	const prefix = (env.filePathRelative || "").replace(/\.md$/, "").replace(/\//g, "-");
	const basename = path$1.basename(filepath).replace(/-|\./g, "_");
	const name = `Demo${basename[0].toUpperCase()}${basename.slice(1)}`;
	const demo = {
		type: "normal",
		export: name,
		path: filepath
	};
	const output = app.dir.temp(target$1, `${prefix}-${name}.js`);
	compileCode(source, output);
	addTask(app, filepath, output);
	env.demoFiles ??= [];
	if (!env.demoFiles.some((d) => d.path === filepath)) {
		env.demoFiles.push(demo);
		insertSetupScript({
			...demo,
			path: output
		}, env);
	}
	return `<VPDemoNormal${stringifyAttrs({
		":config": name,
		title,
		desc,
		expanded
	})}>
    ${codeToHtml(md, source, codeSetting)}
  </VPDemoNormal>`;
}
const normalContainerRender = {
	before(app, md, env, meta, codeMap) {
		const { url, title, desc, expanded = false } = meta;
		const name = `DemoContainer${url}`;
		const prefix = (env.filePathRelative || "").replace(/\.md$/, "").replace(/\//g, "-");
		const output = app.dir.temp(path$1.join(target$1, `${prefix}-${name}.js`));
		env.demoFiles ??= [];
		if (!env.demoFiles.some((d) => d.path === output)) {
			const demo = {
				type: "normal",
				export: name,
				gitignore: true,
				path: output
			};
			env.demoFiles.push(demo);
			insertSetupScript(demo, env);
		}
		compileCode(parseContainerCode(codeMap), output);
		return `<VPDemoNormal${stringifyAttrs({
			":config": name,
			title,
			desc,
			expanded
		})}>`;
	},
	after: () => "</VPDemoNormal>",
	token(token) {
		if (token.info.trim().startsWith("json")) {
			token.hidden = true;
			token.type = "text";
			token.content = "";
		}
	}
};
function parseContainerCode(map) {
	const res = {
		html: "",
		css: "",
		script: "",
		imports: "",
		jsType: "js",
		cssType: "css"
	};
	if (map.html) res.html = map.html;
	if (map.js) {
		res.script = map.js;
		res.jsType = "js";
	}
	if (map.ts) {
		res.script = map.ts;
		res.jsType = "ts";
	}
	if (map.css) {
		res.css = map.css;
		res.cssType = "css";
	}
	if (map.less) {
		res.css = map.less;
		res.cssType = "less";
	}
	if (map.scss) {
		res.css = map.scss;
		res.cssType = "scss";
	}
	if (map.styl || map.stylus) {
		res.css = map.styl || map.stylus;
		res.cssType = "stylus";
	}
	if (map.json) res.imports = map.json;
	return res;
}
//#endregion
//#region src/node/demo/supports/alias.ts
function normalizeAlias(info) {
	const [lang] = info.trim().split(/\s+|:|\{/);
	switch (lang) {
		case "vue": return "vue";
		case "js":
		case "javascript": return "js";
		case "ts":
		case "typescript": return "ts";
		case "stylus":
		case "styl": return "stylus";
		case "md":
		case "markdown": return "md";
	}
	return lang;
}
//#endregion
//#region src/node/demo/vue.ts
function vueEmbed(app, md, env, { url, title, desc, codeSetting = "", expanded = false }) {
	const filepath = findFile(app, env, url);
	const code = readFileSync(filepath);
	if (code === false) {
		logger$1.warn("demo-vue", `Cannot read vue demo file: ${colors.gray(filepath)}\n  at: ${colors.gray(env.filePathRelative || "")}`);
		return "";
	}
	const basename = path$1.basename(filepath).replace(/-|\./g, "_");
	const ext = path$1.extname(filepath).slice(1);
	const name = `Demo${basename[0].toUpperCase()}${basename.slice(1)}`;
	const demo = {
		type: "vue",
		export: name,
		path: filepath
	};
	env.demoFiles ??= [];
	if (!env.demoFiles.some((d) => d.path === filepath)) {
		env.demoFiles.push(demo);
		insertSetupScript(demo, env);
	}
	return `<VPDemoBasic${stringifyAttrs({
		type: "vue",
		title,
		desc,
		expanded
	})}>
    <${name} />
    <template #code>
      ${md.render(`\`\`\`${ext}${codeSetting}\n${code}\n\`\`\``, {})}
    </template>
  </VPDemoBasic>`;
}
const target = "md-power/demo/vue";
const vueContainerRender = {
	before: (app, md, env, meta, codeMap) => {
		const { url, title, desc, expanded = false } = meta;
		const componentName = `DemoContainer${url}`;
		const prefix = (env.filePathRelative || "").replace(/\.md$/, "").replace(/\//g, "-");
		env.demoFiles ??= [];
		const output = app.dir.temp(path$1.join(target, `${prefix}-${componentName}`));
		if (codeMap.vue || codeMap.js || codeMap.ts) {
			let scriptOutput = output;
			let content = "";
			if (codeMap.vue) {
				scriptOutput += ".vue";
				content = transformStyle(codeMap.vue);
			} else if (codeMap.ts) {
				scriptOutput += ".ts";
				content = codeMap.ts;
			} else if (codeMap.js) {
				scriptOutput += ".js";
				content = codeMap.js;
			}
			content = transformImports(content, env.filePath || "");
			const script = {
				type: "vue",
				export: componentName,
				path: scriptOutput,
				gitignore: true
			};
			writeFileSync(scriptOutput, content);
			if (!env.demoFiles.some((d) => d.path === scriptOutput)) {
				env.demoFiles.push(script);
				insertSetupScript(script, env);
			}
		}
		if (codeMap.css || codeMap.scss || codeMap.less || codeMap.stylus) {
			let styleOutput = output;
			let content = "";
			if (codeMap.css) {
				styleOutput += ".css";
				content = codeMap.css;
			} else if (codeMap.scss) {
				styleOutput += ".scss";
				content = codeMap.scss;
			} else if (codeMap.less) {
				styleOutput += ".less";
				content = codeMap.less;
			} else if (codeMap.stylus) {
				styleOutput += ".styl";
				content = codeMap.stylus;
			}
			writeFileSync(styleOutput, content);
			const style = {
				type: "css",
				path: styleOutput,
				gitignore: true
			};
			if (!env.demoFiles.some((d) => d.path === styleOutput)) {
				env.demoFiles.push(style);
				insertSetupScript(style, env);
			}
		}
		return `<VPDemoBasic${stringifyAttrs({
			type: "vue",
			title,
			desc,
			expanded
		})}>
    <${componentName} />
    <template #code>\n`;
	},
	after: () => "</template></VPDemoBasic>"
};
const IMPORT_RE = /import\s+(?:\w+\s+from\s+)?['"]([^'"]+)['"]/g;
const STYLE_RE = /<style.*?>/;
function transformImports(code, filepath) {
	return code.replace(IMPORT_RE, (matched, url) => {
		if (url.startsWith("./") || url.startsWith("../")) return matched.replace(url, `${path$1.resolve(path$1.dirname(filepath), url)}`);
		return matched;
	});
}
function transformStyle(code) {
	return code.replace(STYLE_RE, (matched) => {
		if (matched.includes("scoped")) return matched;
		return matched.replace("<style", "<style scoped");
	});
}
//#endregion
//#region src/node/demo/demo.ts
const embedMap = {
	vue: vueEmbed,
	normal: normalEmbed,
	markdown: markdownEmbed
};
/**
* 嵌入语法
* @[demo type info](url)
*/
function demoEmbed(app, md) {
	createEmbedRuleBlock(md, {
		type: "demo",
		syntaxPattern: /^@\[demo(?:\s(vue|normal|markdown))?\s?(.*)\]\((.*)\)/,
		meta: ([, type, info, url]) => ({
			type: type || "normal",
			url,
			...resolveAttrs(info).attrs
		}),
		content: (meta, content, env) => {
			const { url, type } = meta;
			if (!url) {
				logger$1.warn("demo-vue", `Invalid filepath: ${colors.gray(url)}`);
				return content;
			}
			if (embedMap[type]) return embedMap[type](app, md, env, meta);
			return content;
		}
	});
}
const INFO_RE = /(vue|normal|markdown)?\s?(.*)/;
const renderMap = {
	vue: vueContainerRender,
	normal: normalContainerRender,
	markdown: markdownContainerRender
};
function demoContainer(app, md) {
	let currentRender;
	const render = (tokens, index, _, env) => {
		const token = tokens[index];
		if (token.nesting === 1) {
			const meta = getContainerMeta(token.info);
			meta.url = `${index}`;
			currentRender = renderMap[meta.type];
			return currentRender?.before(app, md, env, meta, parseCodeMapping(tokens, index, currentRender.token)) || "";
		} else {
			const res = currentRender?.after() || "";
			currentRender = void 0;
			return res;
		}
	};
	md.use(container, "demo", { render });
}
function parseCodeMapping(tokens, index, cb) {
	const codeMap = {};
	for (let i = index + 1; !(tokens[i].nesting === -1 && tokens[i].type === "container_demo_close"); ++i) {
		const token = tokens[i];
		if (token.type === "fence") {
			codeMap[normalizeAlias(token.info)] = token.content.trim();
			cb?.(token, tokens, i);
		}
	}
	return codeMap;
}
function getContainerMeta(info) {
	const [, type, raw] = (info.trim().slice(4).trim() || "").match(INFO_RE) || [];
	const { attrs } = resolveAttrs(raw);
	return {
		url: "",
		type: type || "normal",
		...attrs
	};
}
//#endregion
//#region src/node/demo/extendPage.ts
function extendsPageWithDemo(page) {
	const demoFiles = page.markdownEnv.demoFiles ?? [];
	page.deps.push(...demoFiles.filter(({ type }) => type === "markdown").map(({ path }) => path));
	(page.frontmatter.gitInclude ??= []).push(...demoFiles.filter(({ gitignore }) => !gitignore).map(({ path }) => path));
}
//#endregion
//#region src/node/demo/index.ts
/**
* Register demo plugin for markdown-it.
*
* 为 markdown-it 注册 demo 插件。
*
* This plugin enables demo syntax in markdown files, allowing users to
* create interactive code demonstrations with live preview.
*
* 该插件在 markdown 文件中启用 demo 语法，允许用户创建带有实时预览的交互式代码演示。
*
* @param app - VuePress app instance / VuePress 应用实例
* @param md - Markdown-it instance / Markdown-it 实例
*/
function demoPlugin(app, md) {
	createDemoRender();
	demoEmbed(app, md);
	demoContainer(app, md);
}
//#endregion
//#region src/node/embed/audio/reader.ts
const audioReader = (state, silent) => {
	const max = state.posMax;
	let start = state.pos;
	let href = "";
	if (state.src.slice(start, start + 13) !== "@[audioReader") return false;
	if (max - start < 17) return false;
	const labelStart = state.pos + 13;
	const labelEnd = state.md.helpers.parseLinkLabel(state, state.pos + 1, true);
	if (labelEnd < 0) return false;
	let pos = labelEnd + 1;
	if (pos < max && state.src.charCodeAt(pos) === 40) {
		pos++;
		const code = state.src.charCodeAt(pos);
		if (code === 10 || code === 32) return false;
		start = pos;
		const res = state.md.helpers.parseLinkDestination(state.src, pos, state.posMax);
		if (res.ok) {
			href = state.md.normalizeLink(res.str);
			if (state.md.validateLink(href)) pos = res.pos;
			else href = "";
		}
		if (pos >= max || state.src.charCodeAt(pos) !== 41) return false;
	} else return false;
	/* istanbul ignore else -- @preserve */
	if (!silent) {
		state.pos = labelStart;
		state.posMax = labelEnd;
		const info = state.src.slice(labelStart, labelEnd).trim();
		const { attrs } = resolveAttrs(info);
		const token = state.push("audio_reader", "AudioReader", 0);
		token.info = info;
		token.meta = {
			src: href,
			...attrs
		};
	}
	state.pos = pos + 1;
	state.posMax = max;
	return true;
};
const audioReaderPlugin = (md) => {
	md.renderer.rules.audio_reader = (tokens, idx) => {
		const meta = tokens[idx].meta;
		if (meta.startTime) meta.startTime = Number(meta.startTime);
		if (meta.endTime) meta.endTime = Number(meta.endTime);
		if (meta.volume) meta.volume = Number(meta.volume);
		return `<AudioReader${stringifyAttrs(meta)} />`;
	};
	md.inline.ruler.before("link", "audio-reader", audioReader);
};
//#endregion
//#region src/node/utils/nanoid.ts
/**
* Generate a nanoid with custom alphabet
*
* 使用自定义字符集生成 nanoid
*
* @param size - ID length / ID 长度
* @returns Nanoid string / Nanoid 字符串
*/
const nanoid = customAlphabet("abcdefghijklmnopqrstuvwxyz", 5);
//#endregion
//#region src/node/embed/caniuse.ts
const UNDERLINE_RE = /_+/g;
/**
* @example
* ```md
* @[caniuse](feature_name)
* ```
*/
const caniusePlugin = (md, { mode: defaultMode = "embed" } = {}) => {
	createEmbedRuleBlock(md, {
		type: "caniuse",
		syntaxPattern: /^@\[caniuse\s*(embed|image|baseline)?(?:\{([0-9,\-]*)\})?\]\(([^)]*)\)/,
		meta: ([, mode, versions = "", feature]) => ({
			feature,
			mode: mode || defaultMode,
			versions
		}),
		content: (meta, _c, env) => resolveCanIUse(meta, env)
	});
};
/**
* @deprecated use caniuse plugin
*
* 兼容旧语法
* @example
* ```md
* :::caniuse <feature_name>
* :::
* ```
*/
function legacyCaniuse(md, { mode = "embed" } = {}) {
	const modeMap = [
		"image",
		"embed",
		"baseline"
	];
	const isMode = (mode) => modeMap.includes(mode);
	mode = isMode(mode) ? mode : modeMap[0];
	createContainerPlugin(md, "caniuse", {
		before: (info, _t, _i, _o, env) => {
			const feature = info.split(/\s+/)[0];
			const versions = info.match(/\{(.*)\}/)?.[1] || "";
			return feature ? resolveCanIUse({
				feature,
				mode,
				versions
			}, env) : "";
		},
		after: () => ""
	});
}
function resolveCanIUse({ feature, mode, versions }, env) {
	if (!feature) return "";
	if (mode === "image") {
		logger$1.warn(`[caniuse] image mode is deprecated, use ${colors.cyan(`@[caniuse](${feature})`)} instead. (${colors.gray(env.filePathRelative || "")})`);
		const link = "https://caniuse.bitsofco.de/image/";
		const alt = `Data on support for the ${feature} feature across the major browsers from caniuse.com`;
		return `<p><picture>
      <source type="image/webp" srcset="${link}${feature}.webp">
      <source type="image/png" srcset="${link}${feature}.png">
      <img src="${link}${feature}.jpg" alt="${alt}" width="100%">
    </picture></p>`;
	}
	feature = feature.replace(UNDERLINE_RE, "_");
	const { past, future } = resolveVersions(versions);
	const meta = nanoid();
	return `<CanIUseViewer${stringifyAttrs({
		feature,
		meta,
		past,
		future,
		baseline: mode === "baseline"
	})} />`;
}
function resolveVersions(versions) {
	if (!versions) return {
		past: 2,
		future: 1
	};
	const list = versions.split(",").map((v) => Number(v.trim())).filter((v) => !Number.isNaN(v) && v >= -5 && v <= 3);
	list.push(0);
	const uniq = [...new Set(list)].sort((a, b) => b - a);
	return {
		future: uniq[0],
		past: Math.abs(uniq[uniq.length - 1])
	};
}
//#endregion
//#region src/node/embed/code/codepen.ts
const codepenPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "codepen",
		syntaxPattern: /^@\[codepen([^\]]*)\]\(([^)]*)\)/,
		meta: ([, info, source]) => {
			const { width, height, title, tab, preview, editable, theme } = resolveAttrs(info).attrs;
			const [user, slash] = source.split("/");
			return {
				title: title || "Code Pen",
				tab: tab || "result",
				width: width ? parseRect(width) : "100%",
				height: height ? parseRect(height) : "400px",
				user,
				slash,
				preview,
				editable,
				theme
			};
		},
		content: (meta) => `<CodePenViewer${stringifyAttrs(meta)} />`
	});
};
//#endregion
//#region src/node/embed/code/codeSandbox.ts
const codeSandboxPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "codesandbox",
		syntaxPattern: /^@\[codesandbox(?:\s+(embed|button))?([^\]]*)\]\(([^)]*)\)/,
		meta([, type, info, source]) {
			const { attrs } = resolveAttrs(info);
			const [profile, filepath = ""] = source.split("#");
			const [user, id] = profile.includes("/") ? profile.split("/") : ["", profile];
			return {
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : "500px",
				user,
				id,
				title: attrs.title ?? "",
				console: attrs.console ?? false,
				navbar: attrs.navbar ?? true,
				layout: attrs.layout ?? "",
				type: type || "embed",
				filepath
			};
		},
		content: (meta) => `<CodeSandboxViewer${stringifyAttrs(meta)} />`
	});
};
//#endregion
//#region src/node/embed/code/jsfiddle.ts
const jsfiddlePlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "jsfiddle",
		syntaxPattern: /^@\[jsfiddle([^\]]*)\]\(([^)]*)\)/,
		meta([, info, source]) {
			const { width, height, title, tab, theme } = resolveAttrs(info).attrs;
			return {
				width: width ? parseRect(width) : "100%",
				height: height ? parseRect(height) : "400px",
				source,
				title: title || "JS Fiddle",
				tab: tab?.replace(/\s+/g, "") || "js,css,html,result",
				theme
			};
		},
		content: (meta) => `<JSFiddleViewer${stringifyAttrs(meta)} />`
	});
};
//#endregion
//#region src/node/embed/code/replit.ts
const replitPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "replit",
		syntaxPattern: /^@\[replit([^\]]*)\]\(([^)]*)\)/,
		meta: ([, info = "", source = ""]) => {
			const { attrs } = resolveAttrs(info);
			return {
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : "450px",
				source: source.startsWith("@") ? source : `@${source}`,
				title: attrs.title,
				theme: attrs.theme || ""
			};
		},
		content: (meta) => `<ReplitViewer${stringifyAttrs(meta)} />`
	});
};
//#endregion
//#region src/node/embed/pdf.ts
const pdfPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "pdf",
		syntaxPattern: /^@\[pdf(?:\s+(\d+))?([^\]]*)\]\(([^)]*)\)/,
		meta([, page, info, src]) {
			const { attrs } = resolveAttrs(info);
			return {
				src,
				page: +page || 1,
				noToolbar: Boolean(attrs.noToolbar ?? false),
				zoom: +attrs.zoom || 50,
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : "",
				ratio: attrs.ratio ? parseRect(attrs.ratio) : "",
				title: path.basename(src || "")
			};
		},
		content: (meta) => `<PDFViewer${stringifyAttrs(meta)} />`
	});
};
//#endregion
//#region src/node/embed/qrcode.ts
const qrcodePlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "qrcode",
		syntaxPattern: /^@\[qrcode([^\]]*)\]\(([^)]*)\)/,
		meta([, info, text]) {
			const { attrs } = resolveAttrs(info);
			const { card, ...rest } = omit(attrs, ["text"]);
			return {
				text,
				...rest,
				mode: rest.mode || (card ? "card" : "img")
			};
		},
		content(meta) {
			return `<VPQRCode${stringifyAttrs(meta)} />`;
		}
	});
	createContainerSyntaxPlugin(md, "qrcode", (tokens, index) => {
		const { content, meta } = tokens[index];
		const { card, ...rest } = omit(meta, ["text"]);
		return `<VPQRCode ${stringifyAttrs({
			text: content?.trim(),
			...rest,
			mode: rest.mode || (card ? "card" : "img")
		})} />`;
	});
};
//#endregion
//#region src/node/embed/video/acfun.ts
const AC_FUN_LINK = "https://www.acfun.cn/player";
const acfunPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "acfun",
		name: "video_acfun",
		syntaxPattern: /^@\[acfun([^\]]*)\]\(([^)]*)\)/,
		meta([, info, id]) {
			const { attrs } = resolveAttrs(info);
			return {
				id,
				title: attrs.title || "AcFun",
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : void 0,
				ratio: attrs.ratio ?? "16:10"
			};
		},
		content(meta) {
			const { id, width, height, ratio, title } = meta;
			return `<VPVideoEmbed${stringifyAttrs({
				src: `${AC_FUN_LINK}/${id}`,
				width,
				height,
				ratio,
				title,
				type: "acfun"
			})} />`;
		}
	});
};
//#endregion
//#region src/node/embed/video/artPlayer.ts
const installed = {
	dashjs: isPackageExists("dashjs"),
	hlsjs: isPackageExists("hls.js"),
	mpegtsjs: isPackageExists("mpegts.js")
};
const SUPPORTED_VIDEO_TYPES = [
	"mp4",
	"mp3",
	"webm",
	"ogg",
	"mpd",
	"dash",
	"m3u8",
	"hls",
	"ts",
	"flv"
];
const artPlayerPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "artPlayer",
		name: "video_artPlayer",
		syntaxPattern: /^@\[artPlayer([^\]]*)\]\(([^)]*)\)/,
		meta([, info, source]) {
			const { attrs } = resolveAttrs(info);
			const url = source.trim();
			checkSupportType(attrs.type ?? url.split(".").pop());
			return {
				url,
				type: attrs.type,
				autoplay: attrs.autoplay ?? false,
				muted: attrs.muted ?? attrs.autoplay ?? false,
				autoMini: attrs.autoMini ?? false,
				loop: attrs.loop ?? false,
				volume: typeof attrs.volume !== "undefined" ? Number(attrs.volume) : .75,
				poster: attrs.poster,
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : void 0,
				ratio: attrs.ratio ? parseRect(`${attrs.ratio}`) : void 0
			};
		},
		content({ url, ...meta }) {
			meta.muted = meta.muted || meta.autoplay;
			return `<ArtPlayer src="${url}" fullscreen flip playback-rate aspect-ratio setting pip${stringifyAttrs(meta)}/>`;
		}
	});
};
function checkSupportType(type) {
	if (!type) return;
	/* istanbul ignore if -- @preserve */
	if (SUPPORTED_VIDEO_TYPES.includes(type)) {
		let name = "";
		switch (type.toLowerCase()) {
			case "m3u8":
			case "hls":
				name = !installed.hlsjs ? "hls.js" : "";
				break;
			case "flv":
			case "ts":
				name = !installed.mpegtsjs ? "mpegts.js" : "";
				break;
			case "mpd":
			case "dash":
				name = !installed.dashjs ? "dashjs" : "";
				break;
		}
		/* istanbul ignore if -- @preserve */
		if (name) logger$1.warn("artPlayer", `${colors.cyan(name)} is not installed, please install it via npm or yarn or pnpm`);
	} else
 /* istanbul ignore next -- @preserve */
	logger$1.warn("artPlayer", `unsupported video type: ${colors.cyan(type)}`);
}
//#endregion
//#region src/node/utils/timeToSeconds.ts
function timeToSeconds(time) {
	if (!time) return 0;
	if (Number.parseFloat(time) === Number(time)) return Number(time);
	const [s, m, h = 0] = time.split(/\s*:\s*/).reverse().map((n) => Number(n) || 0);
	return s + m * 60 + h * 3600;
}
//#endregion
//#region src/node/embed/video/bilibili.ts
const BILIBILI_LINK = "https://player.bilibili.com/player.html";
const bilibiliPlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "bilibili",
		name: "video_bilibili",
		syntaxPattern: /^@\[bilibili(?:\s+p(\d+))?([^\]]*)\]\(([^)]*)\)/,
		meta([, page, info, source]) {
			const { attrs } = resolveAttrs(info);
			const ids = source.trim().split(/\s+/);
			const bvid = ids.find((id) => id.startsWith("BV"));
			const [aid, cid] = ids.filter((id) => !id.startsWith("BV"));
			return {
				page: +page,
				bvid,
				aid,
				cid,
				autoplay: attrs.autoplay ?? false,
				time: timeToSeconds(attrs.time),
				title: attrs.title || "Bilibili",
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : void 0,
				ratio: attrs.ratio
			};
		},
		content(meta) {
			const params = new URLSearchParams();
			if (meta.bvid) params.set("bvid", meta.bvid);
			if (meta.aid) params.set("aid", meta.aid);
			if (meta.cid) params.set("cid", meta.cid);
			if (meta.page) params.set("p", meta.page.toString());
			if (meta.time) params.set("t", meta.time.toString());
			params.set("autoplay", meta.autoplay ? "1" : "0");
			params.set("high_quality", "1");
			const src = `${BILIBILI_LINK}?${params.toString()}`;
			const { width, height, ratio, title } = meta;
			return `<VPVideoEmbed${stringifyAttrs({
				src,
				width,
				height,
				ratio,
				title,
				type: "bilibili"
			})} />`;
		}
	});
};
//#endregion
//#region src/node/embed/video/youtube.ts
const YOUTUBE_LINK = "https://www.youtube.com/embed/";
const youtubePlugin = (md) => {
	createEmbedRuleBlock(md, {
		type: "youtube",
		name: "video_youtube",
		syntaxPattern: /^@\[youtube([^\]]*)\]\(([^)]*)\)/,
		meta([, info, id]) {
			const { attrs } = resolveAttrs(info);
			return {
				id,
				autoplay: attrs.autoplay ?? false,
				loop: attrs.loop ?? false,
				start: timeToSeconds(attrs.start),
				end: timeToSeconds(attrs.end),
				title: attrs.title || "YouTube",
				width: attrs.width ? parseRect(attrs.width) : "100%",
				height: attrs.height ? parseRect(attrs.height) : void 0,
				ratio: attrs.ratio
			};
		},
		content(meta) {
			const params = new URLSearchParams();
			if (meta.autoplay) params.set("autoplay", "1");
			if (meta.loop) params.set("loop", "1");
			if (meta.start) params.set("start", meta.start.toString());
			if (meta.end) params.set("end", meta.end.toString());
			const src = `${YOUTUBE_LINK}/${meta.id}?${params.toString()}`;
			const { width, height, ratio, title } = meta;
			return `<VPVideoEmbed${stringifyAttrs({
				src,
				width,
				height,
				ratio,
				title,
				type: "youtube"
			})} />`;
		}
	});
};
//#endregion
//#region src/node/embed/index.ts
/**
* Embed syntax plugin - Register all embed syntax plugins
*
* 嵌入语法插件 - 注册所有嵌入语法插件
*
* @param md - Markdown instance / Markdown 实例
* @param options - Plugin options / 插件选项
*/
function embedSyntaxPlugin(md, options) {
	if (options.caniuse) {
		const caniuse = options.caniuse === true ? {} : options.caniuse;
		md.use(caniusePlugin, caniuse);
		legacyCaniuse(md, caniuse);
	}
	if (options.pdf) md.use(pdfPlugin);
	if (options.acfun) md.use(acfunPlugin);
	if (options.bilibili) md.use(bilibiliPlugin);
	if (options.youtube) md.use(youtubePlugin);
	if (options.artPlayer) md.use(artPlayerPlugin);
	if (options.audioReader) md.use(audioReaderPlugin);
	if (options.codepen) md.use(codepenPlugin);
	if (options.replit) md.use(replitPlugin);
	if (options.codeSandbox) md.use(codeSandboxPlugin);
	if (options.jsfiddle) md.use(jsfiddlePlugin);
	if (options.qrcode) md.use(qrcodePlugin);
}
//#endregion
//#region src/node/enhance/docsTitle.ts
/**
* Regular expression for matching h1 heading
*
* 匹配 h1 标题的正则表达式
*/
const REG_HEADING = /^#\s*?([^#\s].*)?\n/;
/**
* Docs title plugin - Extract h1 title from markdown to frontmatter
*
* 文档标题插件 - 从 markdown 中提取 h1 标题到 frontmatter
*
* Adapts to theme's document page title by extracting h1 title from markdown to frontmatter
* and removing it from content to avoid duplicate display.
*
* 适配主题的文档页面标题，将 markdown 中的 h1 标题提取到 frontmatter 中，并将其删除，
* 以避免重复显示标题。
*
* @param md - Markdown instance / Markdown 实例
*/
function docsTitlePlugin(md) {
	const render = md.render.bind(md);
	md.render = (source, env) => {
		if (!env.filePathRelative) return render(source, env);
		let { matter, content } = parseSource(source.trim());
		let title = "";
		content = content.trim().replace(REG_HEADING, (_, match) => {
			title = match.trim();
			return "";
		});
		source = `${matter}\n${content}`;
		const result = render(source, env);
		if (title) {
			env.frontmatter ??= {};
			env.frontmatter.title ??= title;
		}
		return result;
	};
}
/**
* Parse markdown source to separate frontmatter and content
*
* 解析 markdown 源文件，分离 frontmatter 和内容
*
* @param source - Markdown source / Markdown 源文件
* @returns Object with matter and content / 包含 matter 和 content 的对象
*/
function parseSource(source) {
	const char = "---";
	if (!source.startsWith(char)) return {
		matter: "",
		content: source
	};
	else {
		const end = source.indexOf(`\n${char}`);
		const len = 4;
		return {
			matter: source.slice(0, end + len),
			content: source.slice(end + len)
		};
	}
}
//#endregion
//#region src/node/enhance/links.ts
/**
* Links plugin - Process internal and external links
*
* 链接插件 - 处理内部和外部链接
*
* @param md - Markdown instance / Markdown 实例
*/
function linksPlugin(md) {
	const externalAttrs = {
		target: "_blank",
		rel: "noopener noreferrer"
	};
	let hasOpenInternalLink = false;
	const internalTag = "VPLink";
	/**
	* Handle link open token
	*
	* 处理链接打开令牌
	*
	* @param tokens - Token array / 令牌数组
	* @param idx - Token index / 令牌索引
	* @param env - Markdown environment / Markdown 环境
	*/
	function handleLinkOpen(tokens, idx, env) {
		hasOpenInternalLink = false;
		const token = tokens[idx];
		const hrefIndex = token.attrIndex("href");
		/* istanbul ignore if -- @preserve */
		if (hrefIndex < 0) return;
		const hrefLink = token.attrs[hrefIndex][1];
		if (isLinkWithProtocol(hrefLink)) {
			Object.entries(externalAttrs).forEach(([key, val]) => {
				token.attrSet(key, val);
			});
			return;
		}
		if (hrefLink[0] === "#") return;
		hasOpenInternalLink = true;
		token.tag = internalTag;
		const matched = hrefLink.match(/^([^#?]*?(?:\/|\.md|\.html))([#?].*)?$/);
		if (matched) {
			const rawPath = matched[1];
			const { absolutePath, relativePath } = resolvePaths(rawPath, env.base || "/", env.filePathRelative ?? null);
			(env.links ??= []).push({
				raw: rawPath,
				absolute: absolutePath,
				relative: relativePath
			});
		}
	}
	md.renderer.rules.link_open = (tokens, idx, opts, env, self) => {
		handleLinkOpen(tokens, idx, env);
		return self.renderToken(tokens, idx, opts);
	};
	md.renderer.rules.link_close = (tokens, idx, opts, _env, self) => {
		if (hasOpenInternalLink) {
			hasOpenInternalLink = false;
			tokens[idx].tag = internalTag;
		}
		return self.renderToken(tokens, idx, opts);
	};
}
/**
* Resolve relative and absolute paths according to the `base` and `filePathRelative`
*
* 根据 `base` 和 `filePathRelative` 解析相对和绝对路径
*
* @param rawPath - Raw path / 原始路径
* @param base - Base URL / 基础 URL
* @param filePathRelative - Relative file path / 相对文件路径
* @returns Object with absolutePath and relativePath / 包含 absolutePath 和 relativePath 的对象
*/
function resolvePaths(rawPath, base, filePathRelative) {
	let absolutePath;
	let relativePath;
	if (rawPath.startsWith("/")) if (rawPath.endsWith(".md")) {
		absolutePath = path$2.join(base, rawPath);
		relativePath = removeLeadingSlash$1(rawPath);
	} else {
		absolutePath = rawPath;
		relativePath = path$2.relative(base, absolutePath);
	}
	else if (filePathRelative) {
		relativePath = path$2.join(path$2.dirname(encodeURI(filePathRelative)), rawPath);
		absolutePath = path$2.join(base, relativePath);
	} else {
		relativePath = rawPath.replace(/^(?:\.\/)?(.*)$/, "$1");
		absolutePath = null;
	}
	return {
		absolutePath,
		relativePath
	};
}
//#endregion
//#region src/node/icon/createIconRule.ts
function createIconRule([l1, l2, r1, r2], deprecated) {
	return (state, silent) => {
		let found = false;
		const max = state.posMax;
		const start = state.pos;
		if (state.src.charCodeAt(start) !== l1 || state.src.charCodeAt(start + 1) !== l2) return false;
		const next = state.src.charCodeAt(start + 2);
		if (next === 32 || next === 58) return false;
		/* istanbul ignore if -- @preserve */
		if (silent) return false;
		if (max - start < 5) return false;
		state.pos = start + 2;
		while (state.pos < max) {
			if (state.src.charCodeAt(state.pos) === r1 && state.src.charCodeAt(state.pos + 1) === r2) {
				found = true;
				break;
			}
			state.md.inline.skipToken(state);
		}
		if (!found || start + 2 === state.pos || state.src.charCodeAt(state.pos - 1) === 32) {
			state.pos = start;
			return false;
		}
		const info = state.src.slice(start + 2, state.pos);
		state.posMax = state.pos;
		state.pos = start + 2;
		const icon = state.push("icon", "i", 0);
		icon.markup = "::";
		icon.content = info;
		icon.meta = { deprecated };
		state.pos = state.posMax + 2;
		state.posMax = max;
		return true;
	};
}
//#endregion
//#region src/node/icon/resolveIcon.ts
const RE_SIZE = /(?<=\s|^)=(.+?)(?:\s|$)/;
const RE_COLOR = /(?<=\s|^)\/(.+?)(?:\s|$)/;
const RE_PROVIDER = /^(iconify|iconfont|fontawesome)\s+/;
const RE_EXTRA_KEY = /(?:^|-)\d-/g;
function resolveIcon(content, options) {
	let size = options.size;
	let color = options.color;
	let provider = options.provider || "iconify";
	content = content.replace(RE_PROVIDER, (_, p) => {
		provider = p;
		return "";
	}).replace(RE_SIZE, (_, s) => {
		size = s;
		return "";
	}).replace(RE_COLOR, (_, c) => {
		color = c;
		return "";
	}).trim();
	const index = content.indexOf(" ");
	const name = index === -1 ? content : content.slice(0, index);
	const extra = index === -1 ? "" : content.slice(index + 1);
	const props = {
		provider,
		size,
		color,
		name
	};
	if (!extra) return props;
	const { attrs } = resolveAttrs(extra);
	const info = [];
	const excludes = [];
	for (const key in attrs) if (attrs[key] === true) {
		excludes.push(key);
		info.push(kebabCase(key).replace(RE_EXTRA_KEY, (m) => `${m.slice(0, -1)}`));
	}
	return {
		...props,
		extra: info.join(" "),
		...omit(attrs, excludes)
	};
}
//#endregion
//#region src/node/icon/icon.ts
function iconRender(content, options) {
	return `<VPIcon${stringifyAttrs(resolveIcon(content, options))} />`;
}
const iconPlugin = (md, options = {}) => {
	/**
	* ::collect:icon_name =size /color::
	*/
	md.inline.ruler.before("link", "icon", createIconRule([
		58,
		58,
		58,
		58
	]));
	/**
	* :[collect:icon_name size/color]:
	* @deprecated
	*/
	md.inline.ruler.before("link", "icon_deprecated", createIconRule([
		58,
		91,
		93,
		58
	], true));
	md.renderer.rules.icon = (tokens, idx, _, env) => {
		const { content, meta } = tokens[idx];
		let icon = content;
		/* istanbul ignore if -- @preserve */
		if (meta.deprecated) {
			const [name, opt = ""] = content.split(" ");
			const [size, color] = opt.trim().split("/");
			icon = `${name}${size ? ` =${size}` : ""}${color ? ` /${color}` : ""}`;
			logger$1.warn("icon", `The icon syntax of \`${colors.yellow(`:[${content}]:`)}\` is deprecated, please use \`${colors.green(`::${icon}::`)}\` instead. (${colors.gray(env.filePathRelative || env.filePath)})`);
		}
		return iconRender(icon, options);
	};
};
//#endregion
//#region src/node/icon/prepareIcon.ts
function getFontAwesomeCDNLink(type) {
	return `https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/js/${type}.min.js`;
}
function prepareIcon(imports, options = {}) {
	const setupContent = [];
	const assets = [];
	if (options.provider === "iconfont") assets.push(...toArray(options.assets).map((asset) => normalizeAsset(asset)).filter(notNullish));
	else if (options.provider === "fontawesome") assets.push(...toArray(options.assets || "fontawesome").map((asset) => {
		if (asset === "fontawesome") return [
			"solid",
			"regular",
			"fontawesome"
		].map(getFontAwesomeCDNLink).map((asset) => normalizeAsset(asset, "fontawesome"));
		if (asset === "fontawesome-with-brands") return normalizeAsset(getFontAwesomeCDNLink("brands"), "fontawesome");
		return normalizeAsset(asset, "fontawesome");
	}).flat().filter(notNullish));
	let hasStyle = false;
	let hasScript = false;
	for (const asset of uniqWith(assets, (a, b) => a.link === b.link)) if (asset.type === "style") {
		hasStyle = true;
		setupContent.push(`useStyleTag('@import url("${asset.link}");')`);
	} else if (asset.type === "script") {
		hasScript = true;
		setupContent.push(asset.provide === "fontawesome" ? `useScriptTag("${asset.link}", () => {}, { attrs: { "data-auto-replace-svg": "nest" } })` : `useScriptTag("${asset.link}")`);
	}
	if (hasScript || hasStyle) {
		const exports = [];
		if (hasScript) exports.push("useScriptTag");
		if (hasStyle) exports.push("useStyleTag");
		imports.add(`import { ${exports.join(", ")} } from '@vueuse/core'`);
	}
	return setupContent.join("\n    ");
}
function normalizeAsset(asset, provide) {
	const link = normalizeLink(asset);
	if (asset.endsWith(".js")) return {
		type: "script",
		link,
		provide
	};
	if (asset.endsWith(".css")) return {
		type: "style",
		link,
		provide
	};
	logger$1.error("icon", `Can not recognize icon link: "${asset}"`);
	return null;
}
function normalizeLink(link) {
	if (isLinkHttp$1(link) || isLinkAbsolute(link)) return link;
	return `//${link}`;
}
//#endregion
//#region src/node/inline/abbr.ts
/**
* Abbreviation plugin - Enable abbreviation syntax
*
* 缩写词插件 - 启用缩写词语法
*
* Definition syntax: *[ABBREV]: Full description
* 定义语法：*[缩写]: 完整描述
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param globalAbbreviations - Global abbreviations preset / 全局缩写词预设
*/
const abbrPlugin = (md, globalAbbreviations = {}) => {
	const { arrayReplaceAt, escapeRE, lib } = md.utils;
	globalAbbreviations = objectMap(globalAbbreviations, (key, value) => [key.startsWith(":") ? key : `:${key}`, value]);
	const WORDING_REGEXP_TEXT = `${lib.ucmicro.P.source}|${lib.ucmicro.Z.source}|[${" \r\n$+<=>^`|~".split("").map(escapeRE).join("")}]`;
	/**
	* Abbreviation definition rule
	*
	* 缩写词定义规则
	*
	* @param state - State block / 状态块
	* @param startLine - Start line number / 开始行号
	* @param _endLine - End line number / 结束行号
	* @param silent - Silent mode / 静默模式
	* @returns Whether matched / 是否匹配
	*/
	const abbrDefinition = (state, startLine, _endLine, silent) => {
		let labelEnd = -1;
		let pos = state.bMarks[startLine] + state.tShift[startLine];
		const max = state.eMarks[startLine];
		if (pos + 2 >= max || state.src.charAt(pos++) !== "*" || state.src.charAt(pos++) !== "[") return false;
		const labelStart = pos;
		while (pos < max) {
			const ch = state.src.charAt(pos);
			if (ch === "[") return false;
			if (ch === "]") {
				labelEnd = pos;
				break;
			}
			if (ch === "\\") pos++;
			pos++;
		}
		if (labelEnd < 0 || state.src.charAt(labelEnd + 1) !== ":") return false;
		/* istanbul ignore if -- @preserve */
		if (silent) return true;
		const label = state.src.slice(labelStart, labelEnd).replace(/\\(.)/g, "$1");
		const title = state.src.slice(labelEnd + 2, max).trim();
		if (!label.length || !title.length) return false;
		(state.env.abbreviations ??= {})[`:${label}`] ??= title;
		state.line = startLine + 1;
		return true;
	};
	/**
	* Abbreviation replace rule
	*
	* 缩写词替换规则
	*
	* @param state - State core / 核心状态
	*/
	const abbrReplace = (state) => {
		const tokens = state.tokens;
		const { abbreviations: localAbbreviations } = state.env;
		if (!localAbbreviations && isEmptyObject(globalAbbreviations)) return;
		const abbreviations = {
			...globalAbbreviations,
			...localAbbreviations
		};
		const abbreviationsRegExpText = Object.keys(abbreviations).map((x) => x.substring(1)).sort((a, b) => b.length - a.length).map(escapeRE).join("|");
		const regexpSimple = new RegExp(`(?:${abbreviationsRegExpText})`);
		const regExp = new RegExp(`(^|${WORDING_REGEXP_TEXT})(${abbreviationsRegExpText})($|${WORDING_REGEXP_TEXT})`, "g");
		for (const token of tokens) {
			if (token.type !== "inline") continue;
			let children = token.children;
			for (let index = children.length - 1; index >= 0; index--) {
				const currentToken = children[index];
				if (currentToken.type !== "text") continue;
				const text = currentToken.content;
				regExp.lastIndex = 0;
				if (!regexpSimple.test(text)) continue;
				const nodes = [];
				let match;
				let pos = 0;
				while (match = regExp.exec(text)) {
					const [, before, word, after] = match;
					if (match.index > 0 || before.length > 0) {
						const token = new state.Token("text", "", 0);
						token.content = text.slice(pos, match.index + before.length);
						nodes.push(token);
					}
					const abbrToken = new state.Token("abbreviation", "Abbreviation", 0);
					abbrToken.content = word;
					abbrToken.info = abbreviations[`:${word}`];
					nodes.push(abbrToken);
					regExp.lastIndex -= after.length;
					pos = regExp.lastIndex;
				}
				/* istanbul ignore if -- @preserve */
				if (!nodes.length) continue;
				if (pos < text.length) {
					const token = new state.Token("text", "", 0);
					token.content = text.slice(pos);
					nodes.push(token);
				}
				token.children = children = arrayReplaceAt(children, index, nodes);
			}
		}
	};
	md.block.ruler.before("reference", "abbr_definition", abbrDefinition, { alt: ["paragraph", "reference"] });
	md.core.ruler.after("linkify", "abbr_replace", abbrReplace);
	md.renderer.rules.abbreviation = (tokens, idx, _, env) => {
		const { content, info } = tokens[idx];
		const rendered = md.renderInline(info, cleanMarkdownEnv(env));
		return `<Abbreviation aria-label="${rendered.replace(/<[^>]*>/g, "")}">${content}${info ? `<template #tooltip>${rendered}</template>` : ""}</Abbreviation>`;
	};
};
//#endregion
//#region src/node/inline/annotation.ts
/**
* Annotation definition rule
*
* 注释定义规则
*
* @param state - State block / 状态块
* @param startLine - Start line number / 开始行号
* @param endLine - End line number / 结束行号
* @param silent - Silent mode / 静默模式
* @returns Whether matched / 是否匹配
*/
const annotationDef = (state, startLine, endLine, silent) => {
	const start = state.bMarks[startLine] + state.tShift[startLine];
	const max = state.eMarks[startLine];
	if (start + 4 > max || state.src.charAt(start) !== "[" || state.src.charAt(start + 1) !== "+") return false;
	let pos = start + 2;
	while (pos < max) {
		if (state.src.charAt(pos) === " ") return false;
		if (state.src.charAt(pos) === "]") break;
		pos++;
	}
	if (pos === start + 2 || pos + 1 >= max || state.src.charAt(++pos) !== ":") return false;
	/* istanbul ignore if -- @preserve */
	if (silent) return true;
	pos++;
	const data = state.env.annotations ??= {};
	const label = state.src.slice(start + 2, pos - 2);
	let annotation = state.src.slice(pos, max).trim();
	let nextLine = startLine + 1;
	while (nextLine < endLine) {
		const nextStart = state.bMarks[nextLine] + state.tShift[nextLine];
		const nextMax = state.eMarks[nextLine];
		const source = state.src.slice(nextStart, nextMax).trim();
		if (state.sCount[nextLine] < state.blkIndent + 2 && source !== "") break;
		annotation += `\n${source}`;
		nextLine++;
	}
	(data[`:${label}`] ??= {
		sources: [],
		rendered: []
	}).sources.push(annotation);
	state.line = nextLine;
	return true;
};
/**
* Annotation reference rule
*
* 注释引用规则
*
* @param state - State inline / 行内状态
* @param silent - Silent mode / 静默模式
* @returns Whether matched / 是否匹配
*/
const annotationRef = (state, silent) => {
	const start = state.pos;
	const max = state.posMax;
	if (start + 3 > max || typeof state.env.annotations === "undefined" || state.src.charAt(start) !== "[" || state.src.charAt(start + 1) !== "+") return false;
	let pos = start + 2;
	while (pos < max) {
		if (state.src.charAt(pos) === " " || state.src.charAt(pos) === "\n") return false;
		if (state.src.charAt(pos) === "]") break;
		pos++;
	}
	if (pos === start + 2 || pos >= max) return false;
	pos++;
	const label = state.src.slice(start + 2, pos - 1);
	if ((state.env.annotations?.[`:${label}`]?.sources ?? []).length === 0) return false;
	/* istanbul ignore if -- @preserve */
	if (!silent) {
		const refToken = state.push("annotation_ref", "", 0);
		refToken.meta = { label };
	}
	state.pos = pos;
	state.posMax = max;
	return true;
};
/**
* Annotation plugin - Enable annotation syntax
*
* 注释插件 - 启用注释语法
*
* Definition syntax: [+label]: annotation content
* Reference syntax: [+label]
*
* 定义语法：[+label]: 注释内容
* 引用语法：[+label]
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param globalAnnotations - Global annotations preset / 全局注释预设
*/
const annotationPlugin = (md, globalAnnotations = {}) => {
	const annotations = objectMap(globalAnnotations, (key, value) => {
		return [key.startsWith(":") ? key : `:${key}`, {
			sources: toArray(value),
			rendered: []
		}];
	});
	md.renderer.rules.annotation_ref = (tokens, idx, _, env) => {
		const label = tokens[idx].meta.label;
		/* istanbul ignore next -- @preserve */
		const data = env.annotations[`:${label}`] || annotations[`:${label}`];
		return `<Annotation label="${label}" :total="${data.sources.length}">${data.sources.map((source, i) => {
			return `<template #item-${i}>${data.rendered[i] ??= md.render(source, cleanMarkdownEnv(env, ["references"]))}</template>`;
		}).join("")}</Annotation>`;
	};
	md.inline.ruler.before("image", "annotation_ref", annotationRef);
	md.block.ruler.before("reference", "annotation", annotationDef, { alt: ["paragraph", "reference"] });
};
//#endregion
//#region src/node/inline/env-preset.ts
/**
* Environment preset plugin - Inject preset references to markdown env
*
* 环境预设插件 - 向 Markdown 环境注入预设引用
*
* @param md - Markdown-it instance / Markdown-it 实例
* @param env - Environment preset configuration / 环境预设配置
*/
const envPresetPlugin = (md, env = {}) => {
	if (isEmptyObject(env)) return;
	const references = objectMap(env.references || {}, (key, value) => {
		return [md.utils.normalizeReference(key), isString(value) ? { href: value } : value];
	});
	const defaultRender = md.render;
	md.render = (source, env) => {
		env ??= {};
		env.references = {
			...references,
			...env.references
		};
		return defaultRender(source, env);
	};
	const defaultRenderInline = md.renderInline;
	md.renderInline = (source, env) => {
		env ??= {};
		env.references = {
			...references,
			...env.references
		};
		return defaultRenderInline(source, env);
	};
};
//#endregion
//#region src/node/inline/plot.ts
/**
* Plot inline rule definition
*
* 黑幕行内规则定义
*
* @param state - Markdown-it state / Markdown-it 状态
* @param silent - Silent mode / 静默模式
* @returns Whether matched / 是否匹配
*/
const plotDef = (state, silent) => {
	let found = false;
	const max = state.posMax;
	const start = state.pos;
	if (state.src.charCodeAt(start) !== 33 || state.src.charCodeAt(start + 1) !== 33) return false;
	const next = state.src.charCodeAt(start + 2);
	if (next === 32 || next === 33) return false;
	/* istanbul ignore if -- @preserve */
	if (silent) return false;
	if (max - start < 5) return false;
	state.pos = start + 2;
	while (state.pos < max) {
		if (state.src.charCodeAt(state.pos) === 33 && state.src.charCodeAt(state.pos + 1) === 33) {
			found = true;
			break;
		}
		state.md.inline.skipToken(state);
	}
	if (!found || start + 2 === state.pos || state.src.charCodeAt(state.pos - 1) === 32) {
		state.pos = start;
		return false;
	}
	const content = state.src.slice(start + 2, state.pos);
	state.posMax = state.pos;
	state.pos = start + 2;
	const openToken = state.push("plot_inline_open", "Plot", 1);
	openToken.markup = "!!";
	openToken.content = content;
	const contentToken = state.push("text", "", 0);
	contentToken.content = content;
	const closeToken = state.push("plot_inline_close", "Plot", -1);
	closeToken.markup = "!!";
	state.pos = state.posMax + 2;
	state.posMax = max;
	return true;
};
/**
* Plot plugin - Hide text with spoiler effect
*
* 黑幕插件 - 使用黑幕效果隐藏文本
*
* Syntax: !!hidden text!!
* 语法：!!隐藏文本!!
*
* @param md - Markdown-it instance / Markdown-it 实例
*/
const plotPlugin = (md) => {
	md.inline.ruler.before("emphasis", "plot", plotDef);
};
//#endregion
//#region src/node/inline/index.ts
/**
* Inline syntax plugin - Register all inline markdown syntax plugins
*
* 行内语法插件 - 注册所有行内 Markdown 语法插件
*
* @param md - Markdown instance / Markdown 实例
* @param options - Plugin options / 插件选项
*/
function inlineSyntaxPlugin(md, options) {
	md.use(cjsFriendly);
	md.use(attrs);
	md.use(mark);
	md.use(sub);
	md.use(sup);
	md.use(footnote);
	md.use(tasklist);
	const env = options.env || {};
	envPresetPlugin(md, env);
	if (options.annotation)
 /**
	* xxx [+foo] xxx
	*
	* [+foo]: xxx
	*/
	md.use(annotationPlugin, {
		...env.annotations,
		...isPlainObject$1(options.annotation) ? options.annotation : {}
	});
	if (options.abbr)
 /**
	* a HTML element
	*
	* [HTML]: A HTML element description
	*/
	md.use(abbrPlugin, {
		...env.abbreviations,
		...isPlainObject$1(options.abbr) ? options.abbr : {}
	});
	if (options.plot === true || isPlainObject$1(options.plot)) md.use(plotPlugin);
}
//#endregion
//#region src/node/locales/index.ts
/**
* Default locale options for the plugin.
*
* 插件的默认本地化选项。
*
* This constant defines the default locale configurations for all supported languages.
* Each locale entry maps language codes to their respective locale data.
*
* 该常量定义了所有支持语言的默认本地化配置。
* 每个本地化条目将语言代码映射到其相应的本地化数据。
*/
const LOCALE_OPTIONS = [
	[["en", "en-US"], {
		common: {
			copy: "Copy",
			copied: "Copied"
		},
		encrypt: {
			hint: "The content is encrypted, please unlock to view.",
			placeholder: "Enter password",
			incPwd: "Incorrect password",
			noContent: "Unlocked, but content failed to load, please try again later.",
			warningTitle: "🚨 Security Warning:",
			warningText: "Your connection is not encrypted with HTTPS, posing a risk of content leakage and preventing access to encrypted content."
		}
	}],
	[[
		"zh",
		"zh-CN",
		"zh-Hans",
		"zh-Hant"
	], {
		common: {
			copy: "复制",
			copied: "已复制"
		},
		encrypt: {
			hint: "内容已加密，请解锁后查看。",
			placeholder: "输入密码",
			incPwd: "密码错误",
			noContent: "已解锁，但内容加载失败，请稍后再试。",
			warningTitle: "🚨 安全警告：",
			warningText: "您的连接未使用HTTPS加密，存在内容泄露风险，无法访问加密内容。"
		}
	}],
	[["zh-TW"], {
		common: {
			copy: "複製",
			copied: "已複製"
		},
		encrypt: {
			hint: "內容已加密，請解鎖後查看。",
			placeholder: "輸入密碼",
			incPwd: "密碼錯誤",
			noContent: "已解鎖，但內容載入失敗，請稍後再試。",
			warningTitle: "🚨 安全警告：",
			warningText: "您的連線未使用 HTTPS 加密，可能導致內容洩露風險，無法存取加密內容。"
		}
	}],
	[["de", "de-DE"], {
		common: {
			copy: "Kopieren",
			copied: "Kopiert"
		},
		encrypt: {
			hint: "Der Inhalt ist verschlüsselt, bitte entsperren Sie ihn, um ihn anzuzeigen.",
			placeholder: "Passwort eingeben",
			incPwd: "Falsches Passwort",
			noContent: "Entsperrt, aber der Inhalt konnte nicht geladen werden. Bitte versuchen Sie es später erneut.",
			warningTitle: "🚨 Sicherheitswarnung:",
			warningText: "Ihre Verbindung ist nicht mit HTTPS verschlüsselt, was ein Risiko für Inhaltslecks darstellt und den Zugriff auf verschlüsselte Inhalte verhindert."
		}
	}],
	[["fr", "fr-FR"], {
		common: {
			copy: "Copier",
			copied: "Copié"
		},
		encrypt: {
			hint: "Le contenu est chiffré, veuillez déverrouiller pour afficher.",
			placeholder: "Entrez le mot de passe",
			incPwd: "Mot de passe incorrect",
			noContent: "Déverrouillé, mais le contenu n'a pas pu être chargé, veuillez réessayer plus tard.",
			warningTitle: "🚨 Avertissement de sécurité :",
			warningText: "Votre connexion n'est pas chiffrée avec HTTPS, ce qui présente un risque de fuite de contenu et empêche l'accès au contenu chiffré."
		}
	}],
	[["ru", "ru-RU"], {
		common: {
			copy: "Копировать",
			copied: "Скопировано"
		},
		encrypt: {
			hint: "Контент зашифрован, разблокируйте для просмотра.",
			placeholder: "Введите пароль",
			incPwd: "Неверный пароль",
			noContent: "Разблокировано, но не удалось загрузить содержимое. Повторите попытку позже.",
			warningTitle: "🚨 Предупреждение безопасности:",
			warningText: "Ваше соединение не защищено HTTPS, что создает риск утечки данных и блокирует доступ к зашифрованному контенту."
		}
	}],
	[["ja", "ja-JP"], {
		common: {
			copy: "コピー",
			copied: "コピー済み"
		},
		encrypt: {
			hint: "コンテンツは暗号化されています。閲覧するにはロックを解除してください。",
			placeholder: "パスワードを入力",
			incPwd: "パスワードが間違っています",
			noContent: "ロックは解除されましたが、コンテンツの読み込みに失敗しました。後ほど再度お試しください。",
			warningTitle: "🚨 セキュリティ警告:",
			warningText: "接続がHTTPSで暗号化されていないため、コンテンツの漏洩リスクがあり、暗号化されたコンテンツへのアクセスができません。"
		}
	}],
	[["ko", "ko-KR"], {
		common: {
			copy: "복사",
			copied: "복사됨"
		},
		encrypt: {
			hint: "내용이 암호화되어 있습니다. 잠금 해제 후 확인하세요.",
			placeholder: "비밀번호 입력",
			incPwd: "잘못된 비밀번호",
			noContent: "잠금이 해제되었지만 내용을 불러오지 못했습니다. 나중에 다시 시도해 주세요.",
			warningTitle: "🚨 보안 경고:",
			warningText: "연결이 HTTPS로 암호화되지 않아 내용 유출 위험이 있으며, 암호화된 콘텐츠에 접근할 수 없습니다."
		}
	}]
];
//#endregion
//#region src/node/prepareConfigFile.ts
const { url: filepath } = import.meta;
const __dirname = getDirname(filepath);
/**
* Client folder path constant.
*
* 客户端文件夹路径常量。
*/
const CLIENT_FOLDER = ensureEndingSlash(path.resolve(__dirname, "../client"));
/**
* Prepare client configuration file for the plugin.
*
* 为插件准备客户端配置文件。
*
* This function dynamically generates the client config file based on the plugin options.
* It imports and registers Vue components conditionally based on which features are enabled.
*
* 该函数根据插件选项动态生成客户端配置文件。
* 它根据启用的功能有条件地导入和注册 Vue 组件。
*
* @param app - VuePress app instance / VuePress 应用实例
* @param options - Plugin options / 插件配置选项
* @returns Path to the generated config file / 生成的配置文件路径
*
* @example
* ```ts
* const configPath = await prepareConfigFile(app, {
*   pdf: true,
*   tabs: true,
* })
* ```
*/
async function prepareConfigFile(app, options) {
	const imports = /* @__PURE__ */ new Set();
	const enhances = /* @__PURE__ */ new Set();
	imports.add(`import VPCopyButton from '${CLIENT_FOLDER}components/VPCopyButton.vue'`);
	enhances.add(`app.component('VPCopyButton', VPCopyButton)`);
	imports.add(`import Tabs from '${CLIENT_FOLDER}components/Tabs.vue'`);
	enhances.add(`app.component('Tabs', Tabs)`);
	imports.add(`import CodeTabs from '${CLIENT_FOLDER}components/CodeTabs.vue'`);
	enhances.add(`app.component('CodeTabs', CodeTabs)`);
	if (options.encrypt) {
		imports.add(`import VPEncryptSnippet from '${CLIENT_FOLDER}components/VPEncryptSnippet.vue'`);
		enhances.add(`app.component('VPEncryptSnippet', VPEncryptSnippet)`);
	}
	if (options.pdf) {
		imports.add(`import PDFViewer from '${CLIENT_FOLDER}components/PDFViewer.vue'`);
		enhances.add(`app.component('PDFViewer', PDFViewer)`);
	}
	if (options.acfun || options.bilibili || options.youtube) {
		imports.add(`import VPVideoEmbed from '${CLIENT_FOLDER}components/VPVideoEmbed.vue'`);
		enhances.add(`app.component('VPVideoEmbed', VPVideoEmbed)`);
	}
	if (options.codepen) {
		imports.add(`import CodePen from '${CLIENT_FOLDER}components/CodePen.vue'`);
		enhances.add(`app.component('CodePenViewer', CodePen)`);
	}
	if (options.jsfiddle) {
		imports.add(`import JSFiddle from '${CLIENT_FOLDER}components/JsFiddle.vue'`);
		enhances.add(`app.component('JSFiddleViewer', JSFiddle)`);
	}
	if (options.replit) {
		imports.add(`import Replit from '${CLIENT_FOLDER}components/Replit.vue'`);
		enhances.add(`app.component('ReplitViewer', Replit)`);
	}
	if (options.codeSandbox) {
		imports.add(`import CodeSandbox from '${CLIENT_FOLDER}components/CodeSandbox.vue'`);
		enhances.add(`app.component('CodeSandboxViewer', CodeSandbox)`);
	}
	if (options.plot) {
		imports.add(`import Plot from '${CLIENT_FOLDER}components/Plot.vue'`);
		enhances.add(`app.component('Plot', Plot)`);
	}
	if (options.repl) {
		imports.add(`import CodeRepl from '${CLIENT_FOLDER}components/CodeRepl.vue'`);
		enhances.add(`app.component('CodeRepl', CodeRepl)`);
	}
	if (options.caniuse) {
		imports.add(`import CanIUse from '${CLIENT_FOLDER}components/CanIUse.vue'`);
		enhances.add(`app.component('CanIUseViewer', CanIUse)`);
	}
	if (options.fileTree || options.codeTree) {
		imports.add(`import FileTreeNode from '${CLIENT_FOLDER}components/FileTreeNode.vue'`);
		enhances.add(`app.component('FileTreeNode', FileTreeNode)`);
	}
	if (options.codeTree) {
		imports.add(`import VPCodeTree from '${CLIENT_FOLDER}components/VPCodeTree.vue'`);
		enhances.add(`app.component('VPCodeTree', VPCodeTree)`);
	}
	if (options.artPlayer) {
		imports.add(`import ArtPlayer from '${CLIENT_FOLDER}components/ArtPlayer.vue'`);
		enhances.add(`app.component('ArtPlayer', ArtPlayer)`);
	}
	if (options.audioReader) {
		imports.add(`import AudioReader from '${CLIENT_FOLDER}components/AudioReader.vue'`);
		enhances.add(`app.component('AudioReader', AudioReader)`);
	}
	if (options.demo) {
		imports.add(`import VPDemoBasic from '${CLIENT_FOLDER}components/VPDemoBasic.vue'`);
		imports.add(`import VPDemoNormal from '${CLIENT_FOLDER}components/VPDemoNormal.vue'`);
		enhances.add(`app.component('VPDemoBasic', VPDemoBasic)`);
		enhances.add(`app.component('VPDemoNormal', VPDemoNormal)`);
	}
	if (options.annotation) {
		imports.add(`import Annotation from '${CLIENT_FOLDER}components/Annotation.vue'`);
		enhances.add(`app.component('Annotation', Annotation)`);
	}
	if (options.abbr) {
		imports.add(`import Abbreviation from '${CLIENT_FOLDER}components/Abbreviation.vue'`);
		enhances.add(`app.component('Abbreviation', Abbreviation)`);
	}
	if (options.timeline) {
		imports.add(`import VPTimeline from '${CLIENT_FOLDER}components/VPTimeline.vue'`);
		imports.add(`import VPTimelineItem from '${CLIENT_FOLDER}components/VPTimelineItem.vue'`);
		enhances.add(`app.component('VPTimeline', VPTimeline)`);
		enhances.add(`app.component('VPTimelineItem', VPTimelineItem)`);
	}
	if (options.collapse) {
		imports.add(`import VPCollapse from '${CLIENT_FOLDER}components/VPCollapse.vue'`);
		imports.add(`import VPCollapseItem from '${CLIENT_FOLDER}components/VPCollapseItem.vue'`);
		enhances.add(`app.component('VPCollapse', VPCollapse)`);
		enhances.add(`app.component('VPCollapseItem', VPCollapseItem)`);
	}
	if (options.chat) imports.add(`import '${CLIENT_FOLDER}styles/chat.css'`);
	if (options.field) {
		imports.add(`import VPField from '${CLIENT_FOLDER}components/VPField.vue'`);
		enhances.add(`app.component('VPField', VPField)`);
	}
	if (options.table) {
		imports.add(`import VPTable from '${CLIENT_FOLDER}components/VPTable.vue'`);
		enhances.add(`app.component('VPTable', VPTable)`);
	}
	if (options.qrcode) {
		imports.add(`import VPQRCode from '${CLIENT_FOLDER}components/VPQRCode.vue'`);
		enhances.add(`app.component('VPQRCode', VPQRCode)`);
	}
	const setupIcon = prepareIcon(imports, options.icon);
	const setupStmts = [];
	const iconSetup = setupIcon.trim();
	if (iconSetup) setupStmts.push(iconSetup);
	imports.add(`import { setupMarkHighlight } from '${CLIENT_FOLDER}composables/mark.js'`);
	setupStmts.push(`setupMarkHighlight(${JSON.stringify(options.mark === "lazy" ? "lazy" : "eager")})`);
	const setupContent = setupStmts.length ? `    ${setupStmts.join("\n    ")}\n` : "";
	return app.writeTemp("md-power/config.js", `\
import { defineClientConfig } from 'vuepress/client'
${Array.from(imports.values()).join("\n")}

import '${CLIENT_FOLDER}styles/index.css'

export default defineClientConfig({
  enhance({ router, app }) {
${Array.from(enhances.values()).map((item) => `    ${item}`).join("\n")}
  },
  setup() {
    ${setupContent}
  }
})
`);
}
//#endregion
//#region src/node/provideData.ts
/**
* Provide data to be injected into the client-side application.
*
* 提供要注入到客户端应用的数据。
*
* This function creates a record of data that will be defined as global constants
* in the client bundle, allowing the client to access plugin configuration and
* runtime information.
*
* 该函数创建一个数据记录，将作为全局常量定义在客户端包中，
* 允许客户端访问插件配置和运行时信息。
*
* @param options - Plugin options / 插件配置选项
* @param locales - Locale configuration / 本地化配置
* @returns Record of data to be defined / 要定义的数据记录
*
* @example
* ```ts
* const data = provideData(options, locales)
* // Returns:
* // {
* //   __MD_POWER_INJECT_OPTIONS__: { plot: true, pdf: true },
* //   __MD_POWER_DASHJS_INSTALLED__: true,
* //   __MD_POWER_HLSJS_INSTALLED__: false,
* //   ...
* // }
* ```
*/
function provideData(options, locales) {
	const markdownOptions = {
		plot: options.plot,
		pdf: options.pdf
	};
	const icon = options.icon ?? { provider: "iconify" };
	return {
		__MD_POWER_INJECT_OPTIONS__: markdownOptions,
		__MD_POWER_DASHJS_INSTALLED__: isPackageExists("dashjs"),
		__MD_POWER_HLSJS_INSTALLED__: isPackageExists("hls.js"),
		__MD_POWER_MPEGTSJS_INSTALLED__: isPackageExists("mpegts.js"),
		__MD_POWER_ICON_PROVIDER__: icon.provider || "iconify",
		__MD_POWER_ICON_PREFIX__: icon.prefix || "",
		__MD_POWER_ENCRYPT_LOCALES__: options.encrypt ? findLocales(locales, "encrypt") : {}
	};
}
//#endregion
//#region src/node/plugin.ts
/**
* Create markdown power plugin for VuePress.
*
* 为 VuePress 创建 markdown 增强插件。
*
* This plugin provides various markdown enhancements including:
* - Custom containers (tabs, collapse, timeline, etc.)
* - Code blocks enhancements (code tabs, file tree, demo)
* - Embed syntax (video, PDF, code playground)
* - Inline syntax (mark, subscript, superscript, footnote)
* - Icon support
*
* 该插件提供多种 markdown 增强功能，包括：
* - 自定义容器（标签页、折叠面板、时间线等）
* - 代码块增强（代码标签页、文件树、演示）
* - 嵌入语法（视频、PDF、代码游乐场）
* - 行内语法（标记、下标、上标、脚注）
* - 图标支持
*
* @param options - Plugin options / 插件配置选项
* @returns VuePress plugin instance / VuePress 插件实例
*
* @example
* ```ts
* // Basic usage
* markdownPowerPlugin()
*
* // With options
* markdownPowerPlugin({
*   tabs: true,
*   collapse: true,
*   pdf: true,
* })
* ```
*/
function markdownPowerPlugin(options = {}) {
	return (app) => {
		const locales = getFullLocaleConfig({
			app,
			name: "vuepress-plugin-md-power",
			default: LOCALE_OPTIONS,
			config: options.locales
		});
		return {
			name: "vuepress-plugin-md-power",
			clientConfigFile: (app) => prepareConfigFile(app, options),
			define: provideData(options, locales),
			alias: (_, isServer) => {
				if (!isServer) return { ...options.encrypt ? { "/^vue$/": "vue/dist/vue.esm-bundler.js" } : void 0 };
				return {};
			},
			extendsBundlerOptions(bundlerOptions, app) {
				if (options.repl) {
					addViteOptimizeDepsInclude(bundlerOptions, app, [
						"shiki/core",
						"shiki/wasm",
						"shiki/engine/oniguruma"
					]);
					if (options.repl.python) addViteOptimizeDepsInclude(bundlerOptions, app, ["pyodide"]);
				}
				if (options.artPlayer) addViteOptimizeDepsInclude(bundlerOptions, app, [
					"artplayer",
					"dashjs",
					"hls.js",
					"mpegts.js/dist/mpegts.js"
				]);
				if (options.qrcode) addViteOptimizeDepsInclude(bundlerOptions, app, ["qrcode"]);
			},
			extendsMarkdown: async (md, app) => {
				linksPlugin(md);
				docsTitlePlugin(md);
				embedSyntaxPlugin(md, options);
				inlineSyntaxPlugin(md, options);
				iconPlugin(md, options.icon ?? (isPlainObject$1(options.icons) ? options.icons : {}));
				if (options.demo) demoPlugin(app, md);
				await containerPlugin(app, md, options, locales);
				await imageSizePlugin(app, md, options.imageSize);
			},
			onPrepared: async () => {
				if (options.demo) await waitDemoRender();
			},
			onWatched(app, watchers) {
				if (options.demo) demoWatcher(app, watchers);
			},
			extendsPage: (page) => {
				if (options.demo) extendsPageWithDemo(page);
				if (options.codeTree) extendsPageWithCodeTree(page);
			}
		};
	};
}
//#endregion
export { createCodeTabIconGetter, getImageOriginalSize, markdownPowerPlugin, resolveImagePath };
