// vuepress-plugin-md-power@1.0.0-rc.193/lib/client/options.js
//#region src/client/options.ts
/**
* Plugin options injected at build time.
*
* 构建时注入的插件选项。
*/
const pluginOptions = __MD_POWER_INJECT_OPTIONS__;
/**
* Package installation status for video streaming libraries.
*
* 视频流媒体库的安装状态。
*/
const installed = {
	dashjs: __MD_POWER_DASHJS_INSTALLED__,
	hlsjs: __MD_POWER_HLSJS_INSTALLED__,
	mpegtsjs: __MD_POWER_MPEGTSJS_INSTALLED__
};
/**
* Supported video types for ArtPlayer.
*
* ArtPlayer 支持的视频类型。
*/
const ART_PLAYER_SUPPORTED_VIDEO_TYPES = [
	"mp4",
	"mp3",
	"webm",
	"ogg"
];
if (installed.dashjs) ART_PLAYER_SUPPORTED_VIDEO_TYPES.push("mpd", "dash");
if (installed.hlsjs) ART_PLAYER_SUPPORTED_VIDEO_TYPES.push("m3u8", "hls");
if (installed.mpegtsjs) ART_PLAYER_SUPPORTED_VIDEO_TYPES.push("ts", "flv");
/**
* Injection key for timeline component communication.
*
* 时间线组件通信的注入键。
*/
const INJECT_TIMELINE_KEY = Symbol(__VUEPRESS_DEV__ ? "timeline" : "");
/**
* Injection key for collapse component communication.
*
* 折叠面板组件通信的注入键。
*/
const INJECT_COLLAPSE_KEY = Symbol(__VUEPRESS_DEV__ ? "collapse" : "");
/**
* Encrypt snippet locale data.
*
* 加密片段本地化数据。
*/
const ENCRYPT_LOCALES = __MD_POWER_ENCRYPT_LOCALES__;
//#endregion
export { ART_PLAYER_SUPPORTED_VIDEO_TYPES, ENCRYPT_LOCALES, INJECT_COLLAPSE_KEY, INJECT_TIMELINE_KEY, installed, pluginOptions };
