// vuepress-plugin-md-power@1.0.0-rc.193/lib/client/composables/codeRepl.js
import { rustExecute } from "./rustRepl.js";
import { onMounted, ref } from "vue";
import { http } from "../utils/http.js";
import { sleep } from "../utils/sleep.js";
//#region src/client/composables/codeRepl.ts
/**
* CSS selectors for nodes to ignore when extracting code.
*
* 提取代码时要忽略的节点的 CSS 选择器。
*/
const ignoredNodes = [".diff.remove", ".vp-copy-ignore"];
/**
* Regular expression for matching language class.
*
* 匹配语言类的正则表达式。
*/
const RE_LANGUAGE = /language-(\w+)/;
/**
* API endpoints for code execution backends.
*
* 代码执行后端的 API 端点。
*/
const api = {
	go: "https://api.pengzhanbo.cn/repl/golang/run",
	kotlin: "https://api.pengzhanbo.cn/repl/kotlin/run"
};
/**
* Pyodide instance for Python execution.
*
* 用于 Python 执行的 Pyodide 实例。
*/
let pyodide = null;
/**
* Language alias mapping.
*
* 语言别名映射。
*/
const langAlias = {
	kt: "kotlin",
	kotlin: "kotlin",
	go: "go",
	rust: "rust",
	rs: "rust",
	py: "python",
	python: "python"
};
/**
* List of supported languages.
*
* 支持的语言列表。
*/
const supportLang = [
	"kotlin",
	"go",
	"rust",
	"python"
];
/**
* Resolve language name from alias.
*
* 从别名解析语言名称。
*
* @param lang - Language or alias / 语言或别名
* @returns Resolved language name / 解析后的语言名称
*/
function resolveLang(lang) {
	return lang ? langAlias[lang] || lang : "";
}
/**
* Resolve code content from HTML element, ignoring specified nodes.
*
* 从 HTML 元素解析代码内容，忽略指定的节点。
*
* @param el - HTML element / HTML 元素
* @returns Code content / 代码内容
*/
function resolveCode(el) {
	const clone = el.cloneNode(true);
	clone.querySelectorAll(ignoredNodes.join(",")).forEach((node) => node.remove());
	return clone.textContent || "";
}
/**
* Resolve code information from HTML element.
*
* 从 HTML 元素解析代码信息。
*
* @param el - HTML element / HTML 元素
* @returns Object with language and code / 包含语言和代码的对象
*/
function resolveCodeInfo(el) {
	const wrapper = el.querySelector("div[class*=language-]");
	const lang = wrapper?.className.match(RE_LANGUAGE)?.[1];
	const codeEl = wrapper?.querySelector("pre");
	let code = "";
	if (codeEl) code = resolveCode(codeEl);
	return {
		lang: resolveLang(lang),
		code
	};
}
/**
* Composable for code REPL functionality.
*
* 代码 REPL 功能的组合式函数。
*
* This composable provides functionality to execute code in various languages
* (Kotlin, Go, Rust, Python) and manage the execution state.
*
* 该组合式函数提供在各种语言（Kotlin、Go、Rust、Python）中执行代码和管理执行状态的功能。
*
* @param el - Reference to the code element / 代码元素的引用
* @returns REPL state and methods / REPL 状态和方法
*
* @example
* ```vue
* <script setup>
* const codeEl = ref(null)
* const { onRunCode, stdout, stderr, loaded } = useCodeRepl(codeEl)
* <\/script>
* ```
*/
function useCodeRepl(el) {
	const lang = ref();
	const loaded = ref(true);
	const firstRun = ref(true);
	const finished = ref(true);
	const stdout = ref([]);
	const stderr = ref([]);
	const error = ref("");
	const backendVersion = ref("");
	onMounted(() => {
		if (el.value) lang.value = resolveCodeInfo(el.value).lang;
	});
	const executeMap = {
		kotlin: executeKotlin,
		go: executeGolang,
		rust: executeRust,
		python: executePython
	};
	function onCleanRun() {
		loaded.value = false;
		finished.value = false;
		stdout.value = [];
		stderr.value = [];
		error.value = "";
		firstRun.value = true;
		backendVersion.value = "";
	}
	async function onRunCode() {
		if (!el.value || !loaded.value) return;
		const info = resolveCodeInfo(el.value);
		lang.value = info.lang;
		if (!lang.value || !info.code || !supportLang.includes(lang.value)) return;
		if (firstRun.value) firstRun.value = false;
		loaded.value = false;
		finished.value = false;
		stdout.value = [];
		stderr.value = [];
		error.value = "";
		await executeMap[lang.value]?.(info.code);
	}
	async function executeGolang(code) {
		const res = await http.post(api.go, { code });
		backendVersion.value = `v${res.version}`;
		loaded.value = true;
		if (res.error) {
			error.value = res.error;
			finished.value = true;
			return;
		}
		const events = res.events || [];
		for (const event of events) if (event.kind === "stdout") {
			if (event.delay) await sleep(event.delay / 1e6);
			stdout.value.push(event.message);
		} else if (event.kind === "stderr") stderr.value.push(event.message);
		finished.value = true;
	}
	async function executeKotlin(code) {
		const filename = "File.kt";
		const res = await http.post(api.kotlin, {
			args: "",
			files: [{
				name: filename,
				publicId: "",
				text: code
			}]
		});
		backendVersion.value = `v${res.version}`;
		loaded.value = true;
		if (res.errors) {
			const errors = Array.isArray(res.errors[filename]) ? res.errors[filename] : [res.errors[filename]];
			if (errors.length) errors.forEach(({ message, severity }) => severity === "ERROR" && stderr.value.push(message));
		}
		stdout.value.push(res.text);
		finished.value = true;
	}
	async function executeRust(code) {
		await rustExecute(code, {
			onBegin: () => {
				loaded.value = true;
				finished.value = false;
				stdout.value = [];
				stderr.value = [];
				error.value = "";
				backendVersion.value = "release";
			},
			onError(message) {
				error.value = message;
			},
			onStdout(message) {
				stdout.value.push(message);
			},
			onStderr(message) {
				stderr.value.push(message);
			},
			onEnd: () => {
				finished.value = true;
			}
		});
	}
	async function executePython(code) {
		loaded.value = false;
		finished.value = false;
		if (pyodide === null) {
			const { loadPyodide, version } = await import(
				/* webpackChunkName: "pyodide" */
				"pyodide"
);
			pyodide = await loadPyodide({ indexURL: `https://cdn.jsdelivr.net/pyodide/v${version}/full/` });
		}
		pyodide.setStdout({ batched: (msg) => stdout.value.push(msg) });
		try {
			stdout.value.push(pyodide.runPython(code));
		} catch (e) {
			stderr.value.push(String(e));
		}
		loaded.value = true;
		finished.value = true;
	}
	return {
		onRunCode,
		onCleanRun,
		lang,
		backendVersion,
		firstRun,
		stderr,
		stdout,
		loaded,
		finished,
		error
	};
}
//#endregion
export { resolveCode, resolveCodeInfo, useCodeRepl };
